// @rivetkit/traces/dist/tsup/chunk-2D7JND4Z.js
// src/noop.ts
var U32_MAX = 4294967295;
var NOOP_SPAN = {
  spanId: new Uint8Array(8),
  traceId: new Uint8Array(16),
  isActive: () => false
};
function createEmptyOtlpExport() {
  return {
    resourceSpans: [
      {
        scopeSpans: [{ spans: [] }]
      }
    ]
  };
}
function createNoopTraces() {
  return {
    startSpan(_name, _options) {
      return NOOP_SPAN;
    },
    updateSpan(_handle, _options) {
    },
    setAttributes(_handle, _attributes) {
    },
    setStatus(_handle, _status) {
    },
    endSpan(_handle, _options) {
    },
    emitEvent(_handle, _name, _options) {
    },
    withSpan(_handle, fn) {
      return fn();
    },
    getCurrentSpan() {
      return null;
    },
    async flush() {
      return false;
    },
    async readRange(_options) {
      return {
        otlp: createEmptyOtlpExport(),
        clamped: false
      };
    },
    async readRangeWire(options) {
      return {
        startTimeMs: BigInt(options.startMs),
        endTimeMs: BigInt(options.endMs),
        limit: Math.max(0, Math.min(U32_MAX, Math.floor(options.limit))),
        clamped: false,
        baseChunks: [],
        chunks: []
      };
    }
  };
}

export {
  createNoopTraces
};
