// vbare/dist/index.mjs
// src/index.ts
var VersionedDataHandler = class {
  constructor(config) {
    this.config = config;
  }
  // Deserialize bytes of a given version into latest L
  deserialize(bytes, version) {
    let data = this.config.deserializeVersion(bytes, version);
    const converters = this.config.deserializeConverters();
    for (let i = Math.max(0, version - 1); i < converters.length; i++) {
      data = converters[i](data);
    }
    return data;
  }
  // Serialize an S (which may represent any version) to target `version`
  serialize(data, version) {
    let cur = data;
    const converters = this.config.serializeConverters();
    for (let i = Math.max(0, version - 1); i < converters.length; i++) {
      cur = converters[i](cur);
    }
    return this.config.serializeVersion(cur, version);
  }
  // Helpers that embed a u16 (LE) version prefix, like Rust
  serializeWithEmbeddedVersion(data, version) {
    const payload = this.serialize(data, version);
    const versionBytes = new Uint8Array(2);
    new DataView(versionBytes.buffer).setUint16(0, version, true);
    const out = new Uint8Array(2 + payload.length);
    out.set(versionBytes, 0);
    out.set(payload, 2);
    return out;
  }
  deserializeWithEmbeddedVersion(bytes) {
    if (bytes.length < 2) {
      throw new Error("payload too short for embedded version");
    }
    const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    const version = dv.getUint16(0, true);
    const payload = bytes.slice(2);
    return this.deserialize(payload, version);
  }
  // Utility kept for completeness in case callers want it.
  embedVersion(data) {
    const versionBytes = new Uint8Array(2);
    new DataView(versionBytes.buffer).setUint16(0, data.version, true);
    const result = new Uint8Array(versionBytes.length + data.data.length);
    result.set(versionBytes);
    result.set(data.data, versionBytes.length);
    return result;
  }
  extractVersion(bytes) {
    if (bytes.length < 2) {
      throw new Error("Invalid versioned data: too short");
    }
    const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    const version = dv.getUint16(0, true);
    const data = bytes.slice(2);
    return { version, data };
  }
};
function createVersionedDataHandler(config) {
  return new VersionedDataHandler(config);
}
export {
  VersionedDataHandler,
  createVersionedDataHandler
};
