// rivetkit/dist/schemas/file-system-driver/v1.ts
// @generated - post-processed by compile-bare.ts
import * as bare from "@rivetkit/bare-ts"

const config = /* @__PURE__ */ bare.Config({})

export type u64 = bigint
export type uint = bigint

function read0(bc: bare.ByteCursor): readonly string[] {
    const len = bare.readUintSafe(bc)
    if (len === 0) { return [] }
    const result = [bare.readString(bc)]
    for (let i = 1; i < len; i++) {
        result[i] = bare.readString(bc)
    }
    return result
}

function write0(bc: bare.ByteCursor, x: readonly string[]): void {
    bare.writeUintSafe(bc, x.length)
    for (let i = 0; i < x.length; i++) {
        bare.writeString(bc, x[i])
    }
}

export type ActorState = {
    readonly actorId: string,
    readonly name: string,
    readonly key: readonly string[],
    readonly persistedData: ArrayBuffer,
    readonly createdAt: u64,
}

export function readActorState(bc: bare.ByteCursor): ActorState {
    return {
        actorId: bare.readString(bc),
        name: bare.readString(bc),
        key: read0(bc),
        persistedData: bare.readData(bc),
        createdAt: bare.readU64(bc),
    }
}

export function writeActorState(bc: bare.ByteCursor, x: ActorState): void {
    bare.writeString(bc, x.actorId)
    bare.writeString(bc, x.name)
    write0(bc, x.key)
    bare.writeData(bc, x.persistedData)
    bare.writeU64(bc, x.createdAt)
}

export function encodeActorState(x: ActorState): Uint8Array {
    const bc = new bare.ByteCursor(
        new Uint8Array(config.initialBufferLength),
        config
    )
    writeActorState(bc, x)
    return new Uint8Array(bc.view.buffer, bc.view.byteOffset, bc.offset)
}

export function decodeActorState(bytes: Uint8Array): ActorState {
    const bc = new bare.ByteCursor(bytes, config)
    const result = readActorState(bc)
    if (bc.offset < bc.view.byteLength) {
        throw new bare.BareError(bc.offset, "remaining bytes")
    }
    return result
}

export type ActorAlarm = {
    readonly actorId: string,
    readonly timestamp: uint,
}

export function readActorAlarm(bc: bare.ByteCursor): ActorAlarm {
    return {
        actorId: bare.readString(bc),
        timestamp: bare.readUint(bc),
    }
}

export function writeActorAlarm(bc: bare.ByteCursor, x: ActorAlarm): void {
    bare.writeString(bc, x.actorId)
    bare.writeUint(bc, x.timestamp)
}

export function encodeActorAlarm(x: ActorAlarm): Uint8Array {
    const bc = new bare.ByteCursor(
        new Uint8Array(config.initialBufferLength),
        config
    )
    writeActorAlarm(bc, x)
    return new Uint8Array(bc.view.buffer, bc.view.byteOffset, bc.offset)
}

export function decodeActorAlarm(bytes: Uint8Array): ActorAlarm {
    const bc = new bare.ByteCursor(bytes, config)
    const result = readActorAlarm(bc)
    if (bc.offset < bc.view.byteLength) {
        throw new bare.BareError(bc.offset, "remaining bytes")
    }
    return result
}


function assert(condition: boolean, message?: string): asserts condition {
    if (!condition) throw new Error(message ?? "Assertion failed")
}
