// rivetkit/dist/schemas/file-system-driver/v3.ts
// @generated - post-processed by compile-bare.ts
import * as bare from "@rivetkit/bare-ts"

const config = /* @__PURE__ */ bare.Config({})

export type u64 = bigint
export type uint = bigint

export type ActorKvEntry = {
    readonly key: ArrayBuffer,
    readonly value: ArrayBuffer,
}

export function readActorKvEntry(bc: bare.ByteCursor): ActorKvEntry {
    return {
        key: bare.readData(bc),
        value: bare.readData(bc),
    }
}

export function writeActorKvEntry(bc: bare.ByteCursor, x: ActorKvEntry): void {
    bare.writeData(bc, x.key)
    bare.writeData(bc, x.value)
}

function read0(bc: bare.ByteCursor): readonly string[] {
    const len = bare.readUintSafe(bc)
    if (len === 0) { return [] }
    const result = [bare.readString(bc)]
    for (let i = 1; i < len; i++) {
        result[i] = bare.readString(bc)
    }
    return result
}

function write0(bc: bare.ByteCursor, x: readonly string[]): void {
    bare.writeUintSafe(bc, x.length)
    for (let i = 0; i < x.length; i++) {
        bare.writeString(bc, x[i])
    }
}

function read1(bc: bare.ByteCursor): readonly ActorKvEntry[] {
    const len = bare.readUintSafe(bc)
    if (len === 0) { return [] }
    const result = [readActorKvEntry(bc)]
    for (let i = 1; i < len; i++) {
        result[i] = readActorKvEntry(bc)
    }
    return result
}

function write1(bc: bare.ByteCursor, x: readonly ActorKvEntry[]): void {
    bare.writeUintSafe(bc, x.length)
    for (let i = 0; i < x.length; i++) {
        writeActorKvEntry(bc, x[i])
    }
}

function read2(bc: bare.ByteCursor): u64 | null {
    return bare.readBool(bc)
        ? bare.readU64(bc)
        : null
}

function write2(bc: bare.ByteCursor, x: u64 | null): void {
    bare.writeBool(bc, x !== null)
    if (x !== null) {
        bare.writeU64(bc, x)
    }
}

export type ActorState = {
    readonly actorId: string,
    readonly name: string,
    readonly key: readonly string[],
    readonly kvStorage: readonly ActorKvEntry[],
    readonly createdAt: u64,
    readonly startTs: u64 | null,
    readonly connectableTs: u64 | null,
    readonly sleepTs: u64 | null,
    readonly destroyTs: u64 | null,
}

export function readActorState(bc: bare.ByteCursor): ActorState {
    return {
        actorId: bare.readString(bc),
        name: bare.readString(bc),
        key: read0(bc),
        kvStorage: read1(bc),
        createdAt: bare.readU64(bc),
        startTs: read2(bc),
        connectableTs: read2(bc),
        sleepTs: read2(bc),
        destroyTs: read2(bc),
    }
}

export function writeActorState(bc: bare.ByteCursor, x: ActorState): void {
    bare.writeString(bc, x.actorId)
    bare.writeString(bc, x.name)
    write0(bc, x.key)
    write1(bc, x.kvStorage)
    bare.writeU64(bc, x.createdAt)
    write2(bc, x.startTs)
    write2(bc, x.connectableTs)
    write2(bc, x.sleepTs)
    write2(bc, x.destroyTs)
}

export function encodeActorState(x: ActorState): Uint8Array {
    const bc = new bare.ByteCursor(
        new Uint8Array(config.initialBufferLength),
        config
    )
    writeActorState(bc, x)
    return new Uint8Array(bc.view.buffer, bc.view.byteOffset, bc.offset)
}

export function decodeActorState(bytes: Uint8Array): ActorState {
    const bc = new bare.ByteCursor(bytes, config)
    const result = readActorState(bc)
    if (bc.offset < bc.view.byteLength) {
        throw new bare.BareError(bc.offset, "remaining bytes")
    }
    return result
}

export type ActorAlarm = {
    readonly actorId: string,
    readonly timestamp: uint,
}

export function readActorAlarm(bc: bare.ByteCursor): ActorAlarm {
    return {
        actorId: bare.readString(bc),
        timestamp: bare.readUint(bc),
    }
}

export function writeActorAlarm(bc: bare.ByteCursor, x: ActorAlarm): void {
    bare.writeString(bc, x.actorId)
    bare.writeUint(bc, x.timestamp)
}

export function encodeActorAlarm(x: ActorAlarm): Uint8Array {
    const bc = new bare.ByteCursor(
        new Uint8Array(config.initialBufferLength),
        config
    )
    writeActorAlarm(bc, x)
    return new Uint8Array(bc.view.buffer, bc.view.byteOffset, bc.offset)
}

export function decodeActorAlarm(bytes: Uint8Array): ActorAlarm {
    const bc = new bare.ByteCursor(bytes, config)
    const result = readActorAlarm(bc)
    if (bc.offset < bc.view.byteLength) {
        throw new bare.BareError(bc.offset, "remaining bytes")
    }
    return result
}


function assert(condition: boolean, message?: string): asserts condition {
    if (!condition) throw new Error(message ?? "Assertion failed")
}
