// rivetkit/dist/tsup/chunk-N4KRDJ56.js
// src/db/shared.ts
function isSqliteBindingValue(value) {
  if (value === null || typeof value === "number" || typeof value === "string" || typeof value === "bigint" || value instanceof Uint8Array) {
    return true;
  }
  if (Array.isArray(value)) {
    return value.every((item) => typeof item === "number");
  }
  return false;
}
function toSqliteBindings(args) {
  for (const value of args) {
    if (!isSqliteBindingValue(value)) {
      throw new Error(
        `unsupported sqlite binding type: ${typeof value}`
      );
    }
  }
  return args;
}
function createActorKvStore(kv) {
  return {
    get: async (key) => {
      const results = await kv.batchGet([key]);
      return results[0] ?? null;
    },
    getBatch: async (keys) => {
      return await kv.batchGet(keys);
    },
    put: async (key, value) => {
      await kv.batchPut([[key, value]]);
    },
    putBatch: async (entries) => {
      await kv.batchPut(entries);
    },
    deleteBatch: async (keys) => {
      await kv.batchDelete(keys);
    }
  };
}
var AsyncMutex = class {
  #locked = false;
  #waiting = [];
  async acquire() {
    while (this.#locked) {
      await new Promise((resolve) => this.#waiting.push(resolve));
    }
    this.#locked = true;
  }
  release() {
    this.#locked = false;
    const next = this.#waiting.shift();
    if (next) {
      next();
    }
  }
  async run(fn) {
    await this.acquire();
    try {
      return await fn();
    } finally {
      this.release();
    }
  }
};

export {
  toSqliteBindings,
  createActorKvStore,
  AsyncMutex
};
