// rivetkit/dist/tsup/chunk-ZFY5J2EP.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { newObj[key] = obj[key]; } } } newObj.default = obj; return newObj; } }



var _chunkSR3KQE7Qcjs = require('./chunk-SR3KQE7Q.cjs');



var _chunkHDQ2JUQTcjs = require('./chunk-HDQ2JUQT.cjs');

// src/db/drizzle/mod.ts
var mod_exports = {};
_chunkHDQ2JUQTcjs.__export.call(void 0, mod_exports, {
  blob: () => _sqlitecore.blob,
  check: () => _sqlitecore.check,
  db: () => db,
  defineConfig: () => defineConfig,
  extractUsedTable: () => _sqlitecore.extractUsedTable,
  foreignKey: () => _sqlitecore.foreignKey,
  getTableConfig: () => _sqlitecore.getTableConfig,
  getViewConfig: () => _sqlitecore.getViewConfig,
  index: () => _sqlitecore.index,
  integer: () => _sqlitecore.integer,
  numeric: () => _sqlitecore.numeric,
  primaryKey: () => _sqlitecore.primaryKey,
  real: () => _sqlitecore.real,
  sqliteTable: () => _sqlitecore.sqliteTable,
  sqliteTableCreator: () => _sqlitecore.sqliteTableCreator,
  sqliteView: () => _sqlitecore.sqliteView,
  text: () => _sqlitecore.text,
  unique: () => _sqlitecore.unique,
  uniqueIndex: () => _sqlitecore.uniqueIndex,
  uniqueKeyName: () => _sqlitecore.uniqueKeyName,
  view: () => _sqlitecore.view
});


var _sqliteproxy = require('drizzle-orm/sqlite-proxy');

// src/db/drizzle/sqlite-core.ts
var sqlite_core_exports = {};
_chunkHDQ2JUQTcjs.__export.call(void 0, sqlite_core_exports, {
  blob: () => _sqlitecore.blob,
  check: () => _sqlitecore.check,
  extractUsedTable: () => _sqlitecore.extractUsedTable,
  foreignKey: () => _sqlitecore.foreignKey,
  getTableConfig: () => _sqlitecore.getTableConfig,
  getViewConfig: () => _sqlitecore.getViewConfig,
  index: () => _sqlitecore.index,
  integer: () => _sqlitecore.integer,
  numeric: () => _sqlitecore.numeric,
  primaryKey: () => _sqlitecore.primaryKey,
  real: () => _sqlitecore.real,
  sqliteTable: () => _sqlitecore.sqliteTable,
  sqliteTableCreator: () => _sqlitecore.sqliteTableCreator,
  sqliteView: () => _sqlitecore.sqliteView,
  text: () => _sqlitecore.text,
  unique: () => _sqlitecore.unique,
  uniqueIndex: () => _sqlitecore.uniqueIndex,
  uniqueKeyName: () => _sqlitecore.uniqueKeyName,
  view: () => _sqlitecore.view
});
_chunkHDQ2JUQTcjs.__reExport.call(void 0, sqlite_core_exports, sqlite_core_star);
var _sqlitecore = require('drizzle-orm/sqlite-core'); var sqlite_core_star = _interopRequireWildcard(_sqlitecore);






















// src/db/drizzle/mod.ts
_chunkHDQ2JUQTcjs.__reExport.call(void 0, mod_exports, sqlite_core_exports);
var _drizzlekit = require('drizzle-kit');
function defineConfig(config) {
  return _drizzlekit.defineConfig.call(void 0, {
    dialect: "sqlite",
    driver: "durable-sqlite",
    ...config
  });
}
function createProxyCallback(waDb, mutex, isClosed) {
  return async (sql, params, method) => {
    return mutex.run(async () => {
      if (isClosed()) {
        throw new Error("database is closed");
      }
      if (method === "run") {
        await waDb.run(sql, _chunkSR3KQE7Qcjs.toSqliteBindings.call(void 0, params));
        return { rows: [] };
      }
      const result = await waDb.query(sql, _chunkSR3KQE7Qcjs.toSqliteBindings.call(void 0, params));
      if (method === "get") {
        return { rows: result.rows[0] };
      }
      return { rows: result.rows };
    });
  };
}
async function runInlineMigrations(waDb, mutex, migrations) {
  await mutex.run(
    () => waDb.exec(`
		CREATE TABLE IF NOT EXISTS __drizzle_migrations (
			id INTEGER PRIMARY KEY AUTOINCREMENT,
			hash TEXT NOT NULL,
			created_at INTEGER
		)
	`)
  );
  let lastCreatedAt = 0;
  await mutex.run(
    () => waDb.exec(
      "SELECT id, hash, created_at FROM __drizzle_migrations ORDER BY created_at DESC LIMIT 1",
      (row) => {
        lastCreatedAt = Number(row[2]) || 0;
      }
    )
  );
  const journal = migrations.journal;
  if (!(journal == null ? void 0 : journal.entries)) return;
  for (const entry of journal.entries) {
    if (entry.when <= lastCreatedAt) continue;
    const migrationKey = `m${String(entry.idx).padStart(4, "0")}`;
    const sql = migrations.migrations[migrationKey];
    if (!sql) continue;
    await mutex.run(() => waDb.exec(sql));
    await mutex.run(
      () => waDb.run(
        "INSERT INTO __drizzle_migrations (hash, created_at) VALUES (?, ?)",
        [entry.tag, entry.when]
      )
    );
  }
}
function db(config) {
  let waDbInstance = null;
  const mutex = new (0, _chunkSR3KQE7Qcjs.AsyncMutex)();
  return {
    createClient: async (ctx) => {
      if (!ctx.sqliteVfs) {
        throw new Error(
          "SqliteVfs instance not provided in context. The driver must provide a sqliteVfs instance."
        );
      }
      const kvStore = _chunkSR3KQE7Qcjs.createActorKvStore.call(void 0, ctx.kv);
      const waDb = await ctx.sqliteVfs.open(ctx.actorId, kvStore);
      waDbInstance = waDb;
      let closed = false;
      const ensureOpen = () => {
        if (closed) {
          throw new Error("database is closed");
        }
      };
      const callback = createProxyCallback(waDb, mutex, () => closed);
      const client = _sqliteproxy.drizzle.call(void 0, callback, config);
      return Object.assign(client, {
        execute: async (query, ...args) => {
          return mutex.run(async () => {
            ensureOpen();
            if (args.length > 0) {
              const result = await waDb.query(
                query,
                _chunkSR3KQE7Qcjs.toSqliteBindings.call(void 0, args)
              );
              return result.rows.map((row) => {
                const obj = {};
                for (let i = 0; i < result.columns.length; i++) {
                  obj[result.columns[i]] = row[i];
                }
                return obj;
              });
            }
            const results = [];
            let columnNames = null;
            await waDb.exec(
              query,
              (row, columns) => {
                if (!columnNames) {
                  columnNames = columns;
                }
                const obj = {};
                for (let i = 0; i < row.length; i++) {
                  obj[columnNames[i]] = row[i];
                }
                results.push(obj);
              }
            );
            return results;
          });
        },
        close: async () => {
          await mutex.run(async () => {
            if (closed) {
              return;
            }
            closed = true;
            await waDb.close();
            waDbInstance = null;
          });
        }
      });
    },
    onMigrate: async (_client) => {
      if ((config == null ? void 0 : config.migrations) && waDbInstance) {
        await runInlineMigrations(
          waDbInstance,
          mutex,
          config.migrations
        );
      }
    },
    onDestroy: async (client) => {
      await client.close();
    }
  };
}
























exports.blob = _sqlitecore.blob; exports.check = _sqlitecore.check; exports.extractUsedTable = _sqlitecore.extractUsedTable; exports.foreignKey = _sqlitecore.foreignKey; exports.getTableConfig = _sqlitecore.getTableConfig; exports.getViewConfig = _sqlitecore.getViewConfig; exports.index = _sqlitecore.index; exports.integer = _sqlitecore.integer; exports.numeric = _sqlitecore.numeric; exports.primaryKey = _sqlitecore.primaryKey; exports.real = _sqlitecore.real; exports.sqliteTable = _sqlitecore.sqliteTable; exports.sqliteTableCreator = _sqlitecore.sqliteTableCreator; exports.sqliteView = _sqlitecore.sqliteView; exports.text = _sqlitecore.text; exports.unique = _sqlitecore.unique; exports.uniqueIndex = _sqlitecore.uniqueIndex; exports.uniqueKeyName = _sqlitecore.uniqueKeyName; exports.view = _sqlitecore.view; exports.defineConfig = defineConfig; exports.db = db; exports.mod_exports = mod_exports;
