// rivetkit/dist/tsup/chunk-6LJAZ5R4.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});



var _chunkSR3KQE7Qcjs = require('./chunk-SR3KQE7Q.cjs');

// src/db/mod.ts
function db({
  onMigrate
} = {}) {
  return {
    createClient: async (ctx) => {
      const override = ctx.overrideRawDatabaseClient ? await ctx.overrideRawDatabaseClient() : void 0;
      if (override) {
        return {
          execute: async (query, ...args) => {
            return await override.exec(query, ...args);
          },
          close: async () => {
          }
        };
      }
      if (!ctx.sqliteVfs) {
        throw new Error("SqliteVfs instance not provided in context. The driver must provide a sqliteVfs instance.");
      }
      const kvStore = _chunkSR3KQE7Qcjs.createActorKvStore.call(void 0, ctx.kv);
      const db2 = await ctx.sqliteVfs.open(ctx.actorId, kvStore);
      let closed = false;
      const mutex = new (0, _chunkSR3KQE7Qcjs.AsyncMutex)();
      const ensureOpen = () => {
        if (closed) {
          throw new Error("database is closed");
        }
      };
      return {
        execute: async (query, ...args) => {
          return await mutex.run(async () => {
            ensureOpen();
            if (args.length > 0) {
              const bindings = _chunkSR3KQE7Qcjs.toSqliteBindings.call(void 0, args);
              const token = query.trimStart().slice(0, 16).toUpperCase();
              const returnsRows = token.startsWith("SELECT") || token.startsWith("PRAGMA") || token.startsWith("WITH");
              if (returnsRows) {
                const { rows, columns } = await db2.query(query, bindings);
                return rows.map((row) => {
                  const rowObj = {};
                  for (let i = 0; i < columns.length; i++) {
                    rowObj[columns[i]] = row[i];
                  }
                  return rowObj;
                });
              }
              await db2.run(query, bindings);
              return [];
            }
            const results = [];
            let columnNames = null;
            await db2.exec(query, (row, columns) => {
              if (!columnNames) {
                columnNames = columns;
              }
              const rowObj = {};
              for (let i = 0; i < row.length; i++) {
                rowObj[columnNames[i]] = row[i];
              }
              results.push(rowObj);
            });
            return results;
          });
        },
        close: async () => {
          await mutex.run(async () => {
            if (closed) {
              return;
            }
            closed = true;
            await db2.close();
          });
        }
      };
    },
    onMigrate: async (client) => {
      if (onMigrate) {
        await onMigrate(client);
      }
    },
    onDestroy: async (client) => {
      await client.close();
    }
  };
}



exports.db = db;
