// rivetkit/dist/tsup/chunk-KJXZTBFN.js
import {
  createClientWithDriver
} from "./chunk-DOEWKM5V.js";
import {
  ACTOR_VERSIONED,
  ActionContext,
  BeforeConnectContext,
  CONN_CONNECTED_SYMBOL,
  CONN_DRIVER_SYMBOL,
  CONN_SEND_MESSAGE_SYMBOL,
  CONN_SPEAKS_RIVETKIT_SYMBOL,
  CONN_STATE_MANAGER_SYMBOL,
  CONN_VERSIONED,
  CURRENT_VERSION,
  CURRENT_VERSION2,
  CachedSerializer,
  Conn,
  ConnectContext,
  CreateConnStateContext,
  PATH_CONNECT,
  PATH_INSPECTOR_CONNECT,
  PATH_WEBSOCKET_PREFIX,
  QUEUE_MESSAGE_VERSIONED,
  QUEUE_METADATA_VERSIONED,
  RequestContext,
  TO_CLIENT_VERSIONED,
  ToClientSchema,
  WebSocketContext,
  createHttpDriver,
  getInitialActorKvState,
  getNodeCrypto,
  getNodeFs,
  getNodeFsSync,
  getNodeOs,
  getNodePath,
  getRequireFn,
  handleAction,
  handleQueueSend,
  handleRawRequest,
  handleRouteError,
  handleRouteNotFound,
  importNodeDependencies,
  importSqliteVfs,
  inspectorLogger,
  loggerMiddleware,
  parseWebSocketProtocols,
  processMessage,
  routeWebSocket,
  serializeActorKey,
  timingSafeEqual
} from "./chunk-W735O34B.js";
import {
  ActorContext,
  DeadlineError,
  KEYS,
  assertUnreachable,
  deadline,
  decodeQueueMessageKey,
  generateSecureToken,
  getEventCanSubscribe,
  getQueueCanPublish,
  getRunFunction,
  getRunInspectorConfig,
  isConnStatePath,
  isStatePath,
  loggerWithoutContext,
  makeConnKey,
  makeQueueMessageKey,
  queueMessagesPrefix,
  queueMetadataKey,
  tracesStoragePrefix
} from "./chunk-SFC4YXTL.js";
import {
  EXTRA_ERROR_LOG,
  SinglePromiseQueue,
  VERSION,
  arrayBuffersEqual,
  bufferToArrayBuffer,
  getBaseLogger,
  getIncludeTarget,
  getLogger,
  getRivetExperimentalOtel,
  isCborSerializable,
  isDev,
  noopNext,
  promiseWithResolvers,
  setLongTimeout,
  stringifyError
} from "./chunk-ITLKCAXM.js";
import {
  ActionNotFound,
  ActionTimedOut,
  ActorAborted,
  ActorDuplicateKey,
  ActorStopping,
  DatabaseNotEnabled,
  Forbidden,
  InternalError,
  InvalidCanPublishResponse,
  InvalidCanSubscribeResponse,
  InvalidRequestHandlerResponse,
  InvalidStateType,
  OutgoingMessageTooLong,
  QueueFull,
  QueueMessageInvalid,
  QueueMessageTooLarge,
  RequestHandlerNotDefined,
  StateNotEnabled,
  VarsNotEnabled
} from "./chunk-7HTNH26M.js";

// src/actor/instance/mod.ts
import {
  createNoopTraces,
  createTraces
} from "@rivetkit/traces";
import invariant2 from "invariant";

// src/inspector/actor-inspector.ts
import * as cbor from "cbor-x";
import { createNanoEvents } from "nanoevents";
var ActorInspector = class {
  constructor(actor) {
    this.actor = actor;
    var _a, _b;
    this.#lastQueueSize = ((_a = actor.queueManager) == null ? void 0 : _a.size) ?? 0;
    const runInspector = getRunInspectorConfig(actor.config.run);
    this.#workflowInspector = runInspector == null ? void 0 : runInspector.workflow;
    if ((_b = this.#workflowInspector) == null ? void 0 : _b.onHistoryUpdated) {
      this.#workflowInspector.onHistoryUpdated((history) => {
        this.emitter.emit(
          "workflowHistoryUpdated",
          history
        );
      });
    }
  }
  emitter = createNanoEvents();
  #lastQueueSize = 0;
  #workflowInspector;
  getQueueSize() {
    return this.#lastQueueSize;
  }
  async getQueueStatus(limit) {
    const maxSize = this.actor.config.options.maxQueueSize;
    const safeLimit = Math.max(0, Math.floor(limit));
    const messages = await this.actor.queueManager.getMessages();
    const sorted = messages.sort(
      (a, b) => a.createdAt - b.createdAt
    );
    const limited = safeLimit > 0 ? sorted.slice(0, safeLimit) : [];
    return {
      size: BigInt(this.#lastQueueSize),
      maxSize: BigInt(maxSize),
      truncated: sorted.length > limited.length,
      messages: limited.map((message) => ({
        id: message.id,
        name: message.name,
        createdAtMs: BigInt(message.createdAt)
      }))
    };
  }
  updateQueueSize(size) {
    if (this.#lastQueueSize === size) {
      return;
    }
    this.#lastQueueSize = size;
    this.emitter.emit("queueUpdated");
  }
  isWorkflowEnabled() {
    return this.#workflowInspector !== void 0;
  }
  getWorkflowHistory() {
    if (!this.#workflowInspector) {
      return null;
    }
    const history = this.#workflowInspector.getHistory();
    return history ?? null;
  }
  // actor accessor methods
  isDatabaseEnabled() {
    try {
      return this.actor.db !== void 0;
    } catch {
      return false;
    }
  }
  async getDatabaseSchema() {
    var _a;
    if (!this.isDatabaseEnabled()) {
      throw new DatabaseNotEnabled();
    }
    const db = this.actor.db;
    const tables = await db.execute(
      "SELECT name, type FROM sqlite_master WHERE type IN ('table', 'view') AND name NOT LIKE 'sqlite_%' AND name NOT LIKE '__drizzle_%'"
    );
    const tableInfos = [];
    for (const table of tables) {
      const quoted = `"${escapeDoubleQuotes(table.name)}"`;
      const sample = await db.execute(
        `SELECT * FROM ${quoted} LIMIT 1`
      );
      const countResult = await db.execute(
        `SELECT COUNT(*) as count FROM ${quoted}`
      );
      const columnNames = (sample == null ? void 0 : sample[0]) ? Object.keys(sample[0]) : [];
      tableInfos.push({
        table: { schema: "main", name: table.name, type: table.type },
        columns: columnNames.map((name, cid) => ({
          cid,
          name,
          type: "",
          notnull: 0,
          dflt_value: null,
          pk: 0
        })),
        foreignKeys: [],
        records: ((_a = countResult == null ? void 0 : countResult[0]) == null ? void 0 : _a.count) ?? 0
      });
    }
    return bufferToArrayBuffer(cbor.encode({ tables: tableInfos }));
  }
  async getDatabaseTableRows(table, limit, offset) {
    if (!this.isDatabaseEnabled()) {
      throw new DatabaseNotEnabled();
    }
    const db = this.actor.db;
    const safeLimit = Math.max(0, Math.min(Math.floor(limit), 500));
    const safeOffset = Math.max(0, Math.floor(offset));
    const quoted = `"${escapeDoubleQuotes(table)}"`;
    const result = await db.execute(
      `SELECT * FROM ${quoted} LIMIT ? OFFSET ?`,
      safeLimit,
      safeOffset
    );
    return bufferToArrayBuffer(cbor.encode(result));
  }
  isStateEnabled() {
    return this.actor.stateEnabled;
  }
  getState() {
    if (!this.actor.stateEnabled) {
      throw new StateNotEnabled();
    }
    return bufferToArrayBuffer(
      cbor.encode(this.actor.stateManager.persistRaw.state)
    );
  }
  getRpcs() {
    return this.actor.actions;
  }
  getConnections() {
    return Array.from(
      this.actor.connectionManager.connections.entries()
    ).map(([id, conn]) => {
      var _a, _b;
      const connStateManager = conn[CONN_STATE_MANAGER_SYMBOL];
      return {
        type: (_a = conn[CONN_DRIVER_SYMBOL]) == null ? void 0 : _a.type,
        id,
        details: bufferToArrayBuffer(
          cbor.encode({
            type: (_b = conn[CONN_DRIVER_SYMBOL]) == null ? void 0 : _b.type,
            params: conn.params,
            stateEnabled: connStateManager.stateEnabled,
            state: connStateManager.stateEnabled ? connStateManager.state : void 0,
            subscriptions: conn.subscriptions.size,
            isHibernatable: conn.isHibernatable
            // TODO: Include underlying hibernatable metadata +
            // path + headers
          })
        )
      };
    });
  }
  async setState(state) {
    if (!this.actor.stateEnabled) {
      throw new StateNotEnabled();
    }
    this.actor.stateManager.state = cbor.decode(Buffer.from(state));
    await this.actor.stateManager.saveState({ immediate: true });
  }
  async executeAction(name, params) {
    const conn = await this.actor.connectionManager.prepareAndConnectConn(
      createHttpDriver(),
      // TODO: This may cause issues
      void 0,
      void 0,
      void 0,
      void 0
    );
    try {
      return bufferToArrayBuffer(
        cbor.encode(
          await this.actor.executeAction(
            new ActionContext(this.actor, conn),
            name,
            cbor.decode(Buffer.from(params))
          )
        )
      );
    } finally {
      conn.disconnect();
    }
  }
  // JSON-native methods for the HTTP inspector API. These return raw JS
  // objects suitable for JSON serialization instead of CBOR-encoded buffers.
  getStateJson() {
    if (!this.actor.stateEnabled) {
      throw new StateNotEnabled();
    }
    return this.actor.stateManager.persistRaw.state;
  }
  async setStateJson(state) {
    if (!this.actor.stateEnabled) {
      throw new StateNotEnabled();
    }
    this.actor.stateManager.state = state;
    await this.actor.stateManager.saveState({ immediate: true });
  }
  getConnectionsJson() {
    return Array.from(
      this.actor.connectionManager.connections.entries()
    ).map(([id, conn]) => {
      var _a, _b;
      const connStateManager = conn[CONN_STATE_MANAGER_SYMBOL];
      return {
        type: (_a = conn[CONN_DRIVER_SYMBOL]) == null ? void 0 : _a.type,
        id,
        details: {
          type: (_b = conn[CONN_DRIVER_SYMBOL]) == null ? void 0 : _b.type,
          params: conn.params,
          stateEnabled: connStateManager.stateEnabled,
          state: connStateManager.stateEnabled ? connStateManager.state : void 0,
          subscriptions: conn.subscriptions.size,
          isHibernatable: conn.isHibernatable
        }
      };
    });
  }
  async executeActionJson(name, args) {
    const conn = await this.actor.connectionManager.prepareAndConnectConn(
      createHttpDriver(),
      void 0,
      void 0,
      void 0,
      void 0
    );
    try {
      return await this.actor.executeAction(
        new ActionContext(this.actor, conn),
        name,
        args
      );
    } finally {
      conn.disconnect();
    }
  }
  async getTracesJson(options) {
    const result = await this.actor.traces.readRange(options);
    return result;
  }
  getWorkflowHistoryJson() {
    const bigIntReplacer = (_key, value) => typeof value === "bigint" ? Number(value) : value;
    const history = this.getWorkflowHistory();
    const safeHistory = history ? JSON.parse(JSON.stringify(history, bigIntReplacer)) : null;
    return {
      history: safeHistory,
      isWorkflowEnabled: this.isWorkflowEnabled()
    };
  }
  getQueueStatusJson(limit) {
    return this.getQueueStatus(limit).then((status) => ({
      size: Number(status.size),
      maxSize: Number(status.maxSize),
      truncated: status.truncated,
      messages: status.messages.map((m) => ({
        id: Number(m.id),
        name: m.name,
        createdAtMs: Number(m.createdAtMs)
      }))
    }));
  }
};
function escapeDoubleQuotes(value) {
  return value.replace(/"/g, '""');
}

// src/actor/conn/persisted.ts
import * as cbor2 from "cbor-x";
function convertConnToBarePersistedConn(persist) {
  return {
    id: persist.id,
    parameters: bufferToArrayBuffer(cbor2.encode(persist.parameters)),
    state: bufferToArrayBuffer(cbor2.encode(persist.state)),
    subscriptions: persist.subscriptions.map((sub) => ({
      eventName: sub.eventName
    })),
    gatewayId: persist.gatewayId,
    requestId: persist.requestId,
    serverMessageIndex: persist.serverMessageIndex,
    clientMessageIndex: persist.clientMessageIndex,
    requestPath: persist.requestPath,
    requestHeaders: new Map(Object.entries(persist.requestHeaders))
  };
}
function convertConnFromBarePersistedConn(bareData) {
  return {
    id: bareData.id,
    parameters: cbor2.decode(new Uint8Array(bareData.parameters)),
    state: cbor2.decode(new Uint8Array(bareData.state)),
    subscriptions: bareData.subscriptions.map((sub) => ({
      eventName: sub.eventName
    })),
    gatewayId: bareData.gatewayId,
    requestId: bareData.requestId,
    serverMessageIndex: bareData.serverMessageIndex,
    clientMessageIndex: bareData.clientMessageIndex,
    requestPath: bareData.requestPath,
    requestHeaders: Object.fromEntries(bareData.requestHeaders)
  };
}

// src/actor/schedule.ts
var Schedule = class {
  #actor;
  constructor(actor) {
    this.#actor = actor;
  }
  async after(duration, fn, ...args) {
    await this.#actor.scheduleEvent(Date.now() + duration, fn, args);
  }
  async at(timestamp, fn, ...args) {
    await this.#actor.scheduleEvent(timestamp, fn, args);
  }
};

// src/actor/instance/connection-manager.ts
import invariant from "invariant";
var ConnectionManager = class {
  #actor;
  #connections = /* @__PURE__ */ new Map();
  #pendingDisconnectCount = 0;
  /** Connections that have had their state changed and need to be persisted. */
  #connsWithPersistChanged = /* @__PURE__ */ new Set();
  constructor(actor) {
    this.#actor = actor;
  }
  get connections() {
    return this.#connections;
  }
  getConnForId(id) {
    return this.#connections.get(id);
  }
  get connsWithPersistChanged() {
    return this.#connsWithPersistChanged;
  }
  get pendingDisconnectCount() {
    return this.#pendingDisconnectCount;
  }
  clearConnWithPersistChanged() {
    this.#connsWithPersistChanged.clear();
  }
  markConnWithPersistChanged(conn) {
    invariant(
      conn.isHibernatable,
      "cannot mark non-hibernatable conn for persist"
    );
    this.#actor.rLog.debug({
      msg: "marked connection as changed",
      connId: conn.id,
      totalChanged: this.#connsWithPersistChanged.size
    });
    this.#connsWithPersistChanged.add(conn.id);
    this.#actor.stateManager.savePersistThrottled();
  }
  // MARK: - Connection Lifecycle
  /**
   * Handles pre-connection logic (i.e. auth & create state) before actually connecting the connection.
   */
  async prepareConn(driver, params, request, requestPath, requestHeaders, isHibernatable, isRestoringHibernatable) {
    this.#actor.assertReady();
    if (isRestoringHibernatable) {
      return this.#reconnectHibernatableConn(driver);
    }
    if (this.#actor.config.onBeforeConnect) {
      const ctx = new BeforeConnectContext(this.#actor, request);
      await this.#actor.runInTraceSpan(
        "actor.onBeforeConnect",
        {
          "rivet.conn.type": driver.type
        },
        () => this.#actor.config.onBeforeConnect(ctx, params)
      );
    }
    let connState;
    if (this.#actor.connStateEnabled) {
      connState = await this.#createConnState(params, request);
    }
    let connData;
    if (isHibernatable) {
      const hibernatable = driver.hibernatable;
      invariant(hibernatable, "must have hibernatable");
      invariant(requestPath, "missing requestPath for hibernatable ws");
      invariant(
        requestHeaders,
        "missing requestHeaders for hibernatable ws"
      );
      connData = {
        hibernatable: {
          id: crypto.randomUUID(),
          parameters: params,
          state: connState,
          subscriptions: [],
          gatewayId: hibernatable.gatewayId,
          requestId: hibernatable.requestId,
          clientMessageIndex: 0,
          // First message index will be 1, so we start at 0
          serverMessageIndex: 0,
          requestPath,
          requestHeaders
        }
      };
    } else {
      connData = {
        ephemeral: {
          id: crypto.randomUUID(),
          parameters: params,
          state: connState
        }
      };
    }
    const conn = new Conn(this.#actor, connData);
    conn[CONN_DRIVER_SYMBOL] = driver;
    return conn;
  }
  /**
   * Adds a connection form prepareConn to the actor and calls onConnect.
   *
   * This method is intentionally not async since it needs to be called in
   * `onOpen` for WebSockets. If this is async, the order of open events will
   * be messed up and cause race conditions that can drop WebSocket messages.
   * So all async work in prepareConn.
   */
  connectConn(conn) {
    invariant(!this.#connections.has(conn.id), "conn already connected");
    this.#connections.set(conn.id, conn);
    if (this.#actor.driver.onCreateConn) {
      this.#actor.driver.onCreateConn(conn);
    }
    if (conn.isHibernatable) {
      this.markConnWithPersistChanged(conn);
    }
    this.#callOnConnect(conn);
    this.#actor.inspector.emitter.emit("connectionsUpdated");
    this.#actor.resetSleepTimer();
    conn[CONN_CONNECTED_SYMBOL] = true;
    if (conn[CONN_SPEAKS_RIVETKIT_SYMBOL]) {
      const initData = { actorId: this.#actor.id, connectionId: conn.id };
      conn[CONN_SEND_MESSAGE_SYMBOL](
        new CachedSerializer(
          initData,
          TO_CLIENT_VERSIONED,
          CURRENT_VERSION,
          ToClientSchema,
          // JSON: identity conversion (no nested data to encode)
          (value) => ({
            body: {
              tag: "Init",
              val: value
            }
          }),
          // BARE/CBOR: identity conversion (no nested data to encode)
          (value) => ({
            body: {
              tag: "Init",
              val: value
            }
          })
        )
      );
    }
  }
  #reconnectHibernatableConn(driver) {
    invariant(driver.hibernatable, "missing requestIdBuf");
    const existingConn = this.findHibernatableConn(
      driver.hibernatable.gatewayId,
      driver.hibernatable.requestId
    );
    invariant(
      existingConn,
      "cannot find connection for restoring connection"
    );
    this.#actor.rLog.debug({
      msg: "reconnecting hibernatable websocket connection",
      connectionId: existingConn.id
    });
    if (existingConn[CONN_DRIVER_SYMBOL]) {
      this.#disconnectExistingDriver(existingConn);
    }
    existingConn[CONN_DRIVER_SYMBOL] = driver;
    this.#actor.resetSleepTimer();
    existingConn[CONN_CONNECTED_SYMBOL] = true;
    this.#actor.inspector.emitter.emit("connectionsUpdated");
    return existingConn;
  }
  #disconnectExistingDriver(conn) {
    const driver = conn[CONN_DRIVER_SYMBOL];
    if (driver == null ? void 0 : driver.disconnect) {
      driver.disconnect(
        this.#actor,
        conn,
        "Reconnecting hibernatable websocket with new driver state"
      );
    }
  }
  /**
   * Handle connection disconnection.
   *
   * This is called by `Conn.disconnect`. This should not call `Conn.disconnect.`
   */
  async connDisconnected(conn) {
    var _a;
    this.#connections.delete(conn.id);
    this.#actor.rLog.debug({ msg: "removed conn", connId: conn.id });
    if (this.#actor.driver.onDestroyConn) {
      this.#actor.driver.onDestroyConn(conn);
    }
    for (const eventName of [...conn.subscriptions.values()]) {
      this.#actor.eventManager.removeSubscription(eventName, conn, true);
    }
    this.#actor.inspector.emitter.emit("connectionsUpdated");
    this.#pendingDisconnectCount += 1;
    const attributes = {
      "rivet.conn.id": conn.id,
      "rivet.conn.type": (_a = conn[CONN_DRIVER_SYMBOL]) == null ? void 0 : _a.type,
      "rivet.conn.hibernatable": conn.isHibernatable
    };
    const span = this.#actor.startTraceSpan(
      "actor.onDisconnect",
      attributes
    );
    try {
      if (this.#actor.config.onDisconnect) {
        const result = this.#actor.traces.withSpan(
          span,
          () => this.#actor.config.onDisconnect(
            this.#actor.actorContext,
            conn
          )
        );
        this.#actor.emitTraceEvent(
          "connection.disconnect",
          attributes,
          span
        );
        if (result instanceof Promise) {
          await result;
        }
        this.#actor.endTraceSpan(span, { code: "OK" });
      } else {
        this.#actor.emitTraceEvent(
          "connection.disconnect",
          attributes,
          span
        );
        this.#actor.endTraceSpan(span, { code: "OK" });
      }
    } catch (error) {
      this.#actor.endTraceSpan(span, {
        code: "ERROR",
        message: stringifyError(error)
      });
      this.#actor.rLog.error({
        msg: "error in `onDisconnect`",
        error: stringifyError(error)
      });
    } finally {
      this.#connsWithPersistChanged.delete(conn.id);
      if (conn.isHibernatable) {
        const key = makeConnKey(conn.id);
        try {
          await this.#actor.driver.kvBatchDelete(this.#actor.id, [key]);
          this.#actor.rLog.debug({
            msg: "removed connection from KV",
            connId: conn.id
          });
        } catch (err) {
          this.#actor.rLog.error({
            msg: "kvBatchDelete failed for conn",
            err: stringifyError(err)
          });
        }
      }
      this.#pendingDisconnectCount = Math.max(
        0,
        this.#pendingDisconnectCount - 1
      );
      this.#actor.resetSleepTimer();
    }
  }
  async cleanupPersistedHibernatableConnections(reason) {
    const staleConnections = Array.from(this.#connections.values()).filter(
      (conn) => conn.isHibernatable && conn[CONN_DRIVER_SYMBOL] === void 0
    );
    if (staleConnections.length === 0) {
      return 0;
    }
    this.#actor.rLog.info({
      msg: "cleaning up persisted hibernatable connections",
      reason: reason ?? "unknown",
      count: staleConnections.length
    });
    for (const conn of staleConnections) {
      await this.connDisconnected(conn);
    }
    return staleConnections.length;
  }
  /**
   * Utilify function for call sites that don't need a separate prepare and connect phase.
   */
  async prepareAndConnectConn(driver, params, request, requestPath, requestHeaders) {
    const conn = await this.prepareConn(
      driver,
      params,
      request,
      requestPath,
      requestHeaders,
      false,
      false
    );
    this.connectConn(conn);
    return conn;
  }
  // MARK: - Persistence
  /**
   * Restores connections from persisted data during actor initialization.
   */
  restoreConnections(connections) {
    for (const connPersist of connections) {
      const conn = new Conn(this.#actor, {
        hibernatable: connPersist
      });
      this.#connections.set(conn.id, conn);
      if (this.#actor.driver.onCreateConn) {
        this.#actor.driver.onCreateConn(conn);
      }
      for (const sub of connPersist.subscriptions) {
        this.#actor.eventManager.addSubscription(
          sub.eventName,
          conn,
          true
        );
      }
    }
  }
  // MARK: - Private Helpers
  findHibernatableConn(gatewayIdBuf, requestIdBuf) {
    return Array.from(this.#connections.values()).find((conn) => {
      const connStateManager = conn[CONN_STATE_MANAGER_SYMBOL];
      const h = connStateManager.hibernatableDataRaw;
      return h && arrayBuffersEqual(h.gatewayId, gatewayIdBuf) && arrayBuffersEqual(h.requestId, requestIdBuf);
    });
  }
  async #createConnState(params, request) {
    if ("createConnState" in this.#actor.config) {
      const createConnState = this.#actor.config.createConnState;
      const ctx = new CreateConnStateContext(this.#actor, request);
      return await this.#actor.runInTraceSpan(
        "actor.createConnState",
        void 0,
        () => {
          const dataOrPromise = createConnState(ctx, params);
          if (dataOrPromise instanceof Promise) {
            return deadline(
              dataOrPromise,
              this.#actor.config.options.createConnStateTimeout
            );
          }
          return dataOrPromise;
        }
      );
    } else if ("connState" in this.#actor.config) {
      return structuredClone(this.#actor.config.connState);
    }
    throw new Error(
      "Could not create connection state from 'createConnState' or 'connState'"
    );
  }
  #callOnConnect(conn) {
    var _a;
    const attributes = {
      "rivet.conn.id": conn.id,
      "rivet.conn.type": (_a = conn[CONN_DRIVER_SYMBOL]) == null ? void 0 : _a.type,
      "rivet.conn.hibernatable": conn.isHibernatable
    };
    const span = this.#actor.startTraceSpan("actor.onConnect", attributes);
    try {
      if (this.#actor.config.onConnect) {
        const ctx = new ConnectContext(this.#actor, conn);
        const result = this.#actor.traces.withSpan(
          span,
          () => this.#actor.config.onConnect(ctx, conn)
        );
        this.#actor.emitTraceEvent(
          "connection.connect",
          attributes,
          span
        );
        if (result instanceof Promise) {
          deadline(
            result,
            this.#actor.config.options.onConnectTimeout
          ).then(() => {
            this.#actor.endTraceSpan(span, { code: "OK" });
          }).catch((error) => {
            this.#actor.endTraceSpan(span, {
              code: "ERROR",
              message: stringifyError(error)
            });
            this.#actor.rLog.error({
              msg: "error in `onConnect`, closing socket",
              error
            });
            conn == null ? void 0 : conn.disconnect("`onConnect` failed");
          });
          return;
        }
      }
      this.#actor.emitTraceEvent("connection.connect", attributes, span);
      this.#actor.endTraceSpan(span, { code: "OK" });
    } catch (error) {
      this.#actor.endTraceSpan(span, {
        code: "ERROR",
        message: stringifyError(error)
      });
      this.#actor.rLog.error({
        msg: "error in `onConnect`",
        error: stringifyError(error)
      });
      conn == null ? void 0 : conn.disconnect("`onConnect` failed");
    }
  }
};

// src/actor/instance/event-manager.ts
import * as cbor3 from "cbor-x";
var EventManager = class {
  #actor;
  #subscriptionIndex = /* @__PURE__ */ new Map();
  constructor(actor) {
    this.#actor = actor;
  }
  // MARK: - Public API
  /**
   * Adds a subscription for a connection to an event.
   *
   * @param eventName - The name of the event to subscribe to
   * @param connection - The connection subscribing to the event
   * @param fromPersist - Whether this subscription is being restored from persistence
   */
  addSubscription(eventName, connection, fromPersist) {
    if (connection.subscriptions.has(eventName)) {
      this.#actor.rLog.debug({
        msg: "connection already has subscription",
        eventName,
        connId: connection.id
      });
      return;
    }
    connection.subscriptions.add(eventName);
    let subscribers = this.#subscriptionIndex.get(eventName);
    if (!subscribers) {
      subscribers = /* @__PURE__ */ new Set();
      this.#subscriptionIndex.set(eventName, subscribers);
    }
    subscribers.add(connection);
    if (!fromPersist) {
      connection[CONN_STATE_MANAGER_SYMBOL].addSubscription({
        eventName
      });
      this.#actor.stateManager.saveState({ immediate: true });
    }
    this.#actor.rLog.debug({
      msg: "subscription added",
      eventName,
      connId: connection.id,
      totalSubscribers: subscribers.size
    });
  }
  /**
   * Removes a subscription for a connection from an event.
   *
   * @param eventName - The name of the event to unsubscribe from
   * @param connection - The connection unsubscribing from the event
   * @param fromRemoveConn - Whether this is being called as part of connection removal
   */
  removeSubscription(eventName, connection, fromRemoveConn) {
    if (!connection.subscriptions.has(eventName)) {
      this.#actor.rLog.warn({
        msg: "connection does not have subscription",
        eventName,
        connId: connection.id
      });
      return;
    }
    connection.subscriptions.delete(eventName);
    const subscribers = this.#subscriptionIndex.get(eventName);
    if (subscribers) {
      subscribers.delete(connection);
      if (subscribers.size === 0) {
        this.#subscriptionIndex.delete(eventName);
      }
    }
    if (!fromRemoveConn) {
      const removed = connection[CONN_STATE_MANAGER_SYMBOL].removeSubscription({ eventName });
      if (!removed) {
        this.#actor.rLog.warn({
          msg: "subscription does not exist in persist",
          eventName,
          connId: connection.id
        });
      }
      this.#actor.stateManager.saveState({ immediate: true });
    }
    this.#actor.rLog.debug({
      msg: "subscription removed",
      eventName,
      connId: connection.id,
      remainingSubscribers: (subscribers == null ? void 0 : subscribers.size) || 0
    });
  }
  /**
   * Broadcasts an event to all subscribed connections.
   *
   * @param name - The name of the event to broadcast
   * @param args - The arguments to send with the event
   */
  broadcast(name, ...args) {
    this.#actor.assertReady();
    const subscribers = this.#subscriptionIndex.get(name);
    if (!subscribers || subscribers.size === 0) {
      this.#actor.rLog.debug({
        msg: "no subscribers for event",
        eventName: name
      });
      return;
    }
    this.#actor.emitTraceEvent("message.broadcast", {
      "rivet.event.name": name,
      "rivet.broadcast.subscribers": subscribers.size
    });
    const eventData = { name, args };
    const toClientSerializer = new CachedSerializer(
      eventData,
      TO_CLIENT_VERSIONED,
      CURRENT_VERSION,
      ToClientSchema,
      // JSON: args is the raw value (array of arguments)
      (value) => ({
        body: {
          tag: "Event",
          val: {
            name: value.name,
            args: value.args
          }
        }
      }),
      // BARE/CBOR: args needs to be CBOR-encoded to ArrayBuffer
      (value) => ({
        body: {
          tag: "Event",
          val: {
            name: value.name,
            args: bufferToArrayBuffer(cbor3.encode(value.args))
          }
        }
      })
    );
    let sentCount = 0;
    for (const connection of subscribers) {
      if (connection[CONN_SPEAKS_RIVETKIT_SYMBOL]) {
        try {
          connection[CONN_SEND_MESSAGE_SYMBOL](toClientSerializer);
          sentCount++;
        } catch (error) {
          if (error instanceof OutgoingMessageTooLong) {
            throw error;
          }
          this.#actor.rLog.error({
            msg: "failed to send event to connection",
            eventName: name,
            connId: connection.id,
            error: error instanceof Error ? error.message : String(error)
          });
        }
      }
    }
    this.#actor.rLog.debug({
      msg: "event broadcasted",
      eventName: name,
      subscriberCount: subscribers.size,
      sentCount
    });
  }
  /**
   * Gets all subscribers for a specific event.
   *
   * @param eventName - The name of the event
   * @returns Set of connections subscribed to the event, or undefined if no subscribers
   */
  getSubscribers(eventName) {
    return this.#subscriptionIndex.get(eventName);
  }
  /**
   * Gets all events and their subscriber counts.
   *
   * @returns Map of event names to subscriber counts
   */
  getEventStats() {
    const stats = /* @__PURE__ */ new Map();
    for (const [eventName, subscribers] of this.#subscriptionIndex) {
      stats.set(eventName, subscribers.size);
    }
    return stats;
  }
  /**
   * Clears all subscriptions for a connection.
   * Used during connection cleanup.
   *
   * @param connection - The connection to clear subscriptions for
   */
  clearConnectionSubscriptions(connection) {
    for (const eventName of [...connection.subscriptions.values()]) {
      this.removeSubscription(eventName, connection, true);
    }
  }
  /**
   * Gets the total number of unique events being subscribed to.
   */
  get eventCount() {
    return this.#subscriptionIndex.size;
  }
  /**
   * Gets the total number of subscriptions across all events.
   */
  get totalSubscriptionCount() {
    let total = 0;
    for (const subscribers of this.#subscriptionIndex.values()) {
      total += subscribers.size;
    }
    return total;
  }
  /**
   * Checks if an event has any subscribers.
   *
   * @param eventName - The name of the event to check
   * @returns True if the event has at least one subscriber
   */
  hasSubscribers(eventName) {
    const subscribers = this.#subscriptionIndex.get(eventName);
    return subscribers !== void 0 && subscribers.size > 0;
  }
};

// src/actor/instance/persisted.ts
import * as cbor4 from "cbor-x";
function convertActorToBarePersisted(persist) {
  return {
    input: persist.input !== void 0 ? bufferToArrayBuffer(cbor4.encode(persist.input)) : null,
    hasInitialized: persist.hasInitialized,
    state: bufferToArrayBuffer(cbor4.encode(persist.state)),
    scheduledEvents: persist.scheduledEvents.map((event) => ({
      eventId: event.eventId,
      timestamp: BigInt(event.timestamp),
      action: event.action,
      args: event.args ?? null
    }))
  };
}
function convertActorFromBarePersisted(bareData) {
  return {
    input: bareData.input ? cbor4.decode(new Uint8Array(bareData.input)) : void 0,
    hasInitialized: bareData.hasInitialized,
    state: cbor4.decode(new Uint8Array(bareData.state)),
    scheduledEvents: bareData.scheduledEvents.map((event) => ({
      eventId: event.eventId,
      timestamp: Number(event.timestamp),
      action: event.action,
      args: event.args ?? void 0
    }))
  };
}

// src/actor/instance/queue-manager.ts
import * as cbor5 from "cbor-x";
var DEFAULT_METADATA = {
  nextId: 1n,
  size: 0
};
var QUEUE_METADATA_KEY = queueMetadataKey();
var QUEUE_MESSAGES_PREFIX = queueMessagesPrefix();
var QueueManager = class {
  #actor;
  #driver;
  #waiters = /* @__PURE__ */ new Map();
  #metadata = { ...DEFAULT_METADATA };
  #messageListeners = /* @__PURE__ */ new Set();
  #pendingCompletions = /* @__PURE__ */ new Map();
  constructor(actor, driver) {
    this.#actor = actor;
    this.#driver = driver;
  }
  /** Returns the current number of messages in the queue. */
  get size() {
    return this.#metadata.size;
  }
  /** Loads queue metadata from storage and initializes internal state. */
  async initialize() {
    const [metadataBuffer] = await this.#driver.kvBatchGet(this.#actor.id, [
      QUEUE_METADATA_KEY
    ]);
    if (!metadataBuffer) {
      await this.#driver.kvBatchPut(this.#actor.id, [
        [QUEUE_METADATA_KEY, this.#serializeMetadata()]
      ]);
      this.#actor.inspector.updateQueueSize(this.#metadata.size);
      return;
    }
    try {
      const decoded = QUEUE_METADATA_VERSIONED.deserializeWithEmbeddedVersion(
        metadataBuffer
      );
      this.#metadata.nextId = decoded.nextId;
      this.#metadata.size = Number(decoded.size);
    } catch (error) {
      this.#actor.rLog.error({
        msg: "failed to decode queue metadata, rebuilding from messages",
        error
      });
      await this.#rebuildMetadata();
    }
    this.#actor.inspector.updateQueueSize(this.#metadata.size);
  }
  /** Adds a message to the queue with the given name and body. */
  async enqueue(name, body) {
    this.#actor.assertReady();
    const sizeLimit = this.#actor.config.options.maxQueueSize;
    if (this.#metadata.size >= sizeLimit) {
      throw new QueueFull(sizeLimit);
    }
    let invalidPath = "";
    if (!isCborSerializable(body, (path) => {
      invalidPath = path;
    })) {
      throw new QueueMessageInvalid(invalidPath);
    }
    const createdAt = Date.now();
    const bodyCborBuffer = cbor5.encode(body);
    const encodedMessage = QUEUE_MESSAGE_VERSIONED.serializeWithEmbeddedVersion(
      {
        name,
        body: new Uint8Array(bodyCborBuffer).buffer,
        createdAt: BigInt(createdAt),
        failureCount: null,
        availableAt: null,
        inFlight: null,
        inFlightAt: null
      },
      CURRENT_VERSION2
    );
    const encodedSize = encodedMessage.byteLength;
    if (encodedSize > this.#actor.config.options.maxQueueMessageSize) {
      throw new QueueMessageTooLarge(
        encodedSize,
        this.#actor.config.options.maxQueueMessageSize
      );
    }
    const id = this.#metadata.nextId;
    const messageKey = makeQueueMessageKey(id);
    this.#metadata.nextId = id + 1n;
    this.#metadata.size += 1;
    const encodedMetadata = this.#serializeMetadata();
    await this.#driver.kvBatchPut(this.#actor.id, [
      [messageKey, encodedMessage],
      [QUEUE_METADATA_KEY, encodedMetadata]
    ]);
    this.#actor.inspector.updateQueueSize(this.#metadata.size);
    const message = {
      id,
      name,
      body,
      createdAt
    };
    this.#actor.resetSleepTimer();
    await this.#maybeResolveWaiters();
    this.#notifyMessageListeners(name);
    return message;
  }
  /**
   * Adds a message and waits for completion.
   */
  async enqueueAndWait(name, body, timeout) {
    if (timeout !== void 0 && timeout <= 0) {
      return { status: "timedOut" };
    }
    const message = await this.enqueue(name, body);
    const messageId = message.id.toString();
    const { promise, resolve } = promiseWithResolvers(() => {
    });
    const pending = { resolve };
    if (timeout !== void 0) {
      pending.timeoutHandle = setTimeout(() => {
        this.#pendingCompletions.delete(messageId);
        resolve({ status: "timedOut" });
      }, timeout);
    }
    this.#pendingCompletions.set(messageId, pending);
    return await promise;
  }
  async completeMessage(message, response) {
    await this.completeMessageById(message.id, response);
  }
  async completeMessageById(messageId, response) {
    const messageIdString = messageId.toString();
    const pending = this.#pendingCompletions.get(messageIdString);
    if (pending) {
      if (pending.timeoutHandle) {
        clearTimeout(pending.timeoutHandle);
      }
      this.#pendingCompletions.delete(messageIdString);
      pending.resolve({ status: "completed", response });
    }
    await this.deleteMessagesById([messageId]);
  }
  /** Receives messages from the queue matching the given names. Waits until messages are available or timeout is reached. */
  async receive(names, count, timeout, abortSignal, completable = false) {
    this.#actor.assertReady();
    const limitedCount = Math.max(1, count);
    const nameSet = names && names.length > 0 ? new Set(names) : void 0;
    const immediate = await this.#drainMessages(
      nameSet,
      limitedCount,
      completable
    );
    if (immediate.length > 0) {
      return immediate;
    }
    if (timeout === 0) {
      return [];
    }
    const { promise, resolve, reject } = promiseWithResolvers(() => {
    });
    const waiterId = crypto.randomUUID();
    let timeoutHandle;
    let cleanedUp = false;
    let actorAbortCleanup;
    let signalAbortCleanup;
    const cleanup = () => {
      if (cleanedUp) {
        return;
      }
      cleanedUp = true;
      this.#waiters.delete(waiterId);
      if (timeoutHandle) {
        clearTimeout(timeoutHandle);
        timeoutHandle = void 0;
      }
      actorAbortCleanup == null ? void 0 : actorAbortCleanup();
      signalAbortCleanup == null ? void 0 : signalAbortCleanup();
      this.#actor.endQueueWait();
    };
    const resolveWaiter = (messages) => {
      cleanup();
      resolve(messages);
    };
    const rejectWaiter = (error) => {
      cleanup();
      reject(error);
    };
    const waiter = {
      id: waiterId,
      nameSet,
      count: limitedCount,
      completable,
      resolve: resolveWaiter,
      reject: rejectWaiter
    };
    this.#actor.beginQueueWait();
    if (timeout !== void 0) {
      timeoutHandle = setTimeout(() => {
        resolveWaiter([]);
      }, timeout);
    }
    const onAbort = () => {
      rejectWaiter(new ActorAborted());
    };
    const onStop = () => {
      rejectWaiter(new ActorAborted());
    };
    const actorAbortSignal = this.#actor.abortSignal;
    if (actorAbortSignal.aborted) {
      onStop();
      return promise;
    }
    actorAbortSignal.addEventListener("abort", onStop, { once: true });
    actorAbortCleanup = () => actorAbortSignal.removeEventListener("abort", onStop);
    if (abortSignal) {
      if (abortSignal.aborted) {
        onAbort();
        return promise;
      }
      abortSignal.addEventListener("abort", onAbort, { once: true });
      signalAbortCleanup = () => abortSignal.removeEventListener("abort", onAbort);
    }
    this.#waiters.set(waiterId, waiter);
    return promise;
  }
  async waitForNames(names, abortSignal) {
    const nameSet = names && names.length > 0 ? new Set(names) : void 0;
    const existing = await this.#loadQueueMessages();
    if (nameSet) {
      if (existing.some((message) => nameSet.has(message.name))) {
        return;
      }
    } else if (existing.length > 0) {
      return;
    }
    return await new Promise((resolve, reject) => {
      this.#actor.beginQueueWait();
      const listener = {
        nameSet,
        resolve: () => {
          this.#removeMessageListener(listener);
          this.#actor.endQueueWait();
          resolve();
        },
        reject: (error) => {
          this.#removeMessageListener(listener);
          this.#actor.endQueueWait();
          reject(error);
        }
      };
      const actorAbortSignal = this.#actor.abortSignal;
      const onActorAbort = () => listener.reject(new ActorAborted());
      if (actorAbortSignal.aborted) {
        onActorAbort();
        return;
      }
      actorAbortSignal.addEventListener("abort", onActorAbort, {
        once: true
      });
      listener.actorAbortCleanup = () => actorAbortSignal.removeEventListener("abort", onActorAbort);
      if (abortSignal) {
        const onAbort = () => listener.reject(new ActorAborted());
        if (abortSignal.aborted) {
          onAbort();
          return;
        }
        abortSignal.addEventListener("abort", onAbort, { once: true });
        listener.signalAbortCleanup = () => abortSignal.removeEventListener("abort", onAbort);
      }
      this.#messageListeners.add(listener);
    });
  }
  /** Returns all messages currently in the queue without removing them. */
  async getMessages() {
    return await this.#loadQueueMessages();
  }
  /** Deletes messages matching the provided IDs. Returns the IDs that were removed. */
  async deleteMessagesById(ids) {
    if (ids.length === 0) {
      return [];
    }
    const idSet = new Set(ids.map((id) => id.toString()));
    const entries = await this.#loadQueueMessages();
    const toRemove = entries.filter(
      (entry) => idSet.has(entry.id.toString())
    );
    if (toRemove.length === 0) {
      return [];
    }
    await this.#removeMessages(toRemove);
    return toRemove.map((entry) => entry.id);
  }
  async #drainMessages(nameSet, count, completable) {
    if (this.#metadata.size === 0) {
      return [];
    }
    const entries = await this.#loadQueueMessages();
    const matched = nameSet ? entries.filter((entry) => nameSet.has(entry.name)) : entries;
    if (matched.length === 0) {
      return [];
    }
    const selected = matched.slice(0, count);
    if (!completable) {
      await this.#removeMessages(selected);
    }
    const now = Date.now();
    for (const message of selected) {
      this.#actor.emitTraceEvent("queue.message.receive", {
        "rivet.queue.name": message.name,
        "rivet.queue.message_id": message.id.toString(),
        "rivet.queue.created_at_ms": message.createdAt,
        "rivet.queue.latency_ms": now - message.createdAt
      });
    }
    return selected;
  }
  async #loadQueueMessages() {
    const entries = await this.#driver.kvListPrefix(
      this.#actor.id,
      QUEUE_MESSAGES_PREFIX
    );
    const decoded = [];
    for (const [key, value] of entries) {
      try {
        const messageId = decodeQueueMessageKey(key);
        const decodedPayload = QUEUE_MESSAGE_VERSIONED.deserializeWithEmbeddedVersion(
          value
        );
        const body = cbor5.decode(new Uint8Array(decodedPayload.body));
        decoded.push({
          id: messageId,
          name: decodedPayload.name,
          body,
          createdAt: Number(decodedPayload.createdAt)
        });
      } catch (error) {
        this.#actor.rLog.error({
          msg: "failed to decode queue message",
          error
        });
      }
    }
    decoded.sort((a, b) => a.id < b.id ? -1 : a.id > b.id ? 1 : 0);
    if (this.#metadata.size !== decoded.length) {
      this.#metadata.size = decoded.length;
      this.#actor.inspector.updateQueueSize(this.#metadata.size);
    }
    return decoded;
  }
  #removeMessageListener(listener) {
    var _a, _b;
    if (this.#messageListeners.delete(listener)) {
      (_a = listener.actorAbortCleanup) == null ? void 0 : _a.call(listener);
      (_b = listener.signalAbortCleanup) == null ? void 0 : _b.call(listener);
    }
  }
  #notifyMessageListeners(name) {
    if (this.#messageListeners.size === 0) {
      return;
    }
    for (const listener of [...this.#messageListeners]) {
      if (listener.nameSet && !listener.nameSet.has(name)) {
        continue;
      }
      this.#removeMessageListener(listener);
      listener.resolve();
    }
  }
  async #removeMessages(messages) {
    if (messages.length === 0) {
      return;
    }
    const keys = messages.map((message) => makeQueueMessageKey(message.id));
    this.#metadata.size = Math.max(
      0,
      this.#metadata.size - messages.length
    );
    await this.#driver.kvBatchDelete(this.#actor.id, keys);
    await this.#driver.kvBatchPut(this.#actor.id, [
      [QUEUE_METADATA_KEY, this.#serializeMetadata()]
    ]);
    this.#actor.inspector.updateQueueSize(this.#metadata.size);
  }
  async #maybeResolveWaiters() {
    if (this.#waiters.size === 0) {
      return;
    }
    const pending = [...this.#waiters.values()];
    for (const waiter of pending) {
      const messages = await this.#drainMessages(
        waiter.nameSet,
        waiter.count,
        waiter.completable
      );
      if (messages.length === 0) {
        continue;
      }
      this.#waiters.delete(waiter.id);
      waiter.resolve(messages);
    }
  }
  /** Rebuilds metadata by scanning existing queue messages. Used when metadata is corrupted. */
  async #rebuildMetadata() {
    const entries = await this.#driver.kvListPrefix(
      this.#actor.id,
      QUEUE_MESSAGES_PREFIX
    );
    let maxId = 0n;
    for (const [key] of entries) {
      try {
        const messageId = decodeQueueMessageKey(key);
        if (messageId > maxId) {
          maxId = messageId;
        }
      } catch {
      }
    }
    this.#metadata.nextId = maxId + 1n;
    this.#metadata.size = entries.length;
    await this.#driver.kvBatchPut(this.#actor.id, [
      [QUEUE_METADATA_KEY, this.#serializeMetadata()]
    ]);
    this.#actor.inspector.updateQueueSize(this.#metadata.size);
  }
  #serializeMetadata() {
    return QUEUE_METADATA_VERSIONED.serializeWithEmbeddedVersion(
      {
        nextId: this.#metadata.nextId,
        size: this.#metadata.size
      },
      CURRENT_VERSION2
    );
  }
};

// src/actor/instance/schedule-manager.ts
import * as cbor6 from "cbor-x";
var ScheduleManager = class {
  #actor;
  #actorDriver;
  #alarmWriteQueue = new SinglePromiseQueue();
  #config;
  // ActorConfig type
  #persist;
  // Reference to PersistedActor
  constructor(actor, actorDriver, config4) {
    this.#actor = actor;
    this.#actorDriver = actorDriver;
    this.#config = config4;
  }
  // MARK: - Public API
  /**
   * Sets the persist object reference.
   * Called after StateManager initializes the persist proxy.
   */
  setPersist(persist) {
    this.#persist = persist;
  }
  /**
   * Schedules an event to be executed at a specific timestamp.
   *
   * @param timestamp - Unix timestamp in milliseconds when the event should fire
   * @param action - The name of the action to execute
   * @param args - Arguments to pass to the action
   */
  async scheduleEvent(timestamp, action, args) {
    const newEvent = {
      eventId: crypto.randomUUID(),
      timestamp,
      action,
      args: bufferToArrayBuffer(cbor6.encode(args))
    };
    this.#actor.emitTraceEvent("schedule.created", {
      "rivet.schedule.event_id": newEvent.eventId,
      "rivet.schedule.action": newEvent.action,
      "rivet.schedule.timestamp_ms": newEvent.timestamp
    });
    await this.#scheduleEventInner(newEvent);
  }
  /**
   * Triggers any pending alarms that are due.
   * This method is idempotent and safe to call multiple times.
   */
  async onAlarm() {
    var _a, _b, _c;
    const now = Date.now();
    this.#actor.log.debug({
      msg: "alarm triggered",
      now,
      events: ((_b = (_a = this.#persist) == null ? void 0 : _a.scheduledEvents) == null ? void 0 : _b.length) || 0
    });
    if (!((_c = this.#persist) == null ? void 0 : _c.scheduledEvents)) {
      this.#actor.rLog.debug({ msg: "no scheduled events" });
      return;
    }
    const dueIndex = this.#persist.scheduledEvents.findIndex(
      (x) => x.timestamp <= now
    );
    if (dueIndex === -1) {
      this.#actor.rLog.debug({ msg: "no events are due yet" });
      if (this.#persist.scheduledEvents.length > 0) {
        const nextTs = this.#persist.scheduledEvents[0].timestamp;
        this.#actor.log.debug({
          msg: "alarm fired early, rescheduling for next event",
          now,
          nextTs,
          delta: nextTs - now
        });
        await this.#queueSetAlarm(nextTs);
      }
      return;
    }
    const dueEvents = this.#persist.scheduledEvents.splice(0, dueIndex + 1);
    this.#actor.log.debug({
      msg: "running events",
      count: dueEvents.length
    });
    if (this.#persist.scheduledEvents.length > 0) {
      const nextTs = this.#persist.scheduledEvents[0].timestamp;
      this.#actor.log.info({
        msg: "setting next alarm",
        nextTs,
        remainingEvents: this.#persist.scheduledEvents.length
      });
      await this.#queueSetAlarm(nextTs);
    }
    await this.#executeDueEvents(dueEvents);
  }
  /**
   * Initializes alarms on actor startup.
   * Sets the alarm for the next scheduled event if any exist.
   */
  async initializeAlarms() {
    var _a, _b;
    if (((_b = (_a = this.#persist) == null ? void 0 : _a.scheduledEvents) == null ? void 0 : _b.length) > 0) {
      await this.#queueSetAlarm(
        this.#persist.scheduledEvents[0].timestamp
      );
    }
  }
  /**
   * Waits for any pending alarm write operations to complete.
   */
  async waitForPendingAlarmWrites() {
    if (this.#alarmWriteQueue.runningDrainLoop) {
      await this.#alarmWriteQueue.runningDrainLoop;
    }
  }
  /**
   * Gets statistics about scheduled events.
   */
  getScheduleStats() {
    var _a;
    if (!((_a = this.#persist) == null ? void 0 : _a.scheduledEvents)) {
      return {
        totalEvents: 0,
        nextEventTime: null,
        overdueCount: 0
      };
    }
    const now = Date.now();
    const events = this.#persist.scheduledEvents;
    return {
      totalEvents: events.length,
      nextEventTime: events.length > 0 ? events[0].timestamp : null,
      overdueCount: events.filter(
        (e) => e.timestamp <= now
      ).length
    };
  }
  /**
   * Cancels a scheduled event by its ID.
   *
   * @param eventId - The ID of the event to cancel
   * @returns True if the event was found and cancelled
   */
  async cancelEvent(eventId) {
    var _a;
    if (!((_a = this.#persist) == null ? void 0 : _a.scheduledEvents)) {
      return false;
    }
    const index = this.#persist.scheduledEvents.findIndex(
      (e) => e.eventId === eventId
    );
    if (index === -1) {
      return false;
    }
    const wasFirst = index === 0;
    this.#persist.scheduledEvents.splice(index, 1);
    if (wasFirst && this.#persist.scheduledEvents.length > 0) {
      await this.#queueSetAlarm(
        this.#persist.scheduledEvents[0].timestamp
      );
    }
    this.#actor.log.info({
      msg: "cancelled scheduled event",
      eventId,
      remainingEvents: this.#persist.scheduledEvents.length
    });
    return true;
  }
  // MARK: - Private Helpers
  async #scheduleEventInner(newEvent) {
    var _a;
    this.#actor.log.info({
      msg: "scheduling event",
      eventId: newEvent.eventId,
      timestamp: newEvent.timestamp,
      action: newEvent.action
    });
    if (!((_a = this.#persist) == null ? void 0 : _a.scheduledEvents)) {
      throw new Error("Persist not initialized");
    }
    const insertIndex = this.#persist.scheduledEvents.findIndex(
      (x) => x.timestamp > newEvent.timestamp
    );
    if (insertIndex === -1) {
      this.#persist.scheduledEvents.push(newEvent);
    } else {
      this.#persist.scheduledEvents.splice(insertIndex, 0, newEvent);
    }
    if (insertIndex === 0 || this.#persist.scheduledEvents.length === 1) {
      this.#actor.log.info({
        msg: "setting alarm for new event",
        timestamp: newEvent.timestamp,
        eventCount: this.#persist.scheduledEvents.length
      });
      await this.#queueSetAlarm(newEvent.timestamp);
    }
  }
  async #executeDueEvents(events) {
    for (const event of events) {
      const span = this.#actor.startTraceSpan(
        `actor.action.${event.action}`,
        {
          "rivet.action.name": event.action,
          "rivet.action.scheduled": true,
          "rivet.schedule.event_id": event.eventId,
          "rivet.schedule.timestamp_ms": event.timestamp
        }
      );
      try {
        this.#actor.emitTraceEvent(
          "schedule.triggered",
          {
            "rivet.schedule.event_id": event.eventId,
            "rivet.schedule.action": event.action,
            "rivet.schedule.timestamp_ms": event.timestamp
          },
          span
        );
        this.#actor.log.info({
          msg: "executing scheduled event",
          eventId: event.eventId,
          timestamp: event.timestamp,
          action: event.action
        });
        const actions = this.#config.actions ?? {};
        const fn = actions[event.action];
        if (!fn) {
          throw new Error(
            `Missing action for scheduled event: ${event.action}`
          );
        }
        if (typeof fn !== "function") {
          throw new Error(
            `Scheduled event action ${event.action} is not a function (got ${typeof fn})`
          );
        }
        const args = event.args ? cbor6.decode(new Uint8Array(event.args)) : [];
        const result = this.#actor.traces.withSpan(
          span,
          () => fn.call(void 0, this.#actor.actorContext, ...args)
        );
        if (result instanceof Promise) {
          await result;
        }
        this.#actor.endTraceSpan(span, { code: "OK" });
        this.#actor.log.debug({
          msg: "scheduled event completed",
          eventId: event.eventId,
          action: event.action
        });
      } catch (error) {
        this.#actor.traces.setAttributes(span, {
          "error.message": stringifyError(error),
          "error.type": error instanceof Error ? error.name : typeof error
        });
        this.#actor.endTraceSpan(span, {
          code: "ERROR",
          message: stringifyError(error)
        });
        this.#actor.log.error({
          msg: "error executing scheduled event",
          error: stringifyError(error),
          eventId: event.eventId,
          timestamp: event.timestamp,
          action: event.action
        });
      }
    }
  }
  async #queueSetAlarm(timestamp) {
    await this.#alarmWriteQueue.enqueue(async () => {
      await this.#actorDriver.setAlarm(this.#actor, timestamp);
    });
  }
  /**
   * Gets the next scheduled event, if any.
   */
  getNextEvent() {
    var _a;
    if (!((_a = this.#persist) == null ? void 0 : _a.scheduledEvents) || this.#persist.scheduledEvents.length === 0) {
      return null;
    }
    return this.#persist.scheduledEvents[0];
  }
  /**
   * Gets all scheduled events.
   */
  getAllEvents() {
    var _a;
    return ((_a = this.#persist) == null ? void 0 : _a.scheduledEvents) || [];
  }
  /**
   * Clears all scheduled events.
   * Use with caution - this removes all pending scheduled events.
   */
  clearAllEvents() {
    var _a;
    if ((_a = this.#persist) == null ? void 0 : _a.scheduledEvents) {
      this.#persist.scheduledEvents = [];
      this.#actor.log.warn({ msg: "cleared all scheduled events" });
    }
  }
};

// src/actor/instance/state-manager.ts
import { idToStr } from "@rivetkit/engine-runner";
import onChange from "@rivetkit/on-change";
var StateManager = class {
  #actor;
  #actorDriver;
  // State tracking
  #persist;
  #persistRaw;
  #persistChanged = false;
  #isInOnStateChange = false;
  // Save management
  #persistWriteQueue = new SinglePromiseQueue();
  #lastSaveTime = 0;
  #pendingSaveTimeout;
  #pendingSaveScheduledTimestamp;
  #onPersistSavedPromise;
  // Configuration
  #config;
  // ActorConfig type
  #stateSaveInterval;
  constructor(actor, actorDriver, config4) {
    this.#actor = actor;
    this.#actorDriver = actorDriver;
    this.#config = config4;
    this.#stateSaveInterval = config4.options.stateSaveInterval || 100;
  }
  // MARK: - Public API
  get persist() {
    return this.#persist;
  }
  get persistRaw() {
    return this.#persistRaw;
  }
  get persistChanged() {
    return this.#persistChanged;
  }
  get state() {
    this.#validateStateEnabled();
    return this.#persist.state;
  }
  set state(value) {
    this.#validateStateEnabled();
    this.#persist.state = value;
  }
  get stateEnabled() {
    return "createState" in this.#config || "state" in this.#config;
  }
  // MARK: - Initialization
  /**
   * Initializes state from persisted data or creates new state.
   */
  async initializeState(persistData) {
    if (!persistData.hasInitialized) {
      let stateData;
      if (this.stateEnabled) {
        this.#actor.rLog.info({ msg: "actor state initializing" });
        if ("createState" in this.#config) {
          stateData = await this.#actor.runInTraceSpan(
            "actor.createState",
            void 0,
            () => this.#config.createState(
              this.#actor.actorContext,
              persistData.input
            )
          );
        } else if ("state" in this.#config) {
          stateData = structuredClone(this.#config.state);
        } else {
          throw new Error(
            "Both 'createState' or 'state' were not defined"
          );
        }
      } else {
        this.#actor.rLog.debug({ msg: "state not enabled" });
      }
      persistData.state = stateData;
      persistData.hasInitialized = true;
      const bareData = convertActorToBarePersisted(persistData);
      await this.#actorDriver.kvBatchPut(this.#actor.id, [
        [
          KEYS.PERSIST_DATA,
          ACTOR_VERSIONED.serializeWithEmbeddedVersion(
            bareData,
            CURRENT_VERSION2
          )
        ]
      ]);
    }
    this.initPersistProxy(persistData);
  }
  /**
   * Creates proxy for persist object that handles automatic state change detection.
   */
  initPersistProxy(target) {
    this.#persistRaw = target;
    if (target === null || typeof target !== "object") {
      let invalidPath = "";
      if (!isCborSerializable(
        target,
        (path) => {
          invalidPath = path;
        },
        ""
      )) {
        throw new InvalidStateType({ path: invalidPath });
      }
      return target;
    }
    if (this.#persist) {
      onChange.unsubscribe(this.#persist);
    }
    this.#persist = onChange(
      target,
      (path, value, _previousValue, _applyData) => {
        this.#handleStateChange(path, value);
      },
      { ignoreDetached: true }
    );
  }
  // MARK: - State Persistence
  /**
   * Forces the state to get saved.
   */
  async saveState(opts) {
    var _a;
    this.#actor.assertReady(opts.allowStoppingState);
    if (this.#persistChanged) {
      if (opts.immediate) {
        await this.#savePersistInner();
      } else {
        if (!this.#onPersistSavedPromise) {
          this.#onPersistSavedPromise = promiseWithResolvers((reason) => loggerWithoutContext().warn({ msg: "unhandled persist saved promise rejection", reason }));
        }
        this.savePersistThrottled(opts.maxWait);
        await ((_a = this.#onPersistSavedPromise) == null ? void 0 : _a.promise);
      }
    }
  }
  /**
   * Throttled save state method. Used to write to KV at a reasonable cadence.
   *
   * Passing a maxWait will override the stateSaveInterval with the min
   * between that and the maxWait.
   */
  savePersistThrottled(maxWait) {
    const now = Date.now();
    const timeSinceLastSave = now - this.#lastSaveTime;
    let saveDelay = Math.max(
      0,
      this.#stateSaveInterval - timeSinceLastSave
    );
    if (maxWait !== void 0) {
      saveDelay = Math.min(saveDelay, maxWait);
    }
    if (this.#pendingSaveTimeout !== void 0 && this.#pendingSaveScheduledTimestamp !== void 0) {
      const newScheduledTimestamp = now + saveDelay;
      if (newScheduledTimestamp < this.#pendingSaveScheduledTimestamp) {
        clearTimeout(this.#pendingSaveTimeout);
        this.#pendingSaveTimeout = void 0;
        this.#pendingSaveScheduledTimestamp = void 0;
      } else {
        return;
      }
    }
    if (saveDelay > 0) {
      this.#pendingSaveScheduledTimestamp = now + saveDelay;
      this.#pendingSaveTimeout = setTimeout(() => {
        this.#pendingSaveTimeout = void 0;
        this.#pendingSaveScheduledTimestamp = void 0;
        this.#savePersistInner().catch((error) => {
          this.#actor.rLog.error({
            msg: "error saving persist data in scheduled save",
            error: stringifyError(error)
          });
        });
      }, saveDelay);
    } else {
      this.#savePersistInner().catch((error) => {
        this.#actor.rLog.error({
          msg: "error saving persist data immediately",
          error: stringifyError(error)
        });
      });
    }
  }
  /**
   * Clears any pending save timeout.
   */
  clearPendingSaveTimeout() {
    if (this.#pendingSaveTimeout) {
      clearTimeout(this.#pendingSaveTimeout);
      this.#pendingSaveTimeout = void 0;
      this.#pendingSaveScheduledTimestamp = void 0;
    }
  }
  /**
   * Waits for any pending write operations to complete.
   */
  async waitForPendingWrites() {
    if (this.#persistWriteQueue.runningDrainLoop) {
      await this.#persistWriteQueue.runningDrainLoop;
    }
  }
  // MARK: - Private Helpers
  #validateStateEnabled() {
    if (!this.stateEnabled) {
      throw new StateNotEnabled();
    }
  }
  #handleStateChange(path, value) {
    const actorStatePath = isStatePath(path);
    const connStatePath = isConnStatePath(path);
    if (actorStatePath || connStatePath) {
      let invalidPath = "";
      if (!isCborSerializable(
        value,
        (invalidPathPart) => {
          invalidPath = invalidPathPart;
        },
        ""
      )) {
        throw new InvalidStateType({
          path: path + (invalidPath ? `.${invalidPath}` : "")
        });
      }
    }
    this.#actor.rLog.debug({
      msg: "onChange triggered, setting persistChanged=true",
      path
    });
    this.#persistChanged = true;
    if (actorStatePath) {
      this.#actor.inspector.emitter.emit(
        "stateUpdated",
        this.#persist.state
      );
    }
    if (actorStatePath && this.#config.onStateChange && this.#actor.isReady() && !this.#isInOnStateChange) {
      const span = this.#actor.startTraceSpan("actor.onStateChange", {
        "rivet.state.path": path
      });
      try {
        this.#isInOnStateChange = true;
        this.#actor.traces.withSpan(
          span,
          () => this.#config.onStateChange(
            this.#actor.actorContext,
            this.#persistRaw.state
          )
        );
        this.#actor.endTraceSpan(span, { code: "OK" });
      } catch (error) {
        this.#actor.endTraceSpan(span, {
          code: "ERROR",
          message: stringifyError(error)
        });
        this.#actor.rLog.error({
          msg: "error in `_onStateChange`",
          error: stringifyError(error)
        });
      } finally {
        this.#isInOnStateChange = false;
      }
    }
  }
  async #savePersistInner() {
    var _a, _b;
    this.#actor.rLog.info({
      msg: "savePersistInner called",
      persistChanged: this.#persistChanged,
      connsWithPersistChangedSize: this.#actor.connectionManager.connsWithPersistChanged.size,
      connsWithPersistChangedIds: Array.from(
        this.#actor.connectionManager.connsWithPersistChanged
      )
    });
    try {
      this.#lastSaveTime = Date.now();
      const hasChanges = this.#persistChanged || this.#actor.connectionManager.connsWithPersistChanged.size > 0;
      if (hasChanges) {
        await this.#persistWriteQueue.enqueue(async () => {
          this.#actor.rLog.debug({
            msg: "saving persist",
            actorChanged: this.#persistChanged,
            connectionsChanged: this.#actor.connectionManager.connsWithPersistChanged.size
          });
          const entries = [];
          if (this.#persistChanged) {
            this.#persistChanged = false;
            const bareData = convertActorToBarePersisted(
              this.#persistRaw
            );
            entries.push([
              KEYS.PERSIST_DATA,
              ACTOR_VERSIONED.serializeWithEmbeddedVersion(
                bareData,
                CURRENT_VERSION2
              )
            ]);
          }
          const connections = [];
          for (const connId of this.#actor.connectionManager.connsWithPersistChanged) {
            const conn = this.#actor.conns.get(connId);
            if (!conn) {
              this.#actor.rLog.warn({
                msg: "connection not found in conns map",
                connId
              });
              continue;
            }
            const connStateManager = conn[CONN_STATE_MANAGER_SYMBOL];
            const hibernatableDataRaw = connStateManager.hibernatableDataRaw;
            if (!hibernatableDataRaw) {
              this.#actor.log.warn({
                msg: "missing raw hibernatable data for conn in getChangedConnectionsData",
                connId: conn.id
              });
              continue;
            }
            this.#actor.rLog.info({
              msg: "persisting connection",
              connId,
              gatewayId: idToStr(hibernatableDataRaw.gatewayId),
              requestId: idToStr(hibernatableDataRaw.requestId),
              serverMessageIndex: hibernatableDataRaw.serverMessageIndex,
              clientMessageIndex: hibernatableDataRaw.clientMessageIndex,
              hasState: hibernatableDataRaw.state !== void 0
            });
            const bareData = convertConnToBarePersistedConn(
              hibernatableDataRaw
            );
            const connData = CONN_VERSIONED.serializeWithEmbeddedVersion(
              bareData,
              CURRENT_VERSION2
            );
            entries.push([makeConnKey(connId), connData]);
            connections.push(conn);
          }
          this.#actor.rLog.info({
            msg: "prepared entries for kvBatchPut",
            totalEntries: entries.length,
            connectionEntries: connections.length,
            connectionIds: connections.map((c) => c.id)
          });
          if (this.#actorDriver.onBeforePersistConn) {
            for (const conn of connections) {
              this.#actorDriver.onBeforePersistConn(conn);
            }
          }
          this.#actor.connectionManager.clearConnWithPersistChanged();
          this.#actor.rLog.info({
            msg: "calling kvBatchPut",
            actorId: this.#actor.id,
            entriesCount: entries.length
          });
          await this.#actorDriver.kvBatchPut(this.#actor.id, entries);
          this.#actor.rLog.info({
            msg: "kvBatchPut completed successfully"
          });
          if (this.#actorDriver.onAfterPersistConn) {
            for (const conn of connections) {
              this.#actorDriver.onAfterPersistConn(conn);
            }
          }
          this.#actor.rLog.debug({ msg: "persist saved" });
        });
      } else {
        this.#actor.rLog.info({
          msg: "savePersistInner skipped - no changes"
        });
      }
      (_a = this.#onPersistSavedPromise) == null ? void 0 : _a.resolve();
    } catch (error) {
      this.#actor.rLog.error({
        msg: "error saving persist",
        error: stringifyError(error)
      });
      (_b = this.#onPersistSavedPromise) == null ? void 0 : _b.reject(error);
      throw error;
    }
  }
};

// src/actor/instance/traces-driver.ts
function concatPrefix(prefix, key) {
  const merged = new Uint8Array(prefix.length + key.length);
  merged.set(prefix, 0);
  merged.set(key, prefix.length);
  return merged;
}
function stripPrefix(prefix, key) {
  return key.slice(prefix.length);
}
function compareBytes(a, b) {
  const len = Math.min(a.length, b.length);
  for (let i = 0; i < len; i++) {
    if (a[i] !== b[i]) {
      return a[i] - b[i];
    }
  }
  return a.length - b.length;
}
var ActorTracesDriver = class {
  #driver;
  #actorId;
  #prefix;
  constructor(driver, actorId) {
    this.#driver = driver;
    this.#actorId = actorId;
    this.#prefix = tracesStoragePrefix();
  }
  async get(key) {
    const [value] = await this.#driver.kvBatchGet(this.#actorId, [
      concatPrefix(this.#prefix, key)
    ]);
    return value ?? null;
  }
  async set(key, value) {
    await this.#driver.kvBatchPut(this.#actorId, [
      [concatPrefix(this.#prefix, key), value]
    ]);
  }
  async delete(key) {
    await this.#driver.kvBatchDelete(this.#actorId, [
      concatPrefix(this.#prefix, key)
    ]);
  }
  async deletePrefix(prefix) {
    const fullPrefix = concatPrefix(this.#prefix, prefix);
    const entries = await this.#driver.kvListPrefix(
      this.#actorId,
      fullPrefix
    );
    if (entries.length === 0) {
      return;
    }
    await this.#driver.kvBatchDelete(
      this.#actorId,
      entries.map(([key]) => key)
    );
  }
  async list(prefix) {
    const fullPrefix = concatPrefix(this.#prefix, prefix);
    const entries = await this.#driver.kvListPrefix(
      this.#actorId,
      fullPrefix
    );
    return entries.map(([key, value]) => ({
      key: stripPrefix(this.#prefix, key),
      value
    }));
  }
  async listRange(start, end, options) {
    const fullStart = concatPrefix(this.#prefix, start);
    const fullEnd = concatPrefix(this.#prefix, end);
    const entries = await this.#driver.kvListPrefix(
      this.#actorId,
      this.#prefix
    );
    const filtered = entries.filter(([key]) => {
      return compareBytes(key, fullStart) >= 0 && compareBytes(key, fullEnd) < 0;
    }).sort(([keyA], [keyB]) => compareBytes(keyA, keyB));
    if (options == null ? void 0 : options.reverse) {
      filtered.reverse();
    }
    const limited = (options == null ? void 0 : options.limit) ? filtered.slice(0, options.limit) : filtered;
    return limited.map(([key, value]) => ({
      key: stripPrefix(this.#prefix, key),
      value
    }));
  }
  async batch(writes) {
    if (writes.length === 0) {
      return;
    }
    await this.#driver.kvBatchPut(
      this.#actorId,
      writes.map(({ key, value }) => [
        concatPrefix(this.#prefix, key),
        value
      ])
    );
  }
};

// src/actor/instance/mod.ts
var CanSleep = /* @__PURE__ */ ((CanSleep2) => {
  CanSleep2[CanSleep2["Yes"] = 0] = "Yes";
  CanSleep2[CanSleep2["NotReady"] = 1] = "NotReady";
  CanSleep2[CanSleep2["NotStarted"] = 2] = "NotStarted";
  CanSleep2[CanSleep2["ActiveConns"] = 3] = "ActiveConns";
  CanSleep2[CanSleep2["ActiveDisconnectCallbacks"] = 4] = "ActiveDisconnectCallbacks";
  CanSleep2[CanSleep2["ActiveHonoHttpRequests"] = 5] = "ActiveHonoHttpRequests";
  CanSleep2[CanSleep2["ActiveKeepAwake"] = 6] = "ActiveKeepAwake";
  CanSleep2[CanSleep2["ActiveRun"] = 7] = "ActiveRun";
  return CanSleep2;
})(CanSleep || {});
var ActorInstance = class {
  // MARK: - Core Properties
  actorContext;
  #config;
  driver;
  #inlineClient;
  #actorId;
  #name;
  #key;
  #actorKeyString;
  #region;
  // MARK: - Managers
  connectionManager;
  stateManager;
  eventManager;
  #scheduleManager;
  queueManager;
  // MARK: - Logging
  #log;
  #rLog;
  // MARK: - Lifecycle State
  /**
   * If the core actor initiation has set up.
   *
   * Almost all actions on this actor will throw an error if false.
   **/
  #ready = false;
  /**
   * If the actor has fully started.
   *
   * The only purpose of this is to prevent sleeping until started.
   */
  #started = false;
  #sleepCalled = false;
  #destroyCalled = false;
  #stopCalled = false;
  #sleepTimeout;
  #abortController = new AbortController();
  // MARK: - Variables & Database
  #vars;
  #db;
  #sqliteVfs;
  // MARK: - Background Tasks
  #backgroundPromises = [];
  #runPromise;
  #runHandlerActive = false;
  #activeQueueWaitCount = 0;
  // MARK: - HTTP/WebSocket Tracking
  #activeHonoHttpRequests = 0;
  #activeKeepAwakeCount = 0;
  // MARK: - Deprecated (kept for compatibility)
  #schedule;
  // MARK: - Inspector
  #inspectorToken;
  #inspector;
  // MARK: - Tracing
  #traces;
  // MARK: - Driver Overrides
  /**
   * Per-instance config option overrides applied by the driver after creation.
   * When set, the effective option value is the minimum of the base config
   * value and the override value.
   */
  overrides = {};
  // MARK: - Constructor
  constructor(config4) {
    this.#config = config4;
    this.actorContext = new ActorContext(this);
    this.#inspector = new ActorInspector(this);
  }
  // MARK: - Public Getters
  get log() {
    invariant2(this.#log, "log not configured");
    return this.#log;
  }
  get rLog() {
    invariant2(this.#rLog, "log not configured");
    return this.#rLog;
  }
  get isStopping() {
    return this.#stopCalled;
  }
  get id() {
    return this.#actorId;
  }
  get name() {
    return this.#name;
  }
  get key() {
    return this.#key;
  }
  get region() {
    return this.#region;
  }
  get inlineClient() {
    return this.#inlineClient;
  }
  get inspector() {
    return this.#inspector;
  }
  get traces() {
    return this.#traces;
  }
  get inspectorToken() {
    return this.#inspectorToken;
  }
  // MARK: - Tracing
  getCurrentTraceSpan() {
    return this.#traces.getCurrentSpan();
  }
  startTraceSpan(name, attributes) {
    return this.#traces.startSpan(name, {
      parent: this.#traces.getCurrentSpan() ?? void 0,
      attributes: this.#traceAttributes(attributes)
    });
  }
  endTraceSpan(handle, status) {
    this.#traces.endSpan(handle, status ? { status } : void 0);
  }
  async runInTraceSpan(name, attributes, fn) {
    const span = this.startTraceSpan(name, attributes);
    try {
      const result = this.#traces.withSpan(span, fn);
      const resolved = result instanceof Promise ? await result : result;
      this.#traces.endSpan(span, {
        status: { code: "OK" }
      });
      return resolved;
    } catch (error) {
      this.#traces.endSpan(span, {
        status: {
          code: "ERROR",
          message: stringifyError(error)
        }
      });
      throw error;
    }
  }
  emitTraceEvent(name, attributes, handle) {
    const span = handle ?? this.#traces.getCurrentSpan();
    if (!span) {
      return;
    }
    this.#traces.emitEvent(span, name, {
      attributes: this.#traceAttributes(attributes),
      timeUnixMs: Date.now()
    });
  }
  get conns() {
    return this.connectionManager.connections;
  }
  get schedule() {
    return this.#schedule;
  }
  get abortSignal() {
    return this.#abortController.signal;
  }
  get actions() {
    return Object.keys(this.#config.actions ?? {});
  }
  get config() {
    return this.#config;
  }
  // MARK: - State Access
  get persist() {
    return this.stateManager.persist;
  }
  get state() {
    return this.stateManager.state;
  }
  set state(value) {
    this.stateManager.state = value;
  }
  get stateEnabled() {
    return this.stateManager.stateEnabled;
  }
  get connStateEnabled() {
    return "createConnState" in this.#config || "connState" in this.#config;
  }
  // MARK: - Variables & Database
  get vars() {
    this.#validateVarsEnabled();
    invariant2(this.#vars !== void 0, "vars not enabled");
    return this.#vars;
  }
  get db() {
    if (!this.#db) {
      throw new DatabaseNotEnabled();
    }
    return this.#db;
  }
  // MARK: - Initialization
  async start(actorDriver, inlineClient, actorId, name, key, region) {
    var _a, _b;
    this.driver = actorDriver;
    this.#inlineClient = inlineClient;
    this.#actorId = actorId;
    this.#name = name;
    this.#key = key;
    this.#actorKeyString = serializeActorKey(this.#key);
    this.#region = region;
    this.#initializeTraces();
    this.#initializeLogging();
    this.connectionManager = new ConnectionManager(this);
    this.stateManager = new StateManager(this, actorDriver, this.#config);
    this.eventManager = new EventManager(this);
    this.queueManager = new QueueManager(this, actorDriver);
    this.#scheduleManager = new ScheduleManager(
      this,
      actorDriver,
      this.#config
    );
    this.#schedule = new Schedule(this);
    await this.#loadState();
    await this.queueManager.initialize();
    await this.#initializeInspectorToken();
    if (this.#varsEnabled) {
      await this.#initializeVars();
    }
    await this.#callOnStart();
    await this.#setupDatabase();
    await this.#scheduleManager.initializeAlarms();
    this.#ready = true;
    await ((_b = (_a = this.driver).onBeforeActorStart) == null ? void 0 : _b.call(_a, this));
    this.#started = true;
    this.#rLog.info({ msg: "actor started" });
    this.resetSleepTimer();
    this.#startRunHandler();
    await this.onAlarm();
  }
  // MARK: - Ready Check
  isReady() {
    return this.#ready;
  }
  assertReady(allowStoppingState = false) {
    if (!this.#ready) throw new InternalError("Actor not ready");
    if (!allowStoppingState && this.#stopCalled)
      throw new InternalError("Actor is stopping");
  }
  async cleanupPersistedConnections(reason) {
    this.assertReady(true);
    return await this.connectionManager.cleanupPersistedHibernatableConnections(
      reason
    );
  }
  // MARK: - Stop
  async onStop(mode) {
    if (this.#stopCalled) {
      this.#rLog.warn({ msg: "already stopping actor" });
      return;
    }
    this.#stopCalled = true;
    this.#rLog.info({
      msg: "setting stopCalled=true",
      mode
    });
    try {
      if (this.#sleepTimeout) {
        clearTimeout(this.#sleepTimeout);
        this.#sleepTimeout = void 0;
      }
      try {
        this.#abortController.abort();
      } catch {
      }
      await this.#waitForRunHandler(this.overrides.runStopTimeout !== void 0 ? Math.min(this.#config.options.runStopTimeout, this.overrides.runStopTimeout) : this.#config.options.runStopTimeout);
      if (mode === "sleep") {
        await this.#callOnSleep();
      } else if (mode === "destroy") {
        await this.#callOnDestroy();
      } else {
        assertUnreachable(mode);
      }
      await this.#disconnectConnections();
      await this.#waitBackgroundPromises(
        this.overrides.waitUntilTimeout !== void 0 ? Math.min(this.#config.options.waitUntilTimeout, this.overrides.waitUntilTimeout) : this.#config.options.waitUntilTimeout
      );
      this.#rLog.info({ msg: "clearing pending save timeouts" });
      this.stateManager.clearPendingSaveTimeout();
      this.#rLog.info({ msg: "saving state immediately" });
      await this.stateManager.saveState({
        immediate: true,
        allowStoppingState: true
      });
      await this.stateManager.waitForPendingWrites();
      await this.#scheduleManager.waitForPendingAlarmWrites();
    } finally {
      await this.#cleanupDatabase();
    }
  }
  // MARK: - Sleep
  startSleep() {
    var _a;
    if (this.#stopCalled || this.#destroyCalled) {
      this.#rLog.debug({
        msg: "cannot call startSleep if actor already stopping"
      });
      return;
    }
    if (this.#sleepCalled) {
      this.#rLog.warn({
        msg: "cannot call startSleep twice, actor already sleeping"
      });
      return;
    }
    this.#sleepCalled = true;
    const sleep = (_a = this.driver.startSleep) == null ? void 0 : _a.bind(this.driver, this.#actorId);
    invariant2(this.#sleepingSupported, "sleeping not supported");
    invariant2(sleep, "no sleep on driver");
    this.#rLog.info({ msg: "actor sleeping" });
    setImmediate(() => {
      sleep();
    });
  }
  // MARK: - Destroy
  startDestroy() {
    if (this.#stopCalled || this.#sleepCalled) {
      this.#rLog.debug({
        msg: "cannot call startDestroy if actor already stopping or sleeping"
      });
      return;
    }
    if (this.#destroyCalled) {
      this.#rLog.warn({
        msg: "cannot call startDestroy twice, actor already destroying"
      });
      return;
    }
    this.#destroyCalled = true;
    try {
      this.#abortController.abort();
    } catch {
    }
    const destroy = this.driver.startDestroy.bind(
      this.driver,
      this.#actorId
    );
    this.#rLog.info({ msg: "actor destroying" });
    setImmediate(() => {
      destroy();
    });
  }
  // MARK: - HTTP Request Tracking
  beginHonoHttpRequest() {
    this.#activeHonoHttpRequests++;
    this.resetSleepTimer();
  }
  endHonoHttpRequest() {
    this.#activeHonoHttpRequests--;
    if (this.#activeHonoHttpRequests < 0) {
      this.#activeHonoHttpRequests = 0;
      this.#rLog.warn({
        msg: "active hono requests went below 0, this is a RivetKit bug",
        ...EXTRA_ERROR_LOG
      });
    }
    this.resetSleepTimer();
  }
  // MARK: - Message Processing
  async processMessage(message, conn) {
    await processMessage(message, this, conn, {
      onExecuteAction: async (ctx, name, args) => {
        return await this.executeAction(ctx, name, args);
      },
      onSubscribe: async (eventName, conn2) => {
        this.eventManager.addSubscription(eventName, conn2, false);
      },
      onUnsubscribe: async (eventName, conn2) => {
        this.eventManager.removeSubscription(eventName, conn2, false);
      }
    });
  }
  async assertCanSubscribe(ctx, eventName) {
    const canSubscribe = getEventCanSubscribe(this.#config.events, eventName);
    if (!canSubscribe) {
      return;
    }
    const result = await canSubscribe(ctx);
    if (typeof result !== "boolean") {
      throw new InvalidCanSubscribeResponse();
    }
    if (!result) {
      throw new Forbidden();
    }
  }
  async assertCanPublish(ctx, queueName) {
    const canPublish = getQueueCanPublish(this.#config.queues, queueName);
    if (!canPublish) {
      return;
    }
    const result = await canPublish(ctx);
    if (typeof result !== "boolean") {
      throw new InvalidCanPublishResponse();
    }
    if (!result) {
      throw new Forbidden();
    }
  }
  // MARK: - Action Execution
  async executeAction(ctx, actionName, args) {
    this.assertReady();
    const actions = this.#config.actions ?? {};
    if (!(actionName in actions)) {
      this.#rLog.warn({ msg: "action does not exist", actionName });
      throw new ActionNotFound(actionName);
    }
    const actionFunction = actions[actionName];
    if (typeof actionFunction !== "function") {
      this.#rLog.warn({
        msg: "action is not a function",
        actionName,
        type: typeof actionFunction
      });
      throw new ActionNotFound(actionName);
    }
    this.#activeKeepAwakeCount++;
    this.resetSleepTimer();
    const actionSpan = this.startTraceSpan(`actor.action.${actionName}`, {
      "rivet.action.name": actionName
    });
    let spanEnded = false;
    try {
      return await this.#traces.withSpan(actionSpan, async () => {
        this.#rLog.debug({
          msg: "executing action",
          actionName,
          args
        });
        const outputOrPromise = actionFunction.call(
          void 0,
          ctx,
          ...args
        );
        let output;
        const maybeThenable = outputOrPromise;
        if (maybeThenable && typeof maybeThenable.then === "function") {
          output = await deadline(
            Promise.resolve(outputOrPromise),
            this.#config.options.actionTimeout
          );
        } else {
          output = outputOrPromise;
        }
        if (this.#config.onBeforeActionResponse) {
          try {
            output = await this.runInTraceSpan(
              "actor.onBeforeActionResponse",
              { "rivet.action.name": actionName },
              () => this.#config.onBeforeActionResponse(
                this.actorContext,
                actionName,
                args,
                output
              )
            );
          } catch (error) {
            this.#rLog.error({
              msg: "error in `onBeforeActionResponse`",
              error: stringifyError(error)
            });
          }
        }
        return output;
      });
    } catch (error) {
      const isTimeout = error instanceof DeadlineError;
      const message = isTimeout ? "ActionTimedOut" : stringifyError(error);
      this.#traces.setAttributes(actionSpan, {
        "error.message": message,
        "error.type": error instanceof Error ? error.name : typeof error
      });
      this.#traces.endSpan(actionSpan, {
        status: { code: "ERROR", message }
      });
      spanEnded = true;
      if (isTimeout) {
        throw new ActionTimedOut();
      }
      this.#rLog.error({
        msg: "action error",
        actionName,
        error: stringifyError(error)
      });
      throw error;
    } finally {
      if (!spanEnded && actionSpan.isActive()) {
        this.#traces.endSpan(actionSpan, {
          status: { code: "OK" }
        });
      }
      this.#activeKeepAwakeCount--;
      if (this.#activeKeepAwakeCount < 0) {
        this.#activeKeepAwakeCount = 0;
        this.#rLog.warn({
          msg: "active keep awake count went below 0, this is a RivetKit bug",
          ...EXTRA_ERROR_LOG
        });
      }
      this.resetSleepTimer();
      this.stateManager.savePersistThrottled();
    }
  }
  // MARK: - HTTP/WebSocket Handlers
  async handleRawRequest(conn, request) {
    this.assertReady();
    if (!this.#config.onRequest) {
      throw new RequestHandlerNotDefined();
    }
    const onRequest = this.#config.onRequest;
    return await this.runInTraceSpan(
      "actor.onRequest",
      {
        "http.method": request.method,
        "http.url": request.url,
        "rivet.conn.id": conn.id
      },
      async () => {
        const ctx = new RequestContext(this, conn, request);
        try {
          const response = await onRequest(ctx, request);
          if (!response) {
            throw new InvalidRequestHandlerResponse();
          }
          return response;
        } catch (error) {
          this.#rLog.error({
            msg: "onRequest error",
            error: stringifyError(error)
          });
          throw error;
        } finally {
          this.stateManager.savePersistThrottled();
        }
      }
    );
  }
  handleRawWebSocket(conn, websocket, request) {
    this.assertReady();
    if (!this.#config.onWebSocket) {
      throw new InternalError("onWebSocket handler not defined");
    }
    const span = this.startTraceSpan("actor.onWebSocket", {
      "http.url": request == null ? void 0 : request.url,
      "rivet.conn.id": conn.id
    });
    let spanEnded = false;
    try {
      this.resetSleepTimer();
      const ctx = new WebSocketContext(this, conn, request);
      const voidOrPromise = this.#traces.withSpan(
        span,
        () => this.#config.onWebSocket(ctx, websocket)
      );
      if (voidOrPromise instanceof Promise) {
        voidOrPromise.then(() => {
          if (!spanEnded) {
            this.endTraceSpan(span, { code: "OK" });
            spanEnded = true;
          }
        }).catch((error) => {
          if (!spanEnded) {
            this.endTraceSpan(span, {
              code: "ERROR",
              message: stringifyError(error)
            });
            spanEnded = true;
          }
          this.#rLog.error({
            msg: "onWebSocket error",
            error: stringifyError(error)
          });
        }).finally(() => {
          this.stateManager.savePersistThrottled();
        });
      } else {
        if (!spanEnded) {
          this.endTraceSpan(span, { code: "OK" });
          spanEnded = true;
        }
        this.stateManager.savePersistThrottled();
      }
    } catch (error) {
      if (!spanEnded) {
        this.endTraceSpan(span, {
          code: "ERROR",
          message: stringifyError(error)
        });
        spanEnded = true;
      }
      this.#rLog.error({
        msg: "onWebSocket error",
        error: stringifyError(error)
      });
      throw error;
    }
  }
  // MARK: - Scheduling
  async scheduleEvent(timestamp, action, args) {
    await this.#scheduleManager.scheduleEvent(timestamp, action, args);
  }
  async onAlarm() {
    this.resetSleepTimer();
    await this.#scheduleManager.onAlarm();
  }
  // MARK: - Background Tasks
  waitUntil(promise) {
    this.assertReady();
    const nonfailablePromise = promise.then(() => {
      this.#rLog.debug({ msg: "wait until promise complete" });
    }).catch((error) => {
      this.#rLog.error({
        msg: "wait until promise failed",
        error: stringifyError(error)
      });
    });
    this.#backgroundPromises.push(nonfailablePromise);
  }
  /**
   * Prevents the actor from sleeping while the given promise is running.
   *
   * Use this when performing async operations in the `run` handler or other
   * background contexts where you need to ensure the actor stays awake.
   *
   * Returns the resolved value and resets the sleep timer on completion.
   * Errors are propagated to the caller.
   */
  async keepAwake(promise) {
    this.assertReady(true);
    this.#activeKeepAwakeCount++;
    this.resetSleepTimer();
    try {
      return await promise;
    } finally {
      this.#activeKeepAwakeCount--;
      if (this.#activeKeepAwakeCount < 0) {
        this.#activeKeepAwakeCount = 0;
        this.#rLog.warn({
          msg: "active keep awake count went below 0, this is a RivetKit bug",
          ...EXTRA_ERROR_LOG
        });
      }
      this.resetSleepTimer();
    }
  }
  beginQueueWait() {
    this.assertReady(true);
    this.#activeQueueWaitCount++;
    this.resetSleepTimer();
  }
  endQueueWait() {
    this.#activeQueueWaitCount--;
    if (this.#activeQueueWaitCount < 0) {
      this.#activeQueueWaitCount = 0;
      this.#rLog.warn({
        msg: "active queue wait count went below 0, this is a RivetKit bug",
        ...EXTRA_ERROR_LOG
      });
    }
    this.resetSleepTimer();
  }
  // MARK: - Private Helper Methods
  #initializeTraces() {
    if (getRivetExperimentalOtel()) {
      this.#traces = createTraces({
        driver: new ActorTracesDriver(this.driver, this.#actorId)
      });
    } else {
      this.#traces = createNoopTraces();
    }
  }
  #traceAttributes(attributes) {
    return {
      "rivet.actor.id": this.#actorId,
      "rivet.actor.name": this.#name,
      "rivet.actor.key": this.#actorKeyString,
      "rivet.actor.region": this.#region,
      ...attributes ?? {}
    };
  }
  #patchLoggerForTraces(logger3) {
    const levels = ["trace", "debug", "info", "warn", "error", "fatal"];
    for (const level of levels) {
      const original = logger3[level].bind(logger3);
      logger3[level] = ((...args) => {
        this.#emitLogEvent(level, args);
        return original(...args);
      });
    }
  }
  #emitLogEvent(level, args) {
    const span = this.#traces.getCurrentSpan();
    if (!span || !span.isActive()) {
      return;
    }
    let message;
    if (args.length >= 2) {
      message = String(args[1]);
    } else if (args.length === 1) {
      const [value] = args;
      if (typeof value === "string") {
        message = value;
      } else if (typeof value === "number" || typeof value === "boolean") {
        message = String(value);
      } else if (value && typeof value === "object") {
        const maybeMsg = value.msg;
        if (maybeMsg !== void 0) {
          message = String(maybeMsg);
        }
      }
    }
    this.#traces.emitEvent(span, "log", {
      attributes: this.#traceAttributes({
        "log.level": level,
        ...message ? { "log.message": message } : {}
      }),
      timeUnixMs: Date.now()
    });
  }
  #initializeLogging() {
    var _a, _b;
    const logParams = {
      actor: this.#name,
      key: this.#actorKeyString,
      actorId: this.#actorId
    };
    const extraLogParams = (_b = (_a = this.driver).getExtraActorLogParams) == null ? void 0 : _b.call(_a);
    if (extraLogParams) Object.assign(logParams, extraLogParams);
    this.#log = getBaseLogger().child(
      Object.assign(
        getIncludeTarget() ? { target: "actor" } : {},
        logParams
      )
    );
    this.#rLog = getBaseLogger().child(
      Object.assign(
        getIncludeTarget() ? { target: "actor-runtime" } : {},
        logParams
      )
    );
    this.#patchLoggerForTraces(this.#log);
    this.#patchLoggerForTraces(this.#rLog);
  }
  async #loadState() {
    const [persistDataBuffer] = await this.driver.kvBatchGet(
      this.#actorId,
      [KEYS.PERSIST_DATA]
    );
    invariant2(
      persistDataBuffer !== null,
      "persist data has not been set, it should be set when initialized"
    );
    const bareData = ACTOR_VERSIONED.deserializeWithEmbeddedVersion(persistDataBuffer);
    const persistData = convertActorFromBarePersisted(bareData);
    if (persistData.hasInitialized) {
      await this.#restoreExistingActor(persistData);
    } else {
      await this.#createNewActor(persistData);
    }
    this.#scheduleManager.setPersist(this.stateManager.persist);
  }
  async #createNewActor(persistData) {
    this.#rLog.info({ msg: "actor creating" });
    await this.stateManager.initializeState(persistData);
    if (this.#config.onCreate) {
      const onCreate = this.#config.onCreate;
      await this.runInTraceSpan(
        "actor.onCreate",
        void 0,
        () => onCreate(this.actorContext, persistData.input)
      );
    }
  }
  async #restoreExistingActor(persistData) {
    const connEntries = await this.driver.kvListPrefix(
      this.#actorId,
      KEYS.CONN_PREFIX
    );
    const connections = [];
    for (const [_key, value] of connEntries) {
      try {
        const bareData = CONN_VERSIONED.deserializeWithEmbeddedVersion(
          new Uint8Array(value)
        );
        const conn = convertConnFromBarePersistedConn(bareData);
        connections.push(conn);
      } catch (error) {
        this.#rLog.error({
          msg: "failed to decode connection",
          error: stringifyError(error)
        });
      }
    }
    this.#rLog.info({
      msg: "actor restoring",
      connections: connections.length
    });
    this.stateManager.initPersistProxy(persistData);
    this.connectionManager.restoreConnections(connections);
  }
  async #initializeInspectorToken() {
    const [tokenBuffer] = await this.driver.kvBatchGet(this.#actorId, [
      KEYS.INSPECTOR_TOKEN
    ]);
    if (tokenBuffer !== null) {
      const decoder = new TextDecoder();
      this.#inspectorToken = decoder.decode(tokenBuffer);
      this.#rLog.debug({ msg: "loaded existing inspector token" });
    } else {
      this.#inspectorToken = generateSecureToken();
      const tokenBytes = new TextEncoder().encode(this.#inspectorToken);
      await this.driver.kvBatchPut(this.#actorId, [
        [KEYS.INSPECTOR_TOKEN, tokenBytes]
      ]);
      this.#rLog.debug({ msg: "generated new inspector token" });
    }
  }
  async #initializeVars() {
    let vars;
    if ("createVars" in this.#config) {
      const createVars = this.#config.createVars;
      vars = await this.runInTraceSpan(
        "actor.createVars",
        void 0,
        () => {
          const dataOrPromise = createVars(
            this.actorContext,
            this.driver.getContext(this.#actorId)
          );
          if (dataOrPromise instanceof Promise) {
            return deadline(
              dataOrPromise,
              this.#config.options.createVarsTimeout
            );
          }
          return dataOrPromise;
        }
      );
    } else if ("vars" in this.#config) {
      vars = structuredClone(this.#config.vars);
    } else {
      throw new Error(
        "Could not create variables from 'createVars' or 'vars'"
      );
    }
    this.#vars = vars;
  }
  async #callOnStart() {
    this.#rLog.info({ msg: "actor starting" });
    if (this.#config.onWake) {
      const onWake = this.#config.onWake;
      await this.runInTraceSpan(
        "actor.onWake",
        void 0,
        () => onWake(this.actorContext)
      );
    }
  }
  async #callOnSleep() {
    if (this.#config.onSleep) {
      const onSleep = this.#config.onSleep;
      try {
        this.#rLog.debug({ msg: "calling onSleep" });
        await this.runInTraceSpan(
          "actor.onSleep",
          void 0,
          async () => {
            const result = onSleep(this.actorContext);
            if (result instanceof Promise) {
              await deadline(
                result,
                this.overrides.onSleepTimeout !== void 0 ? Math.min(this.#config.options.onSleepTimeout, this.overrides.onSleepTimeout) : this.#config.options.onSleepTimeout
              );
            }
          }
        );
        this.#rLog.debug({ msg: "onSleep completed" });
      } catch (error) {
        if (error instanceof DeadlineError) {
          this.#rLog.error({ msg: "onSleep timed out" });
        } else {
          this.#rLog.error({
            msg: "error in onSleep",
            error: stringifyError(error)
          });
        }
      }
    }
  }
  async #callOnDestroy() {
    if (this.#config.onDestroy) {
      const onDestroy = this.#config.onDestroy;
      try {
        this.#rLog.debug({ msg: "calling onDestroy" });
        await this.runInTraceSpan(
          "actor.onDestroy",
          void 0,
          async () => {
            const result = onDestroy(this.actorContext);
            if (result instanceof Promise) {
              await deadline(
                result,
                this.overrides.onDestroyTimeout !== void 0 ? Math.min(this.#config.options.onDestroyTimeout, this.overrides.onDestroyTimeout) : this.#config.options.onDestroyTimeout
              );
            }
          }
        );
        this.#rLog.debug({ msg: "onDestroy completed" });
      } catch (error) {
        if (error instanceof DeadlineError) {
          this.#rLog.error({ msg: "onDestroy timed out" });
        } else {
          this.#rLog.error({
            msg: "error in onDestroy",
            error: stringifyError(error)
          });
        }
      }
    }
  }
  #startRunHandler() {
    const runFn = getRunFunction(this.#config.run);
    if (!runFn) return;
    this.#rLog.debug({ msg: "starting run handler" });
    this.#runHandlerActive = true;
    this.resetSleepTimer();
    const runSpan = this.startTraceSpan("actor.run");
    const runResult = this.#traces.withSpan(
      runSpan,
      () => runFn(this.actorContext)
    );
    if (runResult instanceof Promise) {
      this.#runPromise = runResult.then(() => {
        if (this.#stopCalled) {
          if (runSpan.isActive()) {
            this.endTraceSpan(runSpan, { code: "OK" });
          }
          this.#rLog.debug({
            msg: "run handler exited during actor stop"
          });
          return;
        }
        this.emitTraceEvent(
          "actor.crash",
          { "rivet.actor.reason": "run_exited" },
          runSpan
        );
        this.endTraceSpan(runSpan, {
          code: "ERROR",
          message: "run exited unexpectedly"
        });
        this.#rLog.warn({
          msg: "run handler exited unexpectedly, crashing actor to reschedule"
        });
        this.startDestroy();
      }).catch((error) => {
        if (this.#stopCalled) {
          if (runSpan.isActive()) {
            this.endTraceSpan(runSpan, { code: "OK" });
          }
          this.#rLog.debug({
            msg: "run handler threw during actor stop",
            error: stringifyError(error)
          });
          return;
        }
        this.emitTraceEvent(
          "actor.crash",
          {
            "rivet.actor.reason": "run_error",
            "error.message": stringifyError(error)
          },
          runSpan
        );
        this.endTraceSpan(runSpan, {
          code: "ERROR",
          message: stringifyError(error)
        });
        this.#rLog.error({
          msg: "run handler threw error, crashing actor to reschedule",
          error: stringifyError(error)
        });
        this.startDestroy();
      }).finally(() => {
        this.#runHandlerActive = false;
        this.resetSleepTimer();
      });
    } else if (runSpan.isActive()) {
      this.endTraceSpan(runSpan, { code: "OK" });
      this.#runHandlerActive = false;
      this.resetSleepTimer();
    }
  }
  async #waitForRunHandler(timeoutMs) {
    if (!this.#runPromise) {
      return;
    }
    this.#rLog.debug({ msg: "waiting for run handler to complete" });
    const timedOut = await Promise.race([
      this.#runPromise.then(() => false).catch(() => false),
      new Promise(
        (resolve) => setTimeout(() => resolve(true), timeoutMs)
      )
    ]);
    if (timedOut) {
      this.#rLog.warn({
        msg: "run handler did not complete in time, it may have leaked - ensure you use c.aborted (or the abort signal c.abortSignal) to exit gracefully",
        timeoutMs
      });
    } else {
      this.#rLog.debug({ msg: "run handler completed" });
    }
  }
  async #setupDatabase() {
    var _a, _b, _c, _d;
    if (!("db" in this.#config) || !this.#config.db) {
      return;
    }
    let client;
    try {
      if (!this.#sqliteVfs && this.driver.createSqliteVfs) {
        this.#sqliteVfs = await this.driver.createSqliteVfs();
      }
      client = await this.#config.db.createClient({
        actorId: this.#actorId,
        overrideRawDatabaseClient: this.driver.overrideRawDatabaseClient ? () => this.driver.overrideRawDatabaseClient(this.#actorId) : void 0,
        overrideDrizzleDatabaseClient: this.driver.overrideDrizzleDatabaseClient ? () => this.driver.overrideDrizzleDatabaseClient(this.#actorId) : void 0,
        kv: {
          batchPut: (entries) => this.driver.kvBatchPut(this.#actorId, entries),
          batchGet: (keys) => this.driver.kvBatchGet(this.#actorId, keys),
          batchDelete: (keys) => this.driver.kvBatchDelete(this.#actorId, keys)
        },
        sqliteVfs: this.#sqliteVfs
      });
      this.#rLog.info({ msg: "database migration starting" });
      await ((_b = (_a = this.#config.db).onMigrate) == null ? void 0 : _b.call(_a, client));
      this.#rLog.info({ msg: "database migration complete" });
      this.#db = client;
    } catch (error) {
      if (client) {
        try {
          await ((_d = (_c = this.#config.db).onDestroy) == null ? void 0 : _d.call(_c, client));
        } catch (cleanupError) {
          this.#rLog.error({
            msg: "database setup cleanup failed",
            error: stringifyError(cleanupError)
          });
        }
      }
      if (this.#sqliteVfs) {
        try {
          await this.#sqliteVfs.destroy();
        } catch (cleanupError) {
          this.#rLog.error({
            msg: "sqlite vfs teardown after setup failure failed",
            error: stringifyError(cleanupError)
          });
        }
      }
      this.#sqliteVfs = void 0;
      if (error instanceof Error) {
        this.#rLog.error({
          msg: "database setup failed",
          error: stringifyError(error)
        });
        throw error;
      }
      const wrappedError = new Error(`Database setup failed: ${String(error)}`);
      this.#rLog.error({
        msg: "database setup failed with non-Error object",
        error: String(error),
        errorType: typeof error
      });
      throw wrappedError;
    }
  }
  async #cleanupDatabase() {
    var _a;
    const client = this.#db;
    const sqliteVfs = this.#sqliteVfs;
    const dbConfig = "db" in this.#config ? this.#config.db : void 0;
    this.#db = void 0;
    this.#sqliteVfs = void 0;
    if (client && dbConfig) {
      try {
        await ((_a = dbConfig.onDestroy) == null ? void 0 : _a.call(dbConfig, client));
      } catch (error) {
        this.#rLog.error({
          msg: "database cleanup failed",
          error: stringifyError(error)
        });
      }
    }
    if (sqliteVfs) {
      try {
        await sqliteVfs.destroy();
      } catch (error) {
        this.#rLog.error({
          msg: "sqlite vfs cleanup failed",
          error: stringifyError(error)
        });
      }
    }
  }
  async #disconnectConnections() {
    const promises = [];
    this.#rLog.debug({
      msg: "disconnecting connections on actor stop",
      totalConns: this.connectionManager.connections.size
    });
    for (const connection of this.connectionManager.connections.values()) {
      this.#rLog.debug({
        msg: "checking connection for disconnect",
        connId: connection.id,
        isHibernatable: connection.isHibernatable
      });
      if (!connection.isHibernatable) {
        this.#rLog.debug({
          msg: "disconnecting non-hibernatable connection on actor stop",
          connId: connection.id
        });
        promises.push(connection.disconnect());
      } else {
        this.#rLog.debug({
          msg: "preserving hibernatable connection on actor stop",
          connId: connection.id
        });
      }
    }
    const res = await Promise.race([
      Promise.all(promises).then(() => false),
      new Promise(
        (res2) => globalThis.setTimeout(() => res2(true), 1500)
      )
    ]);
    if (res) {
      this.#rLog.warn({
        msg: "timed out waiting for connections to close, shutting down anyway"
      });
    }
  }
  async #waitBackgroundPromises(timeoutMs) {
    const pending = this.#backgroundPromises;
    if (pending.length === 0) {
      this.#rLog.debug({ msg: "no background promises" });
      return;
    }
    const timedOut = await Promise.race([
      Promise.allSettled(pending).then(() => false),
      new Promise(
        (resolve) => setTimeout(() => resolve(true), timeoutMs)
      )
    ]);
    if (timedOut) {
      this.#rLog.error({
        msg: "timed out waiting for background tasks",
        count: pending.length,
        timeoutMs
      });
    } else {
      this.#rLog.debug({ msg: "background promises finished" });
    }
  }
  resetSleepTimer() {
    if (this.#config.options.noSleep || !this.#sleepingSupported) return;
    if (this.#stopCalled) return;
    const canSleep = this.#canSleep();
    this.#rLog.debug({
      msg: "resetting sleep timer",
      canSleep: CanSleep[canSleep],
      existingTimeout: !!this.#sleepTimeout,
      timeout: this.#config.options.sleepTimeout
    });
    if (this.#sleepTimeout) {
      clearTimeout(this.#sleepTimeout);
      this.#sleepTimeout = void 0;
    }
    if (this.#sleepCalled) return;
    if (canSleep === 0 /* Yes */) {
      this.#sleepTimeout = setTimeout(() => {
        this.startSleep();
      }, this.#config.options.sleepTimeout);
    }
  }
  #canSleep() {
    if (!this.#ready) return 1 /* NotReady */;
    if (!this.#started) return 1 /* NotReady */;
    if (this.#activeHonoHttpRequests > 0)
      return 5 /* ActiveHonoHttpRequests */;
    if (this.#activeKeepAwakeCount > 0) return 6 /* ActiveKeepAwake */;
    if (this.#runHandlerActive && this.#activeQueueWaitCount === 0) {
      return 7 /* ActiveRun */;
    }
    for (const _conn of this.connectionManager.connections.values()) {
      return 3 /* ActiveConns */;
    }
    if (this.connectionManager.pendingDisconnectCount > 0) {
      return 4 /* ActiveDisconnectCallbacks */;
    }
    return 0 /* Yes */;
  }
  get #sleepingSupported() {
    return this.driver.startSleep !== void 0;
  }
  get #varsEnabled() {
    return "createVars" in this.#config || "vars" in this.#config;
  }
  #validateVarsEnabled() {
    if (!this.#varsEnabled) {
      throw new VarsNotEnabled();
    }
  }
};

// src/actor/definition.ts
var ActorDefinition = class {
  #config;
  constructor(config4) {
    this.#config = config4;
  }
  get config() {
    return this.#config;
  }
  instantiate() {
    return new ActorInstance(this.#config);
  }
};
function lookupInRegistry(config4, name) {
  const definition = config4.use[name];
  if (!definition) throw new Error(`no actor in registry for name ${name}`);
  return definition;
}

// src/actor/router.ts
import { Hono } from "hono";
function createActorRouter(config4, actorDriver, getUpgradeWebSocket, isTest) {
  const router = new Hono({
    strict: false
  });
  router.use("*", loggerMiddleware(loggerWithoutContext()));
  router.use("*", async (c, next) => {
    const actor = await actorDriver.loadActor(c.env.actorId);
    actor.beginHonoHttpRequest();
    try {
      await next();
    } finally {
      actor.endHonoHttpRequest();
    }
  });
  router.get("/", (c) => {
    return c.text(
      "This is an RivetKit actor.\n\nLearn more at https://rivetkit.org"
    );
  });
  router.get("/health", (c) => {
    return c.text("ok");
  });
  router.get("/metadata", async (c) => {
    return c.json({
      runtime: "rivetkit",
      version: VERSION
    });
  });
  if (isTest) {
    router.post("/.test/force-disconnect", async (c) => {
      var _a;
      const connId = c.req.query("conn");
      if (!connId) {
        return c.text("Missing conn query parameter", 400);
      }
      const actor = await actorDriver.loadActor(c.env.actorId);
      const conn = actor.connectionManager.getConnForId(connId);
      if (!conn) {
        return c.text(`Connection not found: ${connId}`, 404);
      }
      if ((_a = conn[CONN_DRIVER_SYMBOL]) == null ? void 0 : _a.terminate) {
        conn[CONN_DRIVER_SYMBOL].terminate(actor, conn);
      }
      return c.json({ success: true });
    });
  }
  if (getUpgradeWebSocket) {
    router.on(
      "GET",
      [PATH_CONNECT, `${PATH_WEBSOCKET_PREFIX}*`, PATH_INSPECTOR_CONNECT],
      async (c) => {
        const upgradeWebSocket = getUpgradeWebSocket();
        if (upgradeWebSocket) {
          return upgradeWebSocket(async (c2) => {
            const protocols = c2.req.header(
              "sec-websocket-protocol"
            );
            const { encoding, connParams } = parseWebSocketProtocols(protocols);
            return await routeWebSocket(
              c2.req.raw,
              c2.req.path,
              c2.req.header(),
              config4,
              actorDriver,
              c2.env.actorId,
              encoding,
              connParams,
              void 0,
              void 0,
              false,
              false
            );
          })(c, noopNext());
        } else {
          return c.text(
            "WebSockets are not enabled for this driver.",
            400
          );
        }
      }
    );
  }
  if (config4.inspector.enabled) {
    const inspectorAuth = async (c) => {
      var _a;
      if (isDev() && !config4.inspector.token()) {
        inspectorLogger().warn({
          msg: "RIVET_INSPECTOR_TOKEN is not set, skipping inspector auth in development mode"
        });
        return void 0;
      }
      const userToken = (_a = c.req.header("Authorization")) == null ? void 0 : _a.replace("Bearer ", "");
      if (!userToken) {
        return c.text("Unauthorized", 401);
      }
      const inspectorToken = config4.inspector.token();
      if (!inspectorToken) {
        return c.text("Unauthorized", 401);
      }
      if (!timingSafeEqual(userToken, inspectorToken)) {
        return c.text("Unauthorized", 401);
      }
      return void 0;
    };
    router.get("/inspector/state", async (c) => {
      const authResponse = await inspectorAuth(c);
      if (authResponse) return authResponse;
      const actor = await actorDriver.loadActor(c.env.actorId);
      const isStateEnabled = actor.inspector.isStateEnabled();
      const state = isStateEnabled ? actor.inspector.getStateJson() : void 0;
      return c.json({ state, isStateEnabled });
    });
    router.patch("/inspector/state", async (c) => {
      const authResponse = await inspectorAuth(c);
      if (authResponse) return authResponse;
      const actor = await actorDriver.loadActor(c.env.actorId);
      const body = await c.req.json();
      await actor.inspector.setStateJson(body.state);
      return c.json({ ok: true });
    });
    router.get("/inspector/connections", async (c) => {
      const authResponse = await inspectorAuth(c);
      if (authResponse) return authResponse;
      const actor = await actorDriver.loadActor(c.env.actorId);
      const connections = actor.inspector.getConnectionsJson();
      return c.json({ connections });
    });
    router.get("/inspector/rpcs", async (c) => {
      const authResponse = await inspectorAuth(c);
      if (authResponse) return authResponse;
      const actor = await actorDriver.loadActor(c.env.actorId);
      const rpcs = actor.inspector.getRpcs();
      return c.json({ rpcs });
    });
    router.post("/inspector/action/:name", async (c) => {
      const authResponse = await inspectorAuth(c);
      if (authResponse) return authResponse;
      const actor = await actorDriver.loadActor(c.env.actorId);
      const name = c.req.param("name");
      const body = await c.req.json();
      const output = await actor.inspector.executeActionJson(
        name,
        body.args ?? []
      );
      return c.json({ output });
    });
    router.get("/inspector/queue", async (c) => {
      const authResponse = await inspectorAuth(c);
      if (authResponse) return authResponse;
      const actor = await actorDriver.loadActor(c.env.actorId);
      const limit = parseInt(c.req.query("limit") ?? "50", 10);
      const status = await actor.inspector.getQueueStatusJson(limit);
      return c.json(status);
    });
    router.get("/inspector/traces", async (c) => {
      const authResponse = await inspectorAuth(c);
      if (authResponse) return authResponse;
      const actor = await actorDriver.loadActor(c.env.actorId);
      const startMs = parseInt(c.req.query("startMs") ?? "0", 10);
      const endMs = parseInt(
        c.req.query("endMs") ?? String(Date.now()),
        10
      );
      const limit = parseInt(c.req.query("limit") ?? "1000", 10);
      await actor.traces.flush();
      const result = await actor.inspector.getTracesJson({
        startMs,
        endMs,
        limit
      });
      return c.json(result);
    });
    router.get("/inspector/workflow-history", async (c) => {
      const authResponse = await inspectorAuth(c);
      if (authResponse) return authResponse;
      const actor = await actorDriver.loadActor(c.env.actorId);
      const result = actor.inspector.getWorkflowHistoryJson();
      return c.json(result);
    });
    router.get("/inspector/summary", async (c) => {
      const authResponse = await inspectorAuth(c);
      if (authResponse) return authResponse;
      const actor = await actorDriver.loadActor(c.env.actorId);
      const isStateEnabled = actor.inspector.isStateEnabled();
      const isDatabaseEnabled = actor.inspector.isDatabaseEnabled();
      const isWorkflowEnabled = actor.inspector.isWorkflowEnabled();
      const state = isStateEnabled ? actor.inspector.getStateJson() : void 0;
      const connections = actor.inspector.getConnectionsJson();
      const rpcs = actor.inspector.getRpcs();
      const queueSize = actor.inspector.getQueueSize();
      const workflowHistory = actor.inspector.getWorkflowHistory();
      const bigIntReplacer = (_key, value) => typeof value === "bigint" ? Number(value) : value;
      const safeWorkflowHistory = workflowHistory ? JSON.parse(JSON.stringify(workflowHistory, bigIntReplacer)) : null;
      return c.json({
        state,
        connections,
        rpcs,
        queueSize,
        isStateEnabled,
        isDatabaseEnabled,
        isWorkflowEnabled,
        workflowHistory: safeWorkflowHistory
      });
    });
  }
  router.post("/action/:action", async (c) => {
    const actionName = c.req.param("action");
    return handleAction(c, config4, actorDriver, actionName, c.env.actorId);
  });
  router.post("/queue", async (c) => {
    return handleQueueSend(c, config4, actorDriver, c.env.actorId);
  });
  router.post("/queue/:name", async (c) => {
    return handleQueueSend(
      c,
      config4,
      actorDriver,
      c.env.actorId,
      c.req.param("name")
    );
  });
  router.all("/request/*", async (c) => {
    const url = new URL(c.req.url);
    const originalPath = url.pathname.replace(/^\/request/, "") || "/";
    const correctedUrl = new URL(originalPath + url.search, url.origin);
    const correctedRequest = new Request(correctedUrl, {
      method: c.req.method,
      headers: c.req.raw.headers,
      body: c.req.raw.body,
      duplex: "half"
    });
    loggerWithoutContext().debug({
      msg: "rewriting http url",
      from: c.req.url,
      to: correctedRequest.url
    });
    return await handleRawRequest(
      c,
      correctedRequest,
      actorDriver,
      c.env.actorId
    );
  });
  router.notFound(handleRouteNotFound);
  router.onError(handleRouteError);
  return router;
}

// src/common/inline-websocket-adapter.ts
import { WSContext } from "hono/ws";
import { VirtualWebSocket } from "@rivetkit/virtual-websocket";
function logger() {
  return getLogger("inline-websocket-adapter");
}
var InlineWebSocketAdapter = class {
  #handler;
  #wsContext;
  #readyState = 0;
  #clientWs;
  #actorWs;
  constructor(handler) {
    this.#handler = handler;
    this.#clientWs = new VirtualWebSocket({
      getReadyState: () => this.#readyState,
      onSend: (data) => {
        try {
          this.#handler.onMessage({ data }, this.#wsContext);
          this.#actorWs.triggerMessage(data);
        } catch (err) {
          this.#handleError(err);
          this.#close(1011, "Internal error processing message");
        }
      },
      onClose: (code, reason) => this.#close(code, reason)
    });
    this.#actorWs = new VirtualWebSocket({
      getReadyState: () => this.#readyState,
      onSend: (data) => this.#clientWs.triggerMessage(data),
      onClose: (code, reason) => this.#close(code, reason)
    });
    this.#wsContext = new WSContext({
      raw: this.#actorWs,
      send: (data) => {
        logger().debug({ msg: "WSContext.send called" });
        this.#clientWs.triggerMessage(data);
      },
      close: (code, reason) => {
        logger().debug({ msg: "WSContext.close called", code, reason });
        this.#close(code || 1e3, reason || "");
      },
      readyState: 1
    });
    setTimeout(() => {
      this.#initialize();
    }, 0);
  }
  /** Get the client-side WebSocket (for proxy/client code) */
  get clientWebSocket() {
    return this.#clientWs;
  }
  /** Get the actor-side WebSocket (passed to actor via wsContext.raw) */
  get actorWebSocket() {
    return this.#actorWs;
  }
  async #initialize() {
    try {
      logger().debug({ msg: "websocket initializing" });
      this.#readyState = 1;
      logger().debug({ msg: "calling handler.onOpen with WSContext" });
      this.#handler.onOpen(void 0, this.#wsContext);
      this.#clientWs.triggerOpen();
      this.#actorWs.triggerOpen();
    } catch (err) {
      this.#handleError(err);
      this.#close(1011, "Internal error during initialization");
    }
  }
  #handleError(err) {
    logger().error({
      msg: "error in websocket",
      error: err,
      errorMessage: err instanceof Error ? err.message : String(err),
      stack: err instanceof Error ? err.stack : void 0
    });
    try {
      this.#handler.onError(err, this.#wsContext);
    } catch (handlerErr) {
      logger().error({ msg: "error in onError handler", error: handlerErr });
    }
    this.#clientWs.triggerError(err);
    this.#actorWs.triggerError(err);
  }
  #close(code, reason) {
    if (this.#readyState === 3 || this.#readyState === 2) {
      return;
    }
    logger().debug({ msg: "closing websocket", code, reason });
    this.#readyState = 2;
    try {
      this.#handler.onClose({ code, reason, wasClean: true }, this.#wsContext);
    } catch (err) {
      logger().error({ msg: "error closing websocket", error: err });
    } finally {
      this.#readyState = 3;
      this.#clientWs.triggerClose(code, reason);
      this.#actorWs.triggerClose(code, reason);
    }
  }
};
function createInlineWebSocket(handler) {
  const adapter = new InlineWebSocketAdapter(handler);
  return adapter.clientWebSocket;
}

// src/drivers/file-system/mod.ts
import { z } from "zod";

// src/drivers/file-system/actor.ts
var FileSystemActorDriver = class {
  #config;
  #managerDriver;
  #inlineClient;
  #state;
  constructor(config4, managerDriver, inlineClient, state) {
    this.#config = config4;
    this.#managerDriver = managerDriver;
    this.#inlineClient = inlineClient;
    this.#state = state;
  }
  async loadActor(actorId) {
    return this.#state.startActor(
      this.#config,
      this.#inlineClient,
      this,
      actorId
    );
  }
  /**
   * Get the current storage directory path
   */
  get storagePath() {
    return this.#state.storagePath;
  }
  getContext(_actorId) {
    return {};
  }
  async kvBatchPut(actorId, entries) {
    await this.#state.kvBatchPut(actorId, entries);
  }
  async kvBatchGet(actorId, keys) {
    return await this.#state.kvBatchGet(actorId, keys);
  }
  async kvBatchDelete(actorId, keys) {
    await this.#state.kvBatchDelete(actorId, keys);
  }
  async kvListPrefix(actorId, prefix) {
    return await this.#state.kvListPrefix(actorId, prefix);
  }
  async setAlarm(actor, timestamp) {
    await this.#state.setActorAlarm(actor.id, timestamp);
  }
  /** Creates a SQLite VFS instance for creating KV-backed databases */
  async createSqliteVfs() {
    return await importSqliteVfs();
  }
  startSleep(actorId) {
    this.#state.sleepActor(actorId);
  }
  async startDestroy(actorId) {
    await this.#state.destroyActor(actorId);
  }
  async onBeforeActorStart(actor) {
    await actor.cleanupPersistedConnections("file-system-driver.start");
  }
};

// src/drivers/file-system/global-state.ts
import invariant3 from "invariant";

// src/schemas/file-system-driver/versioned.ts
import { createVersionedDataHandler } from "vbare";

// dist/schemas/file-system-driver/v1.ts
import * as bare from "@rivetkit/bare-ts";
var config = /* @__PURE__ */ bare.Config({});
function read0(bc) {
  const len = bare.readUintSafe(bc);
  if (len === 0) {
    return [];
  }
  const result = [bare.readString(bc)];
  for (let i = 1; i < len; i++) {
    result[i] = bare.readString(bc);
  }
  return result;
}
function write0(bc, x) {
  bare.writeUintSafe(bc, x.length);
  for (let i = 0; i < x.length; i++) {
    bare.writeString(bc, x[i]);
  }
}
function readActorState(bc) {
  return {
    actorId: bare.readString(bc),
    name: bare.readString(bc),
    key: read0(bc),
    persistedData: bare.readData(bc),
    createdAt: bare.readU64(bc)
  };
}
function writeActorState(bc, x) {
  bare.writeString(bc, x.actorId);
  bare.writeString(bc, x.name);
  write0(bc, x.key);
  bare.writeData(bc, x.persistedData);
  bare.writeU64(bc, x.createdAt);
}
function encodeActorState(x) {
  const bc = new bare.ByteCursor(
    new Uint8Array(config.initialBufferLength),
    config
  );
  writeActorState(bc, x);
  return new Uint8Array(bc.view.buffer, bc.view.byteOffset, bc.offset);
}
function decodeActorState(bytes) {
  const bc = new bare.ByteCursor(bytes, config);
  const result = readActorState(bc);
  if (bc.offset < bc.view.byteLength) {
    throw new bare.BareError(bc.offset, "remaining bytes");
  }
  return result;
}
function readActorAlarm(bc) {
  return {
    actorId: bare.readString(bc),
    timestamp: bare.readUint(bc)
  };
}
function writeActorAlarm(bc, x) {
  bare.writeString(bc, x.actorId);
  bare.writeUint(bc, x.timestamp);
}
function encodeActorAlarm(x) {
  const bc = new bare.ByteCursor(
    new Uint8Array(config.initialBufferLength),
    config
  );
  writeActorAlarm(bc, x);
  return new Uint8Array(bc.view.buffer, bc.view.byteOffset, bc.offset);
}
function decodeActorAlarm(bytes) {
  const bc = new bare.ByteCursor(bytes, config);
  const result = readActorAlarm(bc);
  if (bc.offset < bc.view.byteLength) {
    throw new bare.BareError(bc.offset, "remaining bytes");
  }
  return result;
}

// dist/schemas/file-system-driver/v2.ts
import * as bare2 from "@rivetkit/bare-ts";
var config2 = /* @__PURE__ */ bare2.Config({});
function readActorKvEntry(bc) {
  return {
    key: bare2.readData(bc),
    value: bare2.readData(bc)
  };
}
function writeActorKvEntry(bc, x) {
  bare2.writeData(bc, x.key);
  bare2.writeData(bc, x.value);
}
function read02(bc) {
  const len = bare2.readUintSafe(bc);
  if (len === 0) {
    return [];
  }
  const result = [bare2.readString(bc)];
  for (let i = 1; i < len; i++) {
    result[i] = bare2.readString(bc);
  }
  return result;
}
function write02(bc, x) {
  bare2.writeUintSafe(bc, x.length);
  for (let i = 0; i < x.length; i++) {
    bare2.writeString(bc, x[i]);
  }
}
function read1(bc) {
  const len = bare2.readUintSafe(bc);
  if (len === 0) {
    return [];
  }
  const result = [readActorKvEntry(bc)];
  for (let i = 1; i < len; i++) {
    result[i] = readActorKvEntry(bc);
  }
  return result;
}
function write1(bc, x) {
  bare2.writeUintSafe(bc, x.length);
  for (let i = 0; i < x.length; i++) {
    writeActorKvEntry(bc, x[i]);
  }
}
function readActorState2(bc) {
  return {
    actorId: bare2.readString(bc),
    name: bare2.readString(bc),
    key: read02(bc),
    kvStorage: read1(bc),
    createdAt: bare2.readU64(bc)
  };
}
function writeActorState2(bc, x) {
  bare2.writeString(bc, x.actorId);
  bare2.writeString(bc, x.name);
  write02(bc, x.key);
  write1(bc, x.kvStorage);
  bare2.writeU64(bc, x.createdAt);
}
function encodeActorState2(x) {
  const bc = new bare2.ByteCursor(
    new Uint8Array(config2.initialBufferLength),
    config2
  );
  writeActorState2(bc, x);
  return new Uint8Array(bc.view.buffer, bc.view.byteOffset, bc.offset);
}
function decodeActorState2(bytes) {
  const bc = new bare2.ByteCursor(bytes, config2);
  const result = readActorState2(bc);
  if (bc.offset < bc.view.byteLength) {
    throw new bare2.BareError(bc.offset, "remaining bytes");
  }
  return result;
}
function readActorAlarm2(bc) {
  return {
    actorId: bare2.readString(bc),
    timestamp: bare2.readUint(bc)
  };
}
function writeActorAlarm2(bc, x) {
  bare2.writeString(bc, x.actorId);
  bare2.writeUint(bc, x.timestamp);
}
function encodeActorAlarm2(x) {
  const bc = new bare2.ByteCursor(
    new Uint8Array(config2.initialBufferLength),
    config2
  );
  writeActorAlarm2(bc, x);
  return new Uint8Array(bc.view.buffer, bc.view.byteOffset, bc.offset);
}
function decodeActorAlarm2(bytes) {
  const bc = new bare2.ByteCursor(bytes, config2);
  const result = readActorAlarm2(bc);
  if (bc.offset < bc.view.byteLength) {
    throw new bare2.BareError(bc.offset, "remaining bytes");
  }
  return result;
}

// dist/schemas/file-system-driver/v3.ts
import * as bare3 from "@rivetkit/bare-ts";
var config3 = /* @__PURE__ */ bare3.Config({});
function readActorKvEntry2(bc) {
  return {
    key: bare3.readData(bc),
    value: bare3.readData(bc)
  };
}
function writeActorKvEntry2(bc, x) {
  bare3.writeData(bc, x.key);
  bare3.writeData(bc, x.value);
}
function read03(bc) {
  const len = bare3.readUintSafe(bc);
  if (len === 0) {
    return [];
  }
  const result = [bare3.readString(bc)];
  for (let i = 1; i < len; i++) {
    result[i] = bare3.readString(bc);
  }
  return result;
}
function write03(bc, x) {
  bare3.writeUintSafe(bc, x.length);
  for (let i = 0; i < x.length; i++) {
    bare3.writeString(bc, x[i]);
  }
}
function read12(bc) {
  const len = bare3.readUintSafe(bc);
  if (len === 0) {
    return [];
  }
  const result = [readActorKvEntry2(bc)];
  for (let i = 1; i < len; i++) {
    result[i] = readActorKvEntry2(bc);
  }
  return result;
}
function write12(bc, x) {
  bare3.writeUintSafe(bc, x.length);
  for (let i = 0; i < x.length; i++) {
    writeActorKvEntry2(bc, x[i]);
  }
}
function read2(bc) {
  return bare3.readBool(bc) ? bare3.readU64(bc) : null;
}
function write2(bc, x) {
  bare3.writeBool(bc, x !== null);
  if (x !== null) {
    bare3.writeU64(bc, x);
  }
}
function readActorState3(bc) {
  return {
    actorId: bare3.readString(bc),
    name: bare3.readString(bc),
    key: read03(bc),
    kvStorage: read12(bc),
    createdAt: bare3.readU64(bc),
    startTs: read2(bc),
    connectableTs: read2(bc),
    sleepTs: read2(bc),
    destroyTs: read2(bc)
  };
}
function writeActorState3(bc, x) {
  bare3.writeString(bc, x.actorId);
  bare3.writeString(bc, x.name);
  write03(bc, x.key);
  write12(bc, x.kvStorage);
  bare3.writeU64(bc, x.createdAt);
  write2(bc, x.startTs);
  write2(bc, x.connectableTs);
  write2(bc, x.sleepTs);
  write2(bc, x.destroyTs);
}
function encodeActorState3(x) {
  const bc = new bare3.ByteCursor(
    new Uint8Array(config3.initialBufferLength),
    config3
  );
  writeActorState3(bc, x);
  return new Uint8Array(bc.view.buffer, bc.view.byteOffset, bc.offset);
}
function decodeActorState3(bytes) {
  const bc = new bare3.ByteCursor(bytes, config3);
  const result = readActorState3(bc);
  if (bc.offset < bc.view.byteLength) {
    throw new bare3.BareError(bc.offset, "remaining bytes");
  }
  return result;
}
function readActorAlarm3(bc) {
  return {
    actorId: bare3.readString(bc),
    timestamp: bare3.readUint(bc)
  };
}
function writeActorAlarm3(bc, x) {
  bare3.writeString(bc, x.actorId);
  bare3.writeUint(bc, x.timestamp);
}
function encodeActorAlarm3(x) {
  const bc = new bare3.ByteCursor(
    new Uint8Array(config3.initialBufferLength),
    config3
  );
  writeActorAlarm3(bc, x);
  return new Uint8Array(bc.view.buffer, bc.view.byteOffset, bc.offset);
}
function decodeActorAlarm3(bytes) {
  const bc = new bare3.ByteCursor(bytes, config3);
  const result = readActorAlarm3(bc);
  if (bc.offset < bc.view.byteLength) {
    throw new bare3.BareError(bc.offset, "remaining bytes");
  }
  return result;
}

// src/schemas/file-system-driver/versioned.ts
var CURRENT_VERSION3 = 3;
var v1ToV2 = (v1State) => {
  const kvStorage = [];
  if (v1State.persistedData) {
    const key = new Uint8Array([1]);
    kvStorage.push({
      key: bufferToArrayBuffer(key),
      value: v1State.persistedData
    });
  }
  return {
    actorId: v1State.actorId,
    name: v1State.name,
    key: v1State.key,
    kvStorage,
    createdAt: v1State.createdAt
  };
};
var v2ToV3 = (v2State) => {
  return {
    actorId: v2State.actorId,
    name: v2State.name,
    key: v2State.key,
    kvStorage: v2State.kvStorage,
    createdAt: v2State.createdAt,
    startTs: null,
    connectableTs: null,
    sleepTs: null,
    destroyTs: null
  };
};
var v3ToV2 = (v3State) => {
  return {
    actorId: v3State.actorId,
    name: v3State.name,
    key: v3State.key,
    kvStorage: v3State.kvStorage,
    createdAt: v3State.createdAt
  };
};
var v2ToV1 = (v2State) => {
  const persistDataEntry = v2State.kvStorage.find((entry) => {
    const key = new Uint8Array(entry.key);
    return key.length === 1 && key[0] === 1;
  });
  return {
    actorId: v2State.actorId,
    name: v2State.name,
    key: v2State.key,
    persistedData: (persistDataEntry == null ? void 0 : persistDataEntry.value) || new ArrayBuffer(0),
    createdAt: v2State.createdAt
  };
};
var ACTOR_STATE_VERSIONED = createVersionedDataHandler({
  deserializeVersion: (bytes, version) => {
    switch (version) {
      case 1:
        return decodeActorState(bytes);
      case 2:
        return decodeActorState2(bytes);
      case 3:
        return decodeActorState3(bytes);
      default:
        throw new Error(`Unknown version ${version}`);
    }
  },
  serializeVersion: (data, version) => {
    switch (version) {
      case 1:
        return encodeActorState(data);
      case 2:
        return encodeActorState2(data);
      case 3:
        return encodeActorState3(data);
      default:
        throw new Error(`Unknown version ${version}`);
    }
  },
  deserializeConverters: () => [v1ToV2, v2ToV3],
  serializeConverters: () => [v3ToV2, v2ToV1]
});
var ACTOR_ALARM_VERSIONED = createVersionedDataHandler({
  deserializeVersion: (bytes, version) => {
    switch (version) {
      case 1:
        return decodeActorAlarm(bytes);
      case 2:
        return decodeActorAlarm2(bytes);
      case 3:
        return decodeActorAlarm3(bytes);
      default:
        throw new Error(`Unknown version ${version}`);
    }
  },
  serializeVersion: (data, version) => {
    switch (version) {
      case 1:
        return encodeActorAlarm(data);
      case 2:
        return encodeActorAlarm2(data);
      case 3:
        return encodeActorAlarm3(data);
      default:
        throw new Error(`Unknown version ${version}`);
    }
  },
  deserializeConverters: () => [],
  serializeConverters: () => []
});

// src/drivers/file-system/log.ts
function logger2() {
  return getLogger("driver-fs");
}

// src/drivers/file-system/utils.ts
function generateActorId(name, key) {
  const jsonString = JSON.stringify([name, key]);
  const crypto2 = getNodeCrypto();
  const hash = crypto2.createHash("sha256").update(jsonString).digest("hex").substring(0, 16);
  return hash;
}
function createHashForPath(dirPath) {
  const path = getNodePath();
  const normalizedPath = path.normalize(dirPath);
  const lastComponent = path.basename(normalizedPath);
  const crypto2 = getNodeCrypto();
  const hash = crypto2.createHash("sha256").update(normalizedPath).digest("hex").substring(0, 8);
  return `${lastComponent}-${hash}`;
}
function getStoragePath() {
  const dataPath = getDataPath("rivetkit");
  const dirHash = createHashForPath(process.cwd());
  const path = getNodePath();
  return path.join(dataPath, dirHash);
}
async function pathExists(path) {
  try {
    const fs = getNodeFs();
    await fs.access(path);
    return true;
  } catch {
    return false;
  }
}
async function ensureDirectoryExists(directoryPath) {
  if (!await pathExists(directoryPath)) {
    const fs = getNodeFs();
    await fs.mkdir(directoryPath, { recursive: true });
  }
}
function ensureDirectoryExistsSync(directoryPath) {
  const fsSync = getNodeFsSync();
  if (!fsSync.existsSync(directoryPath)) {
    fsSync.mkdirSync(directoryPath, { recursive: true });
  }
}
function getDataPath(appName) {
  const platform = process.platform;
  const os = getNodeOs();
  const homeDir = os.homedir();
  const path = getNodePath();
  switch (platform) {
    case "win32":
      return path.join(
        process.env.APPDATA || path.join(homeDir, "AppData", "Roaming"),
        appName
      );
    case "darwin":
      return path.join(
        homeDir,
        "Library",
        "Application Support",
        appName
      );
    default:
      return path.join(
        process.env.XDG_DATA_HOME || path.join(homeDir, ".local", "share"),
        appName
      );
  }
}

// src/drivers/file-system/sqlite-runtime.ts
function normalizeParams(params) {
  if (!params || params.length === 0) {
    return [];
  }
  return params.map((value) => {
    if (value instanceof Uint8Array) {
      return Buffer.from(value);
    }
    return value;
  });
}
function createPreparedDatabaseAdapter(rawDb, prepare) {
  return {
    exec: (sql) => {
      rawDb.exec(sql);
    },
    run: (sql, params) => {
      const stmt = prepare(sql);
      stmt.run(...normalizeParams(params));
    },
    get: (sql, params) => {
      const stmt = prepare(sql);
      return stmt.get(...normalizeParams(params));
    },
    all: (sql, params) => {
      const stmt = prepare(sql);
      return stmt.all(...normalizeParams(params));
    },
    close: () => {
      rawDb.close();
    }
  };
}
function configureSqliteRuntimeDatabase(rawDb, path) {
  rawDb.exec("PRAGMA busy_timeout = 5000");
  if (path !== ":memory:") {
    rawDb.exec("PRAGMA journal_mode = WAL");
  }
}
function loadSqliteRuntime() {
  const requireFn = getRequireFn();
  const loadErrors = [];
  try {
    const bunSqlite = requireFn(
      /* webpackIgnore: true */
      "bun:sqlite"
    );
    const BunDatabase = bunSqlite.Database;
    if (BunDatabase) {
      return {
        kind: "bun",
        open: (path) => {
          var _a;
          const rawDb = new BunDatabase(path);
          configureSqliteRuntimeDatabase(rawDb, path);
          const query = (_a = rawDb.query) == null ? void 0 : _a.bind(rawDb);
          if (!query) throw new Error("bun:sqlite database missing query method");
          return createPreparedDatabaseAdapter(rawDb, query);
        }
      };
    }
  } catch (error) {
    loadErrors.push(`bun:sqlite unavailable: ${String(error)}`);
  }
  try {
    const nodeSqlite = requireFn(
      /* webpackIgnore: true */
      "node:sqlite"
    );
    const NodeDatabaseSync = nodeSqlite.DatabaseSync;
    if (NodeDatabaseSync) {
      return {
        kind: "node",
        open: (path) => {
          var _a;
          const rawDb = new NodeDatabaseSync(path);
          configureSqliteRuntimeDatabase(rawDb, path);
          const prepare = (_a = rawDb.prepare) == null ? void 0 : _a.bind(rawDb);
          if (!prepare) {
            throw new Error("node:sqlite DatabaseSync missing prepare method");
          }
          return createPreparedDatabaseAdapter(rawDb, prepare);
        }
      };
    }
  } catch (error) {
    loadErrors.push(`node:sqlite unavailable: ${String(error)}`);
  }
  try {
    const betterSqlite3Module = requireFn(
      /* webpackIgnore: true */
      "better-sqlite3"
    );
    const BetterSqlite3 = typeof betterSqlite3Module === "function" ? betterSqlite3Module : betterSqlite3Module.default;
    if (BetterSqlite3) {
      return {
        kind: "better-sqlite3",
        open: (path) => {
          var _a;
          const rawDb = new BetterSqlite3(path);
          configureSqliteRuntimeDatabase(rawDb, path);
          const prepare = (_a = rawDb.prepare) == null ? void 0 : _a.bind(rawDb);
          if (!prepare) {
            throw new Error("better-sqlite3 database missing prepare method");
          }
          return createPreparedDatabaseAdapter(rawDb, prepare);
        }
      };
    }
  } catch (error) {
    loadErrors.push(`better-sqlite3 unavailable: ${String(error)}`);
    throw new Error(
      `No SQLite runtime available. Tried bun:sqlite, node:sqlite, and better-sqlite3. Install better-sqlite3 (e.g. "pnpm add better-sqlite3") if native runtimes are unavailable.
${loadErrors.join("\n")}`
    );
  }
  throw new Error(
    `No SQLite runtime available. Tried bun:sqlite, node:sqlite, and better-sqlite3.
${loadErrors.join("\n")}`
  );
}
function computePrefixUpperBound(prefix) {
  if (prefix.length === 0) {
    return void 0;
  }
  const upperBound = new Uint8Array(prefix);
  for (let i = upperBound.length - 1; i >= 0; i--) {
    if (upperBound[i] !== 255) {
      upperBound[i] += 1;
      return upperBound.slice(0, i + 1);
    }
  }
  return void 0;
}
function ensureUint8Array(value, fieldName) {
  if (value instanceof Uint8Array) {
    return value;
  }
  if (value instanceof ArrayBuffer) {
    return new Uint8Array(value);
  }
  if (ArrayBuffer.isView(value)) {
    return new Uint8Array(value.buffer, value.byteOffset, value.byteLength);
  }
  throw new Error(`SQLite row field "${fieldName}" is not binary data`);
}

// src/drivers/file-system/kv-limits.ts
var KV_MAX_KEY_SIZE = 2 * 1024;
var KV_MAX_VALUE_SIZE = 128 * 1024;
var KV_MAX_KEYS = 128;
var KV_MAX_PUT_PAYLOAD_SIZE = 976 * 1024;
var KV_MAX_STORAGE_SIZE = 10 * 1024 * 1024 * 1024;
var KV_KEY_WRAPPER_OVERHEAD_SIZE = 2;
function estimateKvSize(db) {
  const row = db.get(
    "SELECT COALESCE(SUM(LENGTH(key) + LENGTH(value)), 0) AS total FROM kv"
  );
  return row ? Number(row.total ?? 0) : 0;
}
function validateKvKey(key, keyLabel = "key") {
  if (key.byteLength + KV_KEY_WRAPPER_OVERHEAD_SIZE > KV_MAX_KEY_SIZE) {
    throw new Error(`${keyLabel} is too long (max 2048 bytes)`);
  }
}
function validateKvKeys(keys) {
  if (keys.length > KV_MAX_KEYS) {
    throw new Error("a maximum of 128 keys is allowed");
  }
  for (const key of keys) {
    validateKvKey(key);
  }
}
function validateKvEntries(entries, totalSize) {
  if (entries.length > KV_MAX_KEYS) {
    throw new Error("A maximum of 128 key-value entries is allowed");
  }
  let payloadSize = 0;
  for (const [key, value] of entries) {
    payloadSize += key.byteLength + KV_KEY_WRAPPER_OVERHEAD_SIZE + value.byteLength;
  }
  if (payloadSize > KV_MAX_PUT_PAYLOAD_SIZE) {
    throw new Error("total payload is too large (max 976 KiB)");
  }
  const storageRemaining = Math.max(0, KV_MAX_STORAGE_SIZE - totalSize);
  if (payloadSize > storageRemaining) {
    throw new Error(
      `not enough space left in storage (${storageRemaining} bytes remaining, current payload is ${payloadSize} bytes)`
    );
  }
  for (const [key, value] of entries) {
    validateKvKey(key);
    if (value.byteLength > KV_MAX_VALUE_SIZE) {
      throw new Error(
        `value is too large (max ${KV_MAX_VALUE_SIZE / 1024} KiB)`
      );
    }
  }
}

// src/drivers/file-system/global-state.ts
var FileSystemGlobalState = class {
  #storagePath;
  #stateDir;
  #dbsDir;
  #alarmsDir;
  #persist;
  #sqliteRuntime;
  #actorKvDatabases = /* @__PURE__ */ new Map();
  // IMPORTANT: Never delete from this map. Doing so will result in race
  // conditions since the actor generation will cease to be tracked
  // correctly. Always increment generation if a new actor is created.
  #actors = /* @__PURE__ */ new Map();
  #actorCountOnStartup = 0;
  #runnerParams;
  get persist() {
    return this.#persist;
  }
  get storagePath() {
    return this.#storagePath;
  }
  get actorCountOnStartup() {
    return this.#actorCountOnStartup;
  }
  constructor(options = {}) {
    const { persist = true, customPath, useNativeSqlite = true } = options;
    if (!useNativeSqlite) {
      throw new Error(
        "File-system driver no longer supports non-SQLite KV storage."
      );
    }
    this.#persist = persist;
    this.#sqliteRuntime = loadSqliteRuntime();
    this.#storagePath = persist ? customPath ?? getStoragePath() : "/tmp";
    const path = getNodePath();
    this.#stateDir = path.join(this.#storagePath, "state");
    this.#dbsDir = path.join(this.#storagePath, "databases");
    this.#alarmsDir = path.join(this.#storagePath, "alarms");
    if (this.#persist) {
      ensureDirectoryExistsSync(this.#stateDir);
      ensureDirectoryExistsSync(this.#dbsDir);
      ensureDirectoryExistsSync(this.#alarmsDir);
      try {
        const fsSync = getNodeFsSync();
        const actorIds = fsSync.readdirSync(this.#stateDir);
        this.#actorCountOnStartup = actorIds.length;
      } catch (error) {
        logger2().error({ msg: "failed to count actors", error });
      }
      logger2().debug({
        msg: "file system driver ready",
        dir: this.#storagePath,
        actorCount: this.#actorCountOnStartup,
        sqliteRuntime: this.#sqliteRuntime.kind
      });
      try {
        this.#cleanupTempFilesSync();
      } catch (err) {
        logger2().error({
          msg: "failed to cleanup temp files",
          error: err
        });
      }
      try {
        this.#migrateLegacyKvToSqliteOnStartupSync();
      } catch (error) {
        logger2().error({
          msg: "failed legacy kv startup migration",
          error
        });
        throw error;
      }
    } else {
      logger2().debug({
        msg: "memory driver ready",
        sqliteRuntime: this.#sqliteRuntime.kind
      });
    }
  }
  getActorStatePath(actorId) {
    return getNodePath().join(this.#stateDir, actorId);
  }
  getActorDbPath(actorId) {
    return getNodePath().join(this.#dbsDir, `${actorId}.db`);
  }
  getActorAlarmPath(actorId) {
    return getNodePath().join(this.#alarmsDir, actorId);
  }
  #getActorKvDatabasePath(actorId) {
    if (this.#persist) {
      return this.getActorDbPath(actorId);
    }
    return ":memory:";
  }
  #ensureActorKvTables(db) {
    db.exec(`
			CREATE TABLE IF NOT EXISTS kv (
				key BLOB PRIMARY KEY NOT NULL,
				value BLOB NOT NULL
			)
		`);
  }
  #getOrCreateActorKvDatabase(actorId) {
    const existing = this.#actorKvDatabases.get(actorId);
    if (existing) {
      return existing;
    }
    const dbPath = this.#getActorKvDatabasePath(actorId);
    if (this.#persist) {
      const path = getNodePath();
      ensureDirectoryExistsSync(path.dirname(dbPath));
    }
    let db;
    try {
      db = this.#sqliteRuntime.open(dbPath);
    } catch (error) {
      throw new Error(
        `failed to open actor kv database for actor ${actorId} at ${dbPath}: ${error}`
      );
    }
    this.#ensureActorKvTables(db);
    this.#actorKvDatabases.set(actorId, db);
    return db;
  }
  #closeActorKvDatabase(actorId) {
    const db = this.#actorKvDatabases.get(actorId);
    if (!db) {
      return;
    }
    try {
      db.close();
    } finally {
      this.#actorKvDatabases.delete(actorId);
    }
  }
  #putKvEntriesInDb(db, entries) {
    if (entries.length === 0) {
      return;
    }
    db.exec("BEGIN");
    try {
      for (const [key, value] of entries) {
        db.run("INSERT OR REPLACE INTO kv (key, value) VALUES (?, ?)", [
          key,
          value
        ]);
      }
      db.exec("COMMIT");
    } catch (error) {
      try {
        db.exec("ROLLBACK");
      } catch {
      }
      throw error;
    }
  }
  #isKvDbPopulated(db) {
    const row = db.get(
      "SELECT COUNT(*) AS count FROM kv"
    );
    const count = row ? Number(row.count) : 0;
    return count > 0;
  }
  #migrateLegacyKvToSqliteOnStartupSync() {
    const fsSync = getNodeFsSync();
    if (!fsSync.existsSync(this.#stateDir)) {
      return;
    }
    const actorIds = fsSync.readdirSync(this.#stateDir).filter((id) => !id.includes(".tmp."));
    for (const actorId of actorIds) {
      const statePath = this.getActorStatePath(actorId);
      let state;
      try {
        const stateBytes = fsSync.readFileSync(statePath);
        state = ACTOR_STATE_VERSIONED.deserializeWithEmbeddedVersion(
          new Uint8Array(stateBytes)
        );
      } catch (error) {
        logger2().warn({
          msg: "failed to parse actor state during startup migration",
          actorId,
          error
        });
        continue;
      }
      if (!state.kvStorage || state.kvStorage.length === 0) {
        continue;
      }
      const dbPath = this.getActorDbPath(actorId);
      const path = getNodePath();
      ensureDirectoryExistsSync(path.dirname(dbPath));
      const db = this.#sqliteRuntime.open(dbPath);
      try {
        this.#ensureActorKvTables(db);
        if (this.#isKvDbPopulated(db)) {
          continue;
        }
        const legacyEntries = state.kvStorage.map((entry) => [
          new Uint8Array(entry.key),
          new Uint8Array(entry.value)
        ]);
        this.#putKvEntriesInDb(db, legacyEntries);
        logger2().info({
          msg: "migrated legacy actor kv storage to sqlite",
          actorId,
          entryCount: legacyEntries.length
        });
      } finally {
        db.close();
      }
    }
  }
  async *getActorsIterator(params) {
    let actorIds = Array.from(this.#actors.keys()).sort();
    const fsSync = getNodeFsSync();
    if (fsSync.existsSync(this.#stateDir)) {
      actorIds = fsSync.readdirSync(this.#stateDir).filter((id) => !id.includes(".tmp")).sort();
    }
    const startIndex = params.cursor ? actorIds.indexOf(params.cursor) + 1 : 0;
    for (let i = startIndex; i < actorIds.length; i++) {
      const actorId = actorIds[i];
      if (!actorId) {
        continue;
      }
      try {
        const state = await this.loadActorStateOrError(actorId);
        yield state;
      } catch (error) {
        logger2().error({
          msg: "failed to load actor state",
          actorId,
          error
        });
      }
    }
  }
  /**
   * Ensures an entry exists for this actor.
   *
   * Used for #createActor and #loadActor.
   */
  #upsertEntry(actorId) {
    let entry = this.#actors.get(actorId);
    if (entry) {
      return entry;
    }
    entry = {
      id: actorId,
      lifecycleState: 0 /* NONEXISTENT */,
      generation: crypto.randomUUID()
    };
    this.#actors.set(actorId, entry);
    return entry;
  }
  /**
   * Creates a new actor and writes to file system.
   */
  async createActor(actorId, name, key, input) {
    await this.#waitForActorStop(actorId);
    let entry = this.#upsertEntry(actorId);
    if (entry.state) {
      throw new ActorDuplicateKey(name, key);
    }
    if (this.isActorStopping(actorId)) {
      await this.#waitForActorStop(actorId);
      entry = this.#upsertEntry(actorId);
    }
    if (entry.lifecycleState === 4 /* DESTROYED */) {
      entry.lifecycleState = 0 /* NONEXISTENT */;
      entry.generation = crypto.randomUUID();
    }
    const initialKvState = getInitialActorKvState(input);
    await this.#withActorWrite(actorId, async (lockedEntry) => {
      lockedEntry.state = {
        actorId,
        name,
        key,
        createdAt: BigInt(Date.now()),
        kvStorage: [],
        startTs: null,
        connectableTs: null,
        sleepTs: null,
        destroyTs: null
      };
      lockedEntry.lifecycleState = 1 /* AWAKE */;
      if (this.#persist) {
        await this.#performWrite(
          actorId,
          lockedEntry.generation,
          lockedEntry.state
        );
      }
      if (initialKvState.length > 0) {
        const db = this.#getOrCreateActorKvDatabase(actorId);
        this.#putKvEntriesInDb(db, initialKvState);
      }
    });
    return entry;
  }
  /**
   * Loads the actor from disk or returns the existing actor entry. This will return an entry even if the actor does not actually exist.
   */
  async loadActor(actorId) {
    const entry = this.#upsertEntry(actorId);
    if (entry.lifecycleState === 4 /* DESTROYED */) {
      return entry;
    }
    if (entry.state) {
      return entry;
    }
    if (!this.#persist) {
      return entry;
    }
    if (entry.loadPromise) {
      await entry.loadPromise;
      return entry;
    }
    entry.loadPromise = this.loadActorState(entry);
    return entry.loadPromise;
  }
  async loadActorState(entry) {
    const stateFilePath = this.getActorStatePath(entry.id);
    try {
      const fs = getNodeFs();
      const stateData = await fs.readFile(stateFilePath);
      const loadedState = ACTOR_STATE_VERSIONED.deserializeWithEmbeddedVersion(
        new Uint8Array(stateData)
      );
      entry.state = {
        ...loadedState,
        kvStorage: []
      };
      return entry;
    } catch (innerError) {
      if (innerError.code === "ENOENT") {
        entry.loadPromise = void 0;
        return entry;
      }
      const error = new Error(
        `Failed to load actor state: ${innerError}`
      );
      throw error;
    }
  }
  async loadOrCreateActor(actorId, name, key, input) {
    await this.#waitForActorStop(actorId);
    const entry = await this.loadActor(actorId);
    if (!entry.state) {
      if (this.isActorStopping(actorId)) {
        await this.#waitForActorStop(actorId);
        return await this.loadOrCreateActor(actorId, name, key, input);
      }
      if (entry.lifecycleState === 4 /* DESTROYED */) {
        entry.lifecycleState = 0 /* NONEXISTENT */;
        entry.generation = crypto.randomUUID();
      }
      const initialKvState = getInitialActorKvState(input);
      await this.#withActorWrite(actorId, async (lockedEntry) => {
        lockedEntry.state = {
          actorId,
          name,
          key,
          createdAt: BigInt(Date.now()),
          kvStorage: [],
          startTs: null,
          connectableTs: null,
          sleepTs: null,
          destroyTs: null
        };
        if (this.#persist) {
          await this.#performWrite(
            actorId,
            lockedEntry.generation,
            lockedEntry.state
          );
        }
        if (initialKvState.length > 0) {
          const db = this.#getOrCreateActorKvDatabase(actorId);
          this.#putKvEntriesInDb(db, initialKvState);
        }
      });
    }
    return entry;
  }
  async sleepActor(actorId) {
    var _a, _b;
    invariant3(
      this.#persist,
      "cannot sleep actor with memory driver, must use file system driver"
    );
    const actor = this.#upsertEntry(actorId);
    invariant3(actor, `tried to sleep ${actorId}, does not exist`);
    if (this.isActorStopping(actorId)) {
      return;
    }
    actor.lifecycleState = 2 /* STARTING_SLEEP */;
    actor.stopPromise = promiseWithResolvers((reason) => logger2().warn({ msg: "unhandled actor sleep stop promise rejection", reason }));
    if (actor.loadPromise) await actor.loadPromise.catch();
    if ((_a = actor.startPromise) == null ? void 0 : _a.promise)
      await actor.startPromise.promise.catch();
    try {
      if (actor.state) {
        await this.#withActorWrite(actorId, async (lockedEntry) => {
          if (!lockedEntry.state) {
            return;
          }
          lockedEntry.state = {
            ...lockedEntry.state,
            sleepTs: BigInt(Date.now())
          };
          if (this.#persist) {
            await this.#performWrite(
              actorId,
              lockedEntry.generation,
              lockedEntry.state
            );
          }
        });
      }
      invariant3(actor.actor, "actor should be loaded");
      await actor.actor.onStop("sleep");
    } finally {
      await this.#withActorWrite(actorId, async () => {
      });
      this.#closeActorKvDatabase(actorId);
      (_b = actor.stopPromise) == null ? void 0 : _b.resolve();
      actor.stopPromise = void 0;
      this.#actors.delete(actorId);
    }
  }
  async destroyActor(actorId) {
    var _a, _b;
    const actor = this.#upsertEntry(actorId);
    if (this.isActorStopping(actorId)) {
      return;
    }
    actor.lifecycleState = 3 /* STARTING_DESTROY */;
    actor.stopPromise = promiseWithResolvers((reason) => logger2().warn({ msg: "unhandled actor destroy stop promise rejection", reason }));
    if (actor.loadPromise) await actor.loadPromise.catch();
    if ((_a = actor.startPromise) == null ? void 0 : _a.promise)
      await actor.startPromise.promise.catch();
    try {
      if (actor.state) {
        await this.#withActorWrite(actorId, async (lockedEntry) => {
          if (!lockedEntry.state) {
            return;
          }
          lockedEntry.state = {
            ...lockedEntry.state,
            destroyTs: BigInt(Date.now())
          };
          if (this.#persist) {
            await this.#performWrite(
              actorId,
              lockedEntry.generation,
              lockedEntry.state
            );
          }
        });
      }
      if (actor.actor) {
        await actor.actor.onStop("destroy");
      }
      await this.#withActorWrite(actorId, async () => {
      });
      this.#closeActorKvDatabase(actorId);
      if (actor.alarmTimeout) {
        actor.alarmTimeout.abort();
      }
      if (this.#persist) {
        const fs = getNodeFs();
        await Promise.all([
          // Delete actor state file
          (async () => {
            try {
              await fs.unlink(this.getActorStatePath(actorId));
            } catch (err) {
              if ((err == null ? void 0 : err.code) !== "ENOENT") {
                logger2().error({
                  msg: "failed to delete actor state file",
                  actorId,
                  error: stringifyError(err)
                });
              }
            }
          })(),
          // Delete actor database file
          (async () => {
            try {
              await fs.unlink(this.getActorDbPath(actorId));
            } catch (err) {
              if ((err == null ? void 0 : err.code) !== "ENOENT") {
                logger2().error({
                  msg: "failed to delete actor database file",
                  actorId,
                  error: stringifyError(err)
                });
              }
            }
          })(),
          // Delete actor alarm file
          (async () => {
            try {
              await fs.unlink(this.getActorAlarmPath(actorId));
            } catch (err) {
              if ((err == null ? void 0 : err.code) !== "ENOENT") {
                logger2().error({
                  msg: "failed to delete actor alarm file",
                  actorId,
                  error: stringifyError(err)
                });
              }
            }
          })()
        ]);
      }
    } finally {
      await this.#withActorWrite(actorId, async () => {
      });
      (_b = actor.stopPromise) == null ? void 0 : _b.resolve();
      actor.stopPromise = void 0;
      actor.state = void 0;
      actor.loadPromise = void 0;
      actor.actor = void 0;
      actor.startPromise = void 0;
      actor.alarmTimeout = void 0;
      actor.alarmTimeout = void 0;
      actor.pendingWriteResolver = void 0;
      actor.lifecycleState = 4 /* DESTROYED */;
    }
  }
  /**
   * Save actor state to disk.
   */
  async writeActor(actorId, generation, state) {
    if (!this.#persist) {
      return;
    }
    await this.#withActorWrite(actorId, async () => {
      await this.#performWrite(actorId, generation, state);
    });
  }
  isGenerationCurrentAndNotDestroyed(actorId, generation) {
    const entry = this.#upsertEntry(actorId);
    if (!entry) return false;
    return entry.generation === generation && entry.lifecycleState !== 3 /* STARTING_DESTROY */;
  }
  isActorStopping(actorId) {
    const entry = this.#upsertEntry(actorId);
    if (!entry) return false;
    return entry.lifecycleState === 2 /* STARTING_SLEEP */ || entry.lifecycleState === 3 /* STARTING_DESTROY */;
  }
  async #waitForActorStop(actorId) {
    while (true) {
      const entry = this.#actors.get(actorId);
      if (!(entry == null ? void 0 : entry.stopPromise)) {
        return;
      }
      try {
        await entry.stopPromise.promise;
      } catch {
        return;
      }
    }
  }
  async #withActorWrite(actorId, fn) {
    const entry = this.#actors.get(actorId);
    invariant3(entry, "actor entry does not exist");
    const previousWrite = entry.pendingWriteResolver;
    const currentWrite = promiseWithResolvers((reason) => logger2().warn({ msg: "unhandled kv write promise rejection", reason }));
    entry.pendingWriteResolver = currentWrite;
    if (previousWrite) {
      try {
        await previousWrite.promise;
      } catch {
      }
    }
    try {
      return await fn(entry);
    } finally {
      currentWrite.resolve();
      if (entry.pendingWriteResolver === currentWrite) {
        entry.pendingWriteResolver = void 0;
      }
    }
  }
  async #waitForPendingWrite(actorId) {
    const entry = this.#actors.get(actorId);
    if (!(entry == null ? void 0 : entry.pendingWriteResolver)) {
      return;
    }
    while (entry.pendingWriteResolver) {
      const pending = entry.pendingWriteResolver;
      try {
        await pending.promise;
      } catch {
      }
    }
  }
  async setActorAlarm(actorId, timestamp) {
    const entry = this.#actors.get(actorId);
    invariant3(entry, "actor entry does not exist");
    const writeGeneration = entry.generation;
    if (this.isActorStopping(actorId)) {
      logger2().info("skipping set alarm since actor stopping");
      return;
    }
    if (this.#persist) {
      const alarmPath = this.getActorAlarmPath(actorId);
      const crypto2 = getNodeCrypto();
      const tempPath = `${alarmPath}.tmp.${crypto2.randomUUID()}`;
      try {
        const path = getNodePath();
        await ensureDirectoryExists(path.dirname(alarmPath));
        const alarmData = {
          actorId,
          timestamp: BigInt(timestamp)
        };
        const data = ACTOR_ALARM_VERSIONED.serializeWithEmbeddedVersion(
          alarmData,
          CURRENT_VERSION3
        );
        const fs = getNodeFs();
        await fs.writeFile(tempPath, data);
        if (!this.isGenerationCurrentAndNotDestroyed(
          actorId,
          writeGeneration
        )) {
          logger2().debug(
            "skipping writing alarm since actor destroying or new generation"
          );
          return;
        }
        await fs.rename(tempPath, alarmPath);
      } catch (error) {
        try {
          const fs = getNodeFs();
          await fs.unlink(tempPath);
        } catch {
        }
        logger2().error({
          msg: "failed to write alarm",
          actorId,
          error
        });
        throw new Error(`Failed to write alarm: ${error}`);
      }
    }
    this.#scheduleAlarmTimeout(actorId, timestamp);
  }
  /**
   * Perform the actual write operation with atomic writes
   */
  async #performWrite(actorId, generation, state) {
    const dataPath = this.getActorStatePath(actorId);
    const crypto2 = getNodeCrypto();
    const tempPath = `${dataPath}.tmp.${crypto2.randomUUID()}`;
    try {
      const path = getNodePath();
      await ensureDirectoryExists(path.dirname(dataPath));
      const bareState = {
        actorId: state.actorId,
        name: state.name,
        key: state.key,
        createdAt: state.createdAt,
        kvStorage: state.kvStorage,
        startTs: state.startTs,
        connectableTs: state.connectableTs,
        sleepTs: state.sleepTs,
        destroyTs: state.destroyTs
      };
      const serializedState = ACTOR_STATE_VERSIONED.serializeWithEmbeddedVersion(
        bareState,
        CURRENT_VERSION3
      );
      const fs = getNodeFs();
      await fs.writeFile(tempPath, serializedState);
      if (!this.isGenerationCurrentAndNotDestroyed(actorId, generation)) {
        logger2().debug(
          "skipping writing alarm since actor destroying or new generation"
        );
        return;
      }
      await fs.rename(tempPath, dataPath);
    } catch (error) {
      try {
        const fs = getNodeFs();
        await fs.unlink(tempPath);
      } catch {
      }
      logger2().error({
        msg: "failed to save actor state",
        actorId,
        error
      });
      throw new Error(`Failed to save actor state: ${error}`);
    }
  }
  /**
   * Call this method after the actor driver has been initiated.
   *
   * This will trigger all initial alarms from the file system.
   *
   * This needs to be sync since DriverConfig.actor is sync
   */
  onRunnerStart(config4, inlineClient, actorDriver) {
    if (this.#runnerParams) {
      return;
    }
    this.#runnerParams = {
      config: config4,
      inlineClient,
      actorDriver
    };
    try {
      this.#loadAlarmsSync();
    } catch (err) {
      logger2().error({
        msg: "failed to load alarms on startup",
        error: err
      });
    }
  }
  async startActor(config4, inlineClient, actorDriver, actorId) {
    var _a;
    await this.#waitForActorStop(actorId);
    let entry = await this.loadActor(actorId);
    if (!entry.state) {
      throw new Error(
        `Actor does not exist and cannot be started: "${actorId}"`
      );
    }
    if (entry.startPromise) {
      await entry.startPromise.promise;
      invariant3(entry.actor, "actor should have loaded");
      return entry.actor;
    }
    if (entry.actor) {
      if (entry.actor.isStopping || this.isActorStopping(actorId)) {
        await this.#waitForActorStop(actorId);
        entry = await this.loadActor(actorId);
        if (!entry.state) {
          throw new Error(
            `Actor does not exist and cannot be started: "${actorId}"`
          );
        }
      } else {
        return entry.actor;
      }
    }
    entry.startPromise = promiseWithResolvers((reason) => logger2().warn({ msg: "unhandled actor start promise rejection", reason }));
    try {
      const definition = lookupInRegistry(config4, entry.state.name);
      entry.actor = await definition.instantiate();
      entry.lifecycleState = 1 /* AWAKE */;
      await entry.actor.start(
        actorDriver,
        inlineClient,
        actorId,
        entry.state.name,
        entry.state.key,
        "unknown"
      );
      const now = BigInt(Date.now());
      await this.#withActorWrite(actorId, async (lockedEntry) => {
        if (!lockedEntry.state) {
          throw new Error(
            `Actor does not exist and cannot be started: "${actorId}"`
          );
        }
        lockedEntry.state = {
          ...lockedEntry.state,
          startTs: now,
          connectableTs: now,
          sleepTs: null
          // Clear sleep timestamp when actor wakes up
        };
        if (this.#persist) {
          await this.#performWrite(
            actorId,
            lockedEntry.generation,
            lockedEntry.state
          );
        }
      });
      entry.startPromise.resolve();
      entry.startPromise = void 0;
      return entry.actor;
    } catch (innerError) {
      const error = new Error(
        `Failed to start actor ${actorId}: ${innerError}`,
        { cause: innerError }
      );
      (_a = entry.startPromise) == null ? void 0 : _a.reject(error);
      entry.startPromise = void 0;
      throw error;
    }
  }
  async loadActorStateOrError(actorId) {
    const state = (await this.loadActor(actorId)).state;
    if (!state) throw new Error(`Actor does not exist: ${actorId}`);
    return state;
  }
  getActorOrError(actorId) {
    const entry = this.#actors.get(actorId);
    if (!entry) throw new Error(`No entry for actor: ${actorId}`);
    return entry;
  }
  async createDatabase(actorId) {
    return this.getActorDbPath(actorId);
  }
  /**
   * Load all persisted alarms from disk and schedule their timers.
   */
  #loadAlarmsSync() {
    try {
      const fsSync = getNodeFsSync();
      const files = fsSync.existsSync(this.#alarmsDir) ? fsSync.readdirSync(this.#alarmsDir) : [];
      for (const file of files) {
        if (file.includes(".tmp.")) continue;
        const path = getNodePath();
        const fullPath = path.join(this.#alarmsDir, file);
        try {
          const buf = fsSync.readFileSync(fullPath);
          const alarmData = ACTOR_ALARM_VERSIONED.deserializeWithEmbeddedVersion(
            new Uint8Array(buf)
          );
          const timestamp = Number(alarmData.timestamp);
          if (Number.isFinite(timestamp)) {
            this.#scheduleAlarmTimeout(
              alarmData.actorId,
              timestamp
            );
          } else {
            logger2().debug({
              msg: "invalid alarm file contents",
              file
            });
          }
        } catch (err) {
          logger2().error({
            msg: "failed to read alarm file",
            file,
            error: stringifyError(err)
          });
        }
      }
    } catch (err) {
      logger2().error({
        msg: "failed to list alarms directory",
        error: err
      });
    }
  }
  /**
   * Schedule an alarm timer for an actor without writing to disk.
   */
  #scheduleAlarmTimeout(actorId, timestamp) {
    var _a;
    const entry = this.#upsertEntry(actorId);
    if (entry.alarmTimestamp !== void 0 && timestamp >= entry.alarmTimestamp) {
      logger2().debug({
        msg: "skipping alarm schedule (later than existing)",
        actorId,
        timestamp,
        current: entry.alarmTimestamp
      });
      return;
    }
    logger2().debug({ msg: "scheduling alarm", actorId, timestamp });
    (_a = entry.alarmTimeout) == null ? void 0 : _a.abort();
    entry.alarmTimestamp = timestamp;
    const delay = Math.max(0, timestamp - Date.now());
    entry.alarmTimeout = setLongTimeout(async () => {
      entry.alarmTimestamp = void 0;
      if (this.#persist) {
        try {
          const fs = getNodeFs();
          await fs.unlink(this.getActorAlarmPath(actorId));
        } catch (err) {
          if ((err == null ? void 0 : err.code) !== "ENOENT") {
            logger2().debug({
              msg: "failed to remove alarm file",
              actorId,
              error: stringifyError(err)
            });
          }
        }
      }
      try {
        logger2().debug({ msg: "triggering alarm", actorId, timestamp });
        const loaded = await this.loadActor(actorId);
        if (!loaded.state)
          throw new Error(`Actor does not exist: ${actorId}`);
        const runnerParams = this.#runnerParams;
        invariant3(runnerParams, "missing runner params");
        if (!loaded.actor) {
          await this.startActor(
            runnerParams.config,
            runnerParams.inlineClient,
            runnerParams.actorDriver,
            actorId
          );
        }
        invariant3(loaded.actor, "actor should be loaded after wake");
        await loaded.actor.onAlarm();
      } catch (err) {
        logger2().error({
          msg: "failed to handle alarm",
          actorId,
          error: stringifyError(err)
        });
      }
    }, delay);
  }
  /**
   * Cleanup stale temp files on startup (synchronous)
   */
  #cleanupTempFilesSync() {
    try {
      const fsSync = getNodeFsSync();
      const files = fsSync.readdirSync(this.#stateDir);
      const tempFiles = files.filter((f) => f.includes(".tmp."));
      const oneHourAgo = Date.now() - 36e5;
      for (const tempFile of tempFiles) {
        try {
          const path = getNodePath();
          const fullPath = path.join(this.#stateDir, tempFile);
          const stat = fsSync.statSync(fullPath);
          if (stat.mtimeMs < oneHourAgo) {
            fsSync.unlinkSync(fullPath);
            logger2().info({
              msg: "cleaned up stale temp file",
              file: tempFile
            });
          }
        } catch (err) {
          logger2().debug({
            msg: "failed to cleanup temp file",
            file: tempFile,
            error: err
          });
        }
      }
    } catch (err) {
      logger2().error({
        msg: "failed to read actors directory for cleanup",
        error: err
      });
    }
  }
  /**
   * Batch put KV entries for an actor.
   */
  async kvBatchPut(actorId, entries) {
    await this.loadActor(actorId);
    await this.#withActorWrite(actorId, async (entry) => {
      if (!entry.state) {
        if (this.isActorStopping(actorId)) {
          return;
        }
        throw new Error(`Actor ${actorId} state not loaded`);
      }
      const db = this.#getOrCreateActorKvDatabase(actorId);
      const totalSize = estimateKvSize(db);
      validateKvEntries(entries, totalSize);
      this.#putKvEntriesInDb(db, entries);
    });
  }
  /**
   * Batch get KV entries for an actor.
   */
  async kvBatchGet(actorId, keys) {
    const entry = await this.loadActor(actorId);
    await this.#waitForPendingWrite(actorId);
    if (!entry.state) {
      if (this.isActorStopping(actorId)) {
        throw new Error(`Actor ${actorId} is stopping`);
      } else {
        throw new Error(`Actor ${actorId} state not loaded`);
      }
    }
    validateKvKeys(keys);
    const db = this.#getOrCreateActorKvDatabase(actorId);
    const results = [];
    for (const key of keys) {
      const row = db.get(
        "SELECT value FROM kv WHERE key = ?",
        [key]
      );
      if (!row) {
        results.push(null);
        continue;
      }
      results.push(ensureUint8Array(row.value, "value"));
    }
    return results;
  }
  /**
   * Batch delete KV entries for an actor.
   */
  async kvBatchDelete(actorId, keys) {
    await this.loadActor(actorId);
    await this.#withActorWrite(actorId, async (entry) => {
      if (!entry.state) {
        if (this.isActorStopping(actorId)) {
          return;
        }
        throw new Error(`Actor ${actorId} state not loaded`);
      }
      if (keys.length === 0) {
        return;
      }
      validateKvKeys(keys);
      const db = this.#getOrCreateActorKvDatabase(actorId);
      db.exec("BEGIN");
      try {
        for (const key of keys) {
          db.run("DELETE FROM kv WHERE key = ?", [key]);
        }
        db.exec("COMMIT");
      } catch (error) {
        try {
          db.exec("ROLLBACK");
        } catch {
        }
        throw error;
      }
    });
  }
  /**
   * List KV entries with a given prefix for an actor.
   */
  async kvListPrefix(actorId, prefix) {
    const entry = await this.loadActor(actorId);
    await this.#waitForPendingWrite(actorId);
    if (!entry.state) {
      if (this.isActorStopping(actorId)) {
        throw new Error(`Actor ${actorId} is destroying`);
      } else {
        throw new Error(`Actor ${actorId} state not loaded`);
      }
    }
    validateKvKey(prefix, "prefix key");
    const db = this.#getOrCreateActorKvDatabase(actorId);
    const upperBound = computePrefixUpperBound(prefix);
    const rows = upperBound ? db.all(
      "SELECT key, value FROM kv WHERE key >= ? AND key < ? ORDER BY key ASC",
      [prefix, upperBound]
    ) : db.all(
      "SELECT key, value FROM kv WHERE key >= ? ORDER BY key ASC",
      [prefix]
    );
    return rows.map((row) => [
      ensureUint8Array(row.key, "key"),
      ensureUint8Array(row.value, "value")
    ]);
  }
};

// src/drivers/file-system/manager.ts
import invariant4 from "invariant";
var FileSystemManagerDriver = class {
  #config;
  #state;
  #driverConfig;
  #getUpgradeWebSocket;
  #actorDriver;
  #actorRouter;
  constructor(config4, state, driverConfig) {
    this.#config = config4;
    this.#state = state;
    this.#driverConfig = driverConfig;
    const inlineClient = createClientWithDriver(this);
    this.#actorDriver = this.#driverConfig.actor(
      config4,
      this,
      inlineClient
    );
    this.#actorRouter = createActorRouter(
      this.#config,
      this.#actorDriver,
      void 0,
      config4.test.enabled
    );
  }
  async sendRequest(actorId, actorRequest) {
    return await this.#actorRouter.fetch(actorRequest, {
      actorId
    });
  }
  async openWebSocket(path, actorId, encoding, params) {
    const normalizedPath = path.startsWith("/") ? path : `/${path}`;
    const fakeUrl = `http://inline-actor${normalizedPath}`;
    const fakeRequest = new Request(fakeUrl, {
      method: "GET"
    });
    const pathOnly = normalizedPath.split("?")[0];
    const { gatewayId, requestId } = createHibernatableRequestMetadata();
    const wsHandler = await routeWebSocket(
      fakeRequest,
      pathOnly,
      {},
      this.#config,
      this.#actorDriver,
      actorId,
      encoding,
      params,
      gatewayId,
      requestId,
      true,
      false
    );
    return createInlineWebSocket(wsHandler);
  }
  async proxyRequest(c, actorRequest, actorId) {
    return await this.#actorRouter.fetch(actorRequest, {
      actorId
    });
  }
  async proxyWebSocket(c, path, actorId, encoding, params) {
    var _a;
    const upgradeWebSocket = (_a = this.#getUpgradeWebSocket) == null ? void 0 : _a.call(this);
    invariant4(upgradeWebSocket, "missing getUpgradeWebSocket");
    const pathOnly = path.split("?")[0];
    const normalizedPath = pathOnly.startsWith("/") ? pathOnly : `/${pathOnly}`;
    const { gatewayId, requestId } = createHibernatableRequestMetadata();
    const wsHandler = await routeWebSocket(
      // TODO: Create new request with new path
      c.req.raw,
      normalizedPath,
      c.req.header(),
      this.#config,
      this.#actorDriver,
      actorId,
      encoding,
      params,
      gatewayId,
      requestId,
      true,
      false
    );
    return upgradeWebSocket(() => wsHandler)(c, noopNext());
  }
  async buildGatewayUrl(actorId) {
    const port = this.#config.managerPort ?? 6420;
    return `http://127.0.0.1:${port}/gateway/${encodeURIComponent(actorId)}`;
  }
  async getForId({
    actorId
  }) {
    const actor = await this.#state.loadActor(actorId);
    if (!actor.state) {
      return void 0;
    }
    if (this.#state.isActorStopping(actorId)) {
      throw new ActorStopping(actorId);
    }
    return actorStateToOutput(actor.state);
  }
  async getWithKey({
    name,
    key
  }) {
    const actorId = generateActorId(name, key);
    const actor = await this.#state.loadActor(actorId);
    if (actor.state) {
      return actorStateToOutput(actor.state);
    }
    return void 0;
  }
  async getOrCreateWithKey(input) {
    const actorId = generateActorId(input.name, input.key);
    await this.#state.loadOrCreateActor(
      actorId,
      input.name,
      input.key,
      input.input
    );
    await this.#actorDriver.loadActor(actorId);
    const state = await this.#state.loadActorStateOrError(actorId);
    return actorStateToOutput(state);
  }
  async createActor({ name, key, input }) {
    const actorId = generateActorId(name, key);
    await this.#state.createActor(actorId, name, key, input);
    await this.#actorDriver.loadActor(actorId);
    const state = await this.#state.loadActorStateOrError(actorId);
    return actorStateToOutput(state);
  }
  async listActors({ name }) {
    const actors = [];
    const itr = this.#state.getActorsIterator({});
    for await (const actor of itr) {
      if (actor.name === name) {
        actors.push(actorStateToOutput(actor));
      }
    }
    actors.sort((a, b) => {
      const aTs = a.createTs ?? 0;
      const bTs = b.createTs ?? 0;
      return bTs - aTs;
    });
    return actors;
  }
  async kvGet(actorId, key) {
    const response = await this.#state.kvBatchGet(actorId, [key]);
    return response[0] !== null ? new TextDecoder().decode(response[0]) : null;
  }
  displayInformation() {
    return {
      properties: {
        ...this.#state.persist ? { Data: this.#state.storagePath } : {},
        Instances: this.#state.actorCountOnStartup.toString()
      }
    };
  }
  extraStartupLog() {
    return {
      instances: this.#state.actorCountOnStartup,
      data: this.#state.storagePath
    };
  }
  setGetUpgradeWebSocket(getUpgradeWebSocket) {
    this.#getUpgradeWebSocket = getUpgradeWebSocket;
  }
};
function actorStateToOutput(state) {
  return {
    actorId: state.actorId,
    name: state.name,
    key: state.key,
    createTs: Number(state.createdAt),
    startTs: state.startTs !== null ? Number(state.startTs) : null,
    connectableTs: state.connectableTs !== null ? Number(state.connectableTs) : null,
    sleepTs: state.sleepTs !== null ? Number(state.sleepTs) : null,
    destroyTs: state.destroyTs !== null ? Number(state.destroyTs) : null
  };
}
function createHibernatableRequestMetadata() {
  const gatewayId = new Uint8Array(4);
  const requestId = new Uint8Array(4);
  crypto.getRandomValues(gatewayId);
  crypto.getRandomValues(requestId);
  return {
    gatewayId: gatewayId.buffer.slice(0),
    requestId: requestId.buffer.slice(0)
  };
}

// src/drivers/file-system/mod.ts
var CreateFileSystemDriverOptionsSchema = z.object({
  /** Custom path for storage. */
  path: z.string().optional(),
  /** Deprecated: file-system driver KV is now always SQLite-backed. */
  useNativeSqlite: z.boolean().optional()
});
function createFileSystemOrMemoryDriver(persist = true, options) {
  importNodeDependencies();
  if ((options == null ? void 0 : options.useNativeSqlite) === false) {
    throw new Error(
      "File-system driver no longer supports non-SQLite KV storage. Remove useNativeSqlite: false."
    );
  }
  const stateOptions = {
    persist,
    customPath: options == null ? void 0 : options.path,
    useNativeSqlite: true
  };
  const state = new FileSystemGlobalState(stateOptions);
  const driverConfig = {
    name: persist ? "file-system" : "memory",
    displayName: persist ? "File System" : "Memory",
    manager: (config4) => new FileSystemManagerDriver(config4, state, driverConfig),
    actor: (config4, managerDriver, inlineClient) => {
      const actorDriver = new FileSystemActorDriver(
        config4,
        managerDriver,
        inlineClient,
        state
      );
      state.onRunnerStart(config4, inlineClient, actorDriver);
      return actorDriver;
    },
    autoStartActorDriver: true
  };
  return driverConfig;
}
function createFileSystemDriver(opts) {
  const validatedOpts = opts ? CreateFileSystemDriverOptionsSchema.parse(opts) : void 0;
  return createFileSystemOrMemoryDriver(true, validatedOpts);
}
function createMemoryDriver() {
  return createFileSystemOrMemoryDriver(false);
}

export {
  ActorDefinition,
  lookupInRegistry,
  createActorRouter,
  InlineWebSocketAdapter,
  getStoragePath,
  ensureDirectoryExists,
  createFileSystemOrMemoryDriver,
  createFileSystemDriver,
  createMemoryDriver
};
