// rivetkit/dist/tsup/chunk-VKVNIQRQ.js
import {
  AsyncMutex,
  createActorKvStore,
  toSqliteBindings
} from "./chunk-N4KRDJ56.js";
import {
  __export,
  __reExport
} from "./chunk-424PT5DM.js";

// src/db/drizzle/mod.ts
var mod_exports = {};
__export(mod_exports, {
  blob: () => blob,
  check: () => check,
  db: () => db,
  defineConfig: () => defineConfig,
  extractUsedTable: () => extractUsedTable,
  foreignKey: () => foreignKey,
  getTableConfig: () => getTableConfig,
  getViewConfig: () => getViewConfig,
  index: () => index,
  integer: () => integer,
  numeric: () => numeric,
  primaryKey: () => primaryKey,
  real: () => real,
  sqliteTable: () => sqliteTable,
  sqliteTableCreator: () => sqliteTableCreator,
  sqliteView: () => sqliteView,
  text: () => text,
  unique: () => unique,
  uniqueIndex: () => uniqueIndex,
  uniqueKeyName: () => uniqueKeyName,
  view: () => view
});
import {
  drizzle as proxyDrizzle
} from "drizzle-orm/sqlite-proxy";

// src/db/drizzle/sqlite-core.ts
var sqlite_core_exports = {};
__export(sqlite_core_exports, {
  blob: () => blob,
  check: () => check,
  extractUsedTable: () => extractUsedTable,
  foreignKey: () => foreignKey,
  getTableConfig: () => getTableConfig,
  getViewConfig: () => getViewConfig,
  index: () => index,
  integer: () => integer,
  numeric: () => numeric,
  primaryKey: () => primaryKey,
  real: () => real,
  sqliteTable: () => sqliteTable,
  sqliteTableCreator: () => sqliteTableCreator,
  sqliteView: () => sqliteView,
  text: () => text,
  unique: () => unique,
  uniqueIndex: () => uniqueIndex,
  uniqueKeyName: () => uniqueKeyName,
  view: () => view
});
__reExport(sqlite_core_exports, sqlite_core_star);
import * as sqlite_core_star from "drizzle-orm/sqlite-core";
import {
  blob,
  check,
  extractUsedTable,
  foreignKey,
  getTableConfig,
  getViewConfig,
  index,
  integer,
  numeric,
  primaryKey,
  real,
  sqliteTable,
  sqliteTableCreator,
  sqliteView,
  text,
  unique,
  uniqueIndex,
  uniqueKeyName,
  view
} from "drizzle-orm/sqlite-core";

// src/db/drizzle/mod.ts
__reExport(mod_exports, sqlite_core_exports);
import { defineConfig as originalDefineConfig } from "drizzle-kit";
function defineConfig(config) {
  return originalDefineConfig({
    dialect: "sqlite",
    driver: "durable-sqlite",
    ...config
  });
}
function createProxyCallback(waDb, mutex, isClosed) {
  return async (sql, params, method) => {
    return mutex.run(async () => {
      if (isClosed()) {
        throw new Error("database is closed");
      }
      if (method === "run") {
        await waDb.run(sql, toSqliteBindings(params));
        return { rows: [] };
      }
      const result = await waDb.query(sql, toSqliteBindings(params));
      if (method === "get") {
        return { rows: result.rows[0] };
      }
      return { rows: result.rows };
    });
  };
}
async function runInlineMigrations(waDb, mutex, migrations) {
  await mutex.run(
    () => waDb.exec(`
		CREATE TABLE IF NOT EXISTS __drizzle_migrations (
			id INTEGER PRIMARY KEY AUTOINCREMENT,
			hash TEXT NOT NULL,
			created_at INTEGER
		)
	`)
  );
  let lastCreatedAt = 0;
  await mutex.run(
    () => waDb.exec(
      "SELECT id, hash, created_at FROM __drizzle_migrations ORDER BY created_at DESC LIMIT 1",
      (row) => {
        lastCreatedAt = Number(row[2]) || 0;
      }
    )
  );
  const journal = migrations.journal;
  if (!(journal == null ? void 0 : journal.entries)) return;
  for (const entry of journal.entries) {
    if (entry.when <= lastCreatedAt) continue;
    const migrationKey = `m${String(entry.idx).padStart(4, "0")}`;
    const sql = migrations.migrations[migrationKey];
    if (!sql) continue;
    await mutex.run(() => waDb.exec(sql));
    await mutex.run(
      () => waDb.run(
        "INSERT INTO __drizzle_migrations (hash, created_at) VALUES (?, ?)",
        [entry.tag, entry.when]
      )
    );
  }
}
function db(config) {
  let waDbInstance = null;
  const mutex = new AsyncMutex();
  return {
    createClient: async (ctx) => {
      if (!ctx.sqliteVfs) {
        throw new Error(
          "SqliteVfs instance not provided in context. The driver must provide a sqliteVfs instance."
        );
      }
      const kvStore = createActorKvStore(ctx.kv);
      const waDb = await ctx.sqliteVfs.open(ctx.actorId, kvStore);
      waDbInstance = waDb;
      let closed = false;
      const ensureOpen = () => {
        if (closed) {
          throw new Error("database is closed");
        }
      };
      const callback = createProxyCallback(waDb, mutex, () => closed);
      const client = proxyDrizzle(callback, config);
      return Object.assign(client, {
        execute: async (query, ...args) => {
          return mutex.run(async () => {
            ensureOpen();
            if (args.length > 0) {
              const result = await waDb.query(
                query,
                toSqliteBindings(args)
              );
              return result.rows.map((row) => {
                const obj = {};
                for (let i = 0; i < result.columns.length; i++) {
                  obj[result.columns[i]] = row[i];
                }
                return obj;
              });
            }
            const results = [];
            let columnNames = null;
            await waDb.exec(
              query,
              (row, columns) => {
                if (!columnNames) {
                  columnNames = columns;
                }
                const obj = {};
                for (let i = 0; i < row.length; i++) {
                  obj[columnNames[i]] = row[i];
                }
                results.push(obj);
              }
            );
            return results;
          });
        },
        close: async () => {
          await mutex.run(async () => {
            if (closed) {
              return;
            }
            closed = true;
            await waDb.close();
            waDbInstance = null;
          });
        }
      });
    },
    onMigrate: async (_client) => {
      if ((config == null ? void 0 : config.migrations) && waDbInstance) {
        await runInlineMigrations(
          waDbInstance,
          mutex,
          config.migrations
        );
      }
    },
    onDestroy: async (client) => {
      await client.close();
    }
  };
}

export {
  blob,
  check,
  extractUsedTable,
  foreignKey,
  getTableConfig,
  getViewConfig,
  index,
  integer,
  numeric,
  primaryKey,
  real,
  sqliteTable,
  sqliteTableCreator,
  sqliteView,
  text,
  unique,
  uniqueIndex,
  uniqueKeyName,
  view,
  defineConfig,
  db,
  mod_exports
};
