// rivetkit/dist/tsup/utils.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


















var _chunk2Q5WS6QHcjs = require('./chunk-2Q5WS6QH.cjs');
require('./chunk-L47L3ZWJ.cjs');
require('./chunk-HDQ2JUQT.cjs');



















exports.EXTRA_ERROR_LOG = _chunk2Q5WS6QHcjs.EXTRA_ERROR_LOG; exports.SinglePromiseQueue = _chunk2Q5WS6QHcjs.SinglePromiseQueue; exports.VERSION = _chunk2Q5WS6QHcjs.VERSION; exports.arrayBuffersEqual = _chunk2Q5WS6QHcjs.arrayBuffersEqual; exports.assertUnreachable = _chunk2Q5WS6QHcjs.assertUnreachable; exports.bufferToArrayBuffer = _chunk2Q5WS6QHcjs.bufferToArrayBuffer; exports.combineUrlPath = _chunk2Q5WS6QHcjs.combineUrlPath; exports.dbg = _chunk2Q5WS6QHcjs.dbg; exports.detectRuntime = _chunk2Q5WS6QHcjs.detectRuntime; exports.getEnvUniversal = _chunk2Q5WS6QHcjs.getEnvUniversal; exports.httpUserAgent = _chunk2Q5WS6QHcjs.httpUserAgent; exports.interval = _chunk2Q5WS6QHcjs.interval; exports.joinSignals = _chunk2Q5WS6QHcjs.joinSignals; exports.promiseWithResolvers = _chunk2Q5WS6QHcjs.promiseWithResolvers; exports.setLongTimeout = _chunk2Q5WS6QHcjs.setLongTimeout; exports.sleep = _chunk2Q5WS6QHcjs.sleep; exports.stringifyError = _chunk2Q5WS6QHcjs.stringifyError; exports.toUint8Array = _chunk2Q5WS6QHcjs.toUint8Array;
