// @rivetkit/virtual-websocket/dist/mod.js
// src/virtual-websocket.ts
var VirtualWebSocket = class {
  static CONNECTING = 0;
  static OPEN = 1;
  static CLOSING = 2;
  static CLOSED = 3;
  CONNECTING = 0;
  OPEN = 1;
  CLOSING = 2;
  CLOSED = 3;
  #options;
  #listeners = /* @__PURE__ */ new Map();
  #onopen = null;
  #onclose = null;
  #onerror = null;
  #onmessage = null;
  constructor(options) {
    this.#options = options;
  }
  // UniversalWebSocket properties
  get readyState() {
    return this.#options.getReadyState();
  }
  /**
   * Binary type for message data. Only "arraybuffer" is supported.
   * Setting to "blob" will throw an error.
   */
  get binaryType() {
    return "arraybuffer";
  }
  set binaryType(value) {
    if (value !== "arraybuffer") {
      throw new DOMException(
        'VirtualWebSocket only supports binaryType "arraybuffer"',
        "NotSupportedError"
      );
    }
  }
  get bufferedAmount() {
    return 0;
  }
  get extensions() {
    return "";
  }
  get protocol() {
    return "";
  }
  get url() {
    return "";
  }
  // UniversalWebSocket methods
  send(data) {
    const state = this.readyState;
    if (state === this.CONNECTING) {
      throw new DOMException(
        "WebSocket is still in CONNECTING state",
        "InvalidStateError"
      );
    }
    if (state === this.CLOSING || state === this.CLOSED) {
      return;
    }
    this.#options.onSend(data);
  }
  close(code = 1e3, reason = "") {
    const state = this.readyState;
    if (state === this.CLOSED || state === this.CLOSING) {
      return;
    }
    if (code !== 1e3 && (code < 3e3 || code > 4999) && code !== 1001 && code !== 1002 && code !== 1003 && code !== 1007 && code !== 1008 && code !== 1009 && code !== 1010 && code !== 1011) {
      throw new DOMException("Invalid close code", "InvalidAccessError");
    }
    if (new TextEncoder().encode(reason).length > 123) {
      throw new DOMException("Close reason too long", "SyntaxError");
    }
    this.#options.onClose(code, reason);
  }
  /** @experimental Immediate close without close frame */
  terminate() {
    if (this.#options.onTerminate) {
      this.#options.onTerminate();
    } else {
      this.#options.onClose(1006, "Abnormal Closure");
    }
  }
  addEventListener(type, listener) {
    if (!this.#listeners.has(type)) {
      this.#listeners.set(type, []);
    }
    this.#listeners.get(type).push(listener);
  }
  removeEventListener(type, listener) {
    const listeners = this.#listeners.get(type);
    if (listeners) {
      const index = listeners.indexOf(listener);
      if (index !== -1) {
        listeners.splice(index, 1);
      }
    }
  }
  dispatchEvent(event) {
    this.#dispatch(event.type, event);
    return true;
  }
  // on* property getters/setters
  get onopen() {
    return this.#onopen;
  }
  set onopen(fn) {
    this.#onopen = fn;
  }
  get onclose() {
    return this.#onclose;
  }
  set onclose(fn) {
    this.#onclose = fn;
  }
  get onerror() {
    return this.#onerror;
  }
  set onerror(fn) {
    this.#onerror = fn;
  }
  get onmessage() {
    return this.#onmessage;
  }
  set onmessage(fn) {
    this.#onmessage = fn;
  }
  // Helper methods to trigger events from the other side
  triggerOpen() {
    const event = {
      type: "open",
      target: this,
      currentTarget: this
    };
    this.#dispatch("open", event);
  }
  triggerMessage(data) {
    const event = {
      type: "message",
      data,
      target: this,
      currentTarget: this
    };
    this.#dispatch("message", event);
  }
  triggerClose(code, reason, wasClean) {
    const event = {
      type: "close",
      code,
      reason,
      wasClean: wasClean ?? code === 1e3,
      target: this,
      currentTarget: this
    };
    this.#dispatch("close", event);
  }
  triggerError(error) {
    const event = {
      type: "error",
      target: this,
      currentTarget: this,
      error,
      message: error instanceof Error ? error.message : String(error)
    };
    this.#dispatch("error", event);
  }
  #dispatch(type, event) {
    const listeners = this.#listeners.get(type);
    if (listeners && listeners.length > 0) {
      for (const listener of listeners) {
        try {
          listener(event);
        } catch (err) {
          console.error("Error in WebSocket event listener:", err);
        }
      }
    }
    switch (type) {
      case "open":
        if (this.#onopen) {
          try {
            this.#onopen(event);
          } catch (err) {
            console.error(
              "Error in WebSocket onopen handler:",
              err
            );
          }
        }
        break;
      case "close":
        if (this.#onclose) {
          try {
            this.#onclose(event);
          } catch (err) {
            console.error(
              "Error in WebSocket onclose handler:",
              err
            );
          }
        }
        break;
      case "error":
        if (this.#onerror) {
          try {
            this.#onerror(event);
          } catch (err) {
            console.error(
              "Error in WebSocket onerror handler:",
              err
            );
          }
        }
        break;
      case "message":
        if (this.#onmessage) {
          try {
            this.#onmessage(event);
          } catch (err) {
            console.error(
              "Error in WebSocket onmessage handler:",
              err
            );
          }
        }
        break;
    }
  }
};
export {
  VirtualWebSocket
};
