// vite-plugin-srvx/dist/index.js
// src/build.ts
import path from "path";
var defaultBuildOptions = {
  entry: "./src/server.ts",
  outDir: "dist",
  serverOutFile: "server.js"
};
function srvxBuild(options) {
  const mergedOptions = { ...defaultBuildOptions, ...options };
  const isVercel = mergedOptions.framework === "vercel" || process.env.VERCEL === "1";
  return {
    name: "vite-plugin-srvx-server-build",
    // Only apply during server builds (not client builds)
    apply(_config, { command, mode }) {
      return command === "build" && mode === "server";
    },
    config() {
      const serverOutDir = isVercel ? path.join(mergedOptions.outDir || "dist", "api") : mergedOptions.outDir || "dist";
      const serverOutFile = isVercel ? "index.js" : mergedOptions.serverOutFile || "server.js";
      return {
        build: {
          ssr: true,
          outDir: serverOutDir,
          copyPublicDir: false,
          // Don't copy public assets during server build
          rollupOptions: {
            input: mergedOptions.entry,
            output: {
              entryFileNames: serverOutFile,
              format: "esm"
            }
          },
          emptyOutDir: false
        }
      };
    },
    buildEnd() {
      const serverOutDir = isVercel ? path.join(mergedOptions.outDir || "dist", "api") : mergedOptions.outDir || "dist";
      const serverOutFile = isVercel ? "index.js" : mergedOptions.serverOutFile || "server.js";
      const serverPath = path.join(serverOutDir, serverOutFile);
      console.log("\n\u2705 Server built successfully!");
      console.log(`\u{1F4E6} Server output: ${serverPath}`);
      if (isVercel) {
        console.log("\n\u{1F680} Deploy to Vercel:");
        console.log("   1. Configure vercel.json with:");
        console.log("      {");
        console.log('        "functions": {');
        console.log('          "dist/api/**": { "runtime": "edge" }');
        console.log("        }");
        console.log("      }");
        console.log("   2. Run: vercel deploy\n");
        console.log("\u{1F4A1} Output structure:");
        console.log("   dist/api/index.js  - Edge Function");
        console.log("   dist/public/       - Static assets");
      } else {
        console.log("\n\u{1F680} Run your app with:");
        console.log(`   srvx ${serverPath}
`);
        console.log("\u{1F4A1} Tip: srvx will automatically serve static files from dist/public/");
      }
    }
  };
}

// src/client-build.ts
var defaultClientBuildOptions = {
  outDir: "public",
  emptyOutDir: true
};
function clientBuild(options) {
  const mergedOptions = { ...defaultClientBuildOptions, ...options };
  return {
    name: "vite-plugin-srvx-client-build",
    // Only apply during client builds (not server builds)
    apply(_config, { command, mode }) {
      return command === "build" && mode !== "server";
    },
    config() {
      return {
        build: {
          outDir: mergedOptions.outDir,
          emptyOutDir: mergedOptions.emptyOutDir
        }
      };
    }
  };
}

// src/dev-server.ts
import fs from "fs";
import path2 from "path";
var defaultOptions = {
  entry: "./src/server.ts",
  exclude: [
    /.*\.tsx?$/,
    /.*\.ts$/,
    /.*\.jsx?$/,
    /.*\.css$/,
    /.*\.scss$/,
    /.*\.sass$/,
    /.*\.less$/,
    /.*\.styl$/,
    /.*\.png$/,
    /.*\.jpg$/,
    /.*\.jpeg$/,
    /.*\.gif$/,
    /.*\.svg$/,
    /.*\.webp$/,
    /^\/@.+$/,
    /^\/node_modules\/.*/
  ],
  injectClientScript: true
};
function createMiddleware(server, options) {
  return async (req, res, next) => {
    const config = server.config;
    const base = config.base === "/" ? "" : config.base;
    if (req.url === "/" || req.url === base || req.url === `${base}/`) {
      const indexPath = path2.join(config.root, "index.html");
      if (fs.existsSync(indexPath)) {
        const html = await server.transformIndexHtml(
          req.url,
          fs.readFileSync(indexPath, "utf-8")
        );
        res.statusCode = 200;
        res.setHeader("Content-Type", "text/html");
        res.setHeader("Content-Length", Buffer.byteLength(html));
        res.end(html);
        return;
      }
    }
    const exclude = options.exclude ?? defaultOptions.exclude ?? [];
    const urlPath = req.url?.split("?")[0];
    for (const pattern of exclude) {
      if (urlPath) {
        if (pattern instanceof RegExp) {
          if (pattern.test(urlPath)) {
            return next();
          }
        } else if (typeof pattern === "string") {
          if (urlPath.startsWith(pattern)) {
            return next();
          }
        }
      }
    }
    if (req.url?.startsWith(base)) {
      const publicDir = config.publicDir;
      if (publicDir && fs.existsSync(publicDir)) {
        const filePath = path2.join(publicDir, req.url.replace(base, ""));
        if (fs.existsSync(filePath) && fs.statSync(filePath).isFile()) {
          return next();
        }
      }
    }
    let app;
    try {
      const loadModule = options.loadModule ?? ((server2, entry) => server2.ssrLoadModule(entry));
      const module = await loadModule(server, options.entry);
      if ("default" in module) {
        app = module.default;
      } else {
        app = module;
      }
      if (!app?.fetch) {
        throw new Error("No fetch handler found in the entry module");
      }
    } catch (e) {
      return next();
    }
    const protocol = req.socket.encrypted ? "https" : "http";
    const host = req.headers.host || "localhost";
    const url = `${protocol}://${host}${req.url}`;
    const headers = new Headers();
    for (const [key, value] of Object.entries(req.headers)) {
      if (value !== void 0) {
        if (Array.isArray(value)) {
          for (const v of value) {
            headers.append(key, v);
          }
        } else {
          headers.set(key, value);
        }
      }
    }
    const body = req.method !== "GET" && req.method !== "HEAD" ? req : void 0;
    const request = new Request(url, {
      method: req.method,
      headers,
      body,
      duplex: "half"
    });
    const response = await app.fetch(request);
    res.statusCode = response.status;
    response.headers.forEach((value, key) => {
      res.setHeader(key, value);
    });
    if (options.injectClientScript !== false) {
      const contentType = response.headers.get("content-type");
      if (contentType?.includes("text/html")) {
        const body2 = await response.text();
        let script = `<script type="module" src="/@vite/client"></script>`;
        const nonce = response.headers.get("content-security-policy-nonce");
        if (nonce) {
          script = `<script type="module" nonce="${nonce}" src="/@vite/client"></script>`;
        }
        const injectedBody = body2.replace("</head>", `${script}</head>`);
        res.setHeader("content-length", Buffer.byteLength(injectedBody));
        res.end(injectedBody);
        return;
      }
    }
    if (response.body) {
      const reader = response.body.getReader();
      const stream = async () => {
        try {
          while (true) {
            const { done, value } = await reader.read();
            if (done) break;
            res.write(value);
          }
          res.end();
        } catch (error) {
          res.end();
        }
      };
      await stream();
    } else {
      res.end();
    }
  };
}
function devServer(options) {
  const entry = options?.entry ?? defaultOptions.entry;
  return {
    name: "vite-plugin-srvx-dev",
    apply: "serve",
    configureServer(server) {
      const mergedOptions = {
        ...defaultOptions,
        ...options,
        entry
      };
      server.middlewares.use(createMiddleware(server, mergedOptions));
    }
  };
}

// src/srvx.ts
var defaultSrvxOptions = {
  entry: "./src/server.ts",
  outDir: "dist",
  clientOutDir: "public",
  serverOutFile: "server.js"
};
function srvx(options) {
  const mergedOptions = { ...defaultSrvxOptions, ...options };
  return [
    // Development server plugin
    devServer({
      entry: mergedOptions.entry,
      exclude: mergedOptions.exclude,
      injectClientScript: mergedOptions.injectClientScript,
      loadModule: mergedOptions.loadModule
    }),
    // Client build plugin
    clientBuild({
      outDir: mergedOptions.clientOutDir,
      emptyOutDir: mergedOptions.clientEmptyOutDir
    }),
    // Server build plugin
    srvxBuild({
      entry: mergedOptions.entry,
      outDir: mergedOptions.outDir,
      serverOutFile: mergedOptions.serverOutFile,
      framework: mergedOptions.framework
    })
  ];
}

// src/index.ts
var index_default = srvx;
export {
  clientBuild,
  index_default as default,
  defaultBuildOptions,
  defaultClientBuildOptions,
  defaultOptions,
  defaultSrvxOptions,
  devServer,
  srvx,
  srvxBuild
};
