// @rivetkit/workflow-engine/dist/tsup/testing.js
import {
  CancelledError,
  CriticalError,
  DEFAULT_LOOP_COMMIT_INTERVAL,
  DEFAULT_LOOP_HISTORY_EVERY,
  DEFAULT_LOOP_HISTORY_KEEP,
  DEFAULT_MAX_RETRIES,
  DEFAULT_RETRY_BACKOFF_BASE,
  DEFAULT_RETRY_BACKOFF_MAX,
  DEFAULT_STEP_TIMEOUT,
  EntryInProgressError,
  EvictedError,
  HistoryDivergedError,
  JoinError,
  Loop,
  MessageWaitError,
  RaceError,
  RollbackCheckpointError,
  RollbackError,
  SleepError,
  StepExhaustedError,
  StepFailedError,
  WorkflowContextImpl,
  appendLoopIteration,
  appendName,
  compareKeys,
  createEntry,
  createHistorySnapshot,
  createStorage,
  deleteEntriesWithPrefix,
  emptyLocation,
  flush,
  generateId,
  getEntry,
  getOrCreateMetadata,
  isLocationPrefix,
  isLoopIterationMarker,
  keyStartsWith,
  keyToHex,
  loadMetadata,
  loadStorage,
  locationToKey,
  locationsEqual,
  parentLocation,
  registerName,
  resolveName,
  runWorkflow,
  setEntry,
  sleep
} from "./chunk-JTLDEP6X.js";

// src/testing.ts
var InMemoryWorkflowMessageDriver = class {
  #messages = [];
  #waiters = /* @__PURE__ */ new Set();
  async addMessage(message) {
    this.#messages.push(message);
    this.#notifyWaiters(message.name);
  }
  async receiveMessages(opts) {
    const limitedCount = Math.max(1, opts.count);
    const nameSet = opts.names && opts.names.length > 0 ? new Set(opts.names) : void 0;
    const selected = [];
    for (let i = 0; i < this.#messages.length && selected.length < limitedCount; i++) {
      const message = this.#messages[i];
      if (nameSet && !nameSet.has(message.name)) {
        continue;
      }
      selected.push({ message, index: i });
    }
    if (selected.length === 0) {
      return [];
    }
    if (!opts.completable) {
      for (let i = selected.length - 1; i >= 0; i--) {
        this.#messages.splice(selected[i].index, 1);
      }
      return selected.map((entry) => entry.message);
    }
    return selected.map((entry) => {
      const { message } = entry;
      return {
        ...message,
        complete: async () => {
          await this.completeMessage(message.id);
        }
      };
    });
  }
  async completeMessage(messageId) {
    const index = this.#messages.findIndex((message) => message.id === messageId);
    if (index !== -1) {
      this.#messages.splice(index, 1);
    }
  }
  async waitForMessages(messageNames, abortSignal) {
    if (abortSignal.aborted) {
      throw new EvictedError();
    }
    const nameSet = messageNames.length > 0 ? new Set(messageNames) : void 0;
    if (this.#messages.some(
      (message) => nameSet ? nameSet.has(message.name) : true
    )) {
      return;
    }
    await new Promise((resolve, reject) => {
      const waiter = {
        nameSet,
        resolve: () => {
          this.#removeWaiter(waiter);
          resolve();
        },
        reject: (error) => {
          this.#removeWaiter(waiter);
          reject(error);
        },
        abortSignal,
        onAbort: () => {
          waiter.reject(new EvictedError());
        }
      };
      abortSignal.addEventListener("abort", waiter.onAbort, { once: true });
      this.#waiters.add(waiter);
    });
  }
  #removeWaiter(waiter) {
    if (this.#waiters.delete(waiter)) {
      waiter.abortSignal.removeEventListener("abort", waiter.onAbort);
    }
  }
  #notifyWaiters(name) {
    for (const waiter of [...this.#waiters]) {
      if (waiter.nameSet && !waiter.nameSet.has(name)) {
        continue;
      }
      waiter.resolve();
    }
  }
  clear() {
    this.#messages = [];
    for (const waiter of [...this.#waiters]) {
      waiter.reject(new Error("cleared"));
    }
  }
};
var InMemoryDriver = class {
  // Map from hex-encoded key to { originalKey, value }
  kv = /* @__PURE__ */ new Map();
  alarms = /* @__PURE__ */ new Map();
  #inMemoryMessageDriver = new InMemoryWorkflowMessageDriver();
  /** Simulated latency per operation (ms) */
  latency = 10;
  /** How often the worker polls for work */
  workerPollInterval = 100;
  messageDriver = this.#inMemoryMessageDriver;
  async get(key) {
    await sleep(this.latency);
    const entry = this.kv.get(keyToHex(key));
    return (entry == null ? void 0 : entry.value) ?? null;
  }
  async set(key, value) {
    await sleep(this.latency);
    this.kv.set(keyToHex(key), { key, value });
  }
  async delete(key) {
    await sleep(this.latency);
    this.kv.delete(keyToHex(key));
  }
  async deletePrefix(prefix) {
    await sleep(this.latency);
    for (const [hexKey, entry] of this.kv) {
      if (keyStartsWith(entry.key, prefix)) {
        this.kv.delete(hexKey);
      }
    }
  }
  async list(prefix) {
    await sleep(this.latency);
    const results = [];
    for (const entry of this.kv.values()) {
      if (keyStartsWith(entry.key, prefix)) {
        results.push({ key: entry.key, value: entry.value });
      }
    }
    return results.sort((a, b) => compareKeys(a.key, b.key));
  }
  async batch(writes) {
    await sleep(this.latency);
    for (const { key, value } of writes) {
      this.kv.set(keyToHex(key), { key, value });
    }
  }
  async setAlarm(workflowId, wakeAt) {
    await sleep(this.latency);
    this.alarms.set(workflowId, wakeAt);
  }
  async clearAlarm(workflowId) {
    await sleep(this.latency);
    this.alarms.delete(workflowId);
  }
  async waitForMessages(messageNames, abortSignal) {
    const driver = this.messageDriver;
    if (driver.waitForMessages) {
      await driver.waitForMessages(messageNames, abortSignal);
      return;
    }
    while (true) {
      if (abortSignal.aborted) {
        throw new EvictedError();
      }
      const messages = await this.messageDriver.receiveMessages({
        names: messageNames.length > 0 ? messageNames : void 0,
        count: 1,
        completable: true
      });
      if (messages.length > 0) {
        return;
      }
      await sleep(Math.max(1, this.latency));
    }
  }
  /**
   * Get the alarm time for a workflow (for testing).
   */
  getAlarm(workflowId) {
    return this.alarms.get(workflowId);
  }
  /**
   * Check if any alarms are due and return their workflow IDs.
   */
  getDueAlarms() {
    const now = Date.now();
    const due = [];
    for (const [workflowId, wakeAt] of this.alarms) {
      if (wakeAt <= now) {
        due.push(workflowId);
      }
    }
    return due;
  }
  /**
   * Clear all data (for testing).
   */
  clear() {
    this.kv.clear();
    this.alarms.clear();
    this.#inMemoryMessageDriver.clear();
  }
  /**
   * Get a snapshot of all data (for testing/debugging).
   */
  snapshot() {
    const kvSnapshot = {};
    for (const [hexKey, entry] of this.kv) {
      kvSnapshot[hexKey] = entry.value;
    }
    return {
      kv: kvSnapshot,
      alarms: Object.fromEntries(this.alarms)
    };
  }
  /**
   * Get all hex-encoded keys (for testing).
   */
  keys() {
    return [...this.kv.keys()];
  }
};
export {
  CancelledError,
  CriticalError,
  DEFAULT_LOOP_COMMIT_INTERVAL,
  DEFAULT_LOOP_HISTORY_EVERY,
  DEFAULT_LOOP_HISTORY_KEEP,
  DEFAULT_MAX_RETRIES,
  DEFAULT_RETRY_BACKOFF_BASE,
  DEFAULT_RETRY_BACKOFF_MAX,
  DEFAULT_STEP_TIMEOUT,
  EntryInProgressError,
  EvictedError,
  HistoryDivergedError,
  InMemoryDriver,
  JoinError,
  Loop,
  MessageWaitError,
  RaceError,
  RollbackCheckpointError,
  RollbackError,
  SleepError,
  StepExhaustedError,
  StepFailedError,
  WorkflowContextImpl,
  appendLoopIteration,
  appendName,
  createEntry,
  createHistorySnapshot,
  createStorage,
  deleteEntriesWithPrefix,
  emptyLocation,
  flush,
  generateId,
  getEntry,
  getOrCreateMetadata,
  isLocationPrefix,
  isLoopIterationMarker,
  loadMetadata,
  loadStorage,
  locationToKey,
  locationsEqual,
  parentLocation,
  registerName,
  resolveName,
  runWorkflow,
  setEntry
};
