// @rivetkit/workflow-engine/dist/tsup/testing.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } var _class;
















































var _chunkKQO2TD7Tcjs = require('./chunk-KQO2TD7T.cjs');

// src/testing.ts
var InMemoryWorkflowMessageDriver = class {
  #messages = [];
  #waiters = /* @__PURE__ */ new Set();
  async addMessage(message) {
    this.#messages.push(message);
    this.#notifyWaiters(message.name);
  }
  async receiveMessages(opts) {
    const limitedCount = Math.max(1, opts.count);
    const nameSet = opts.names && opts.names.length > 0 ? new Set(opts.names) : void 0;
    const selected = [];
    for (let i = 0; i < this.#messages.length && selected.length < limitedCount; i++) {
      const message = this.#messages[i];
      if (nameSet && !nameSet.has(message.name)) {
        continue;
      }
      selected.push({ message, index: i });
    }
    if (selected.length === 0) {
      return [];
    }
    if (!opts.completable) {
      for (let i = selected.length - 1; i >= 0; i--) {
        this.#messages.splice(selected[i].index, 1);
      }
      return selected.map((entry) => entry.message);
    }
    return selected.map((entry) => {
      const { message } = entry;
      return {
        ...message,
        complete: async () => {
          await this.completeMessage(message.id);
        }
      };
    });
  }
  async completeMessage(messageId) {
    const index = this.#messages.findIndex((message) => message.id === messageId);
    if (index !== -1) {
      this.#messages.splice(index, 1);
    }
  }
  async waitForMessages(messageNames, abortSignal) {
    if (abortSignal.aborted) {
      throw new (0, _chunkKQO2TD7Tcjs.EvictedError)();
    }
    const nameSet = messageNames.length > 0 ? new Set(messageNames) : void 0;
    if (this.#messages.some(
      (message) => nameSet ? nameSet.has(message.name) : true
    )) {
      return;
    }
    await new Promise((resolve, reject) => {
      const waiter = {
        nameSet,
        resolve: () => {
          this.#removeWaiter(waiter);
          resolve();
        },
        reject: (error) => {
          this.#removeWaiter(waiter);
          reject(error);
        },
        abortSignal,
        onAbort: () => {
          waiter.reject(new (0, _chunkKQO2TD7Tcjs.EvictedError)());
        }
      };
      abortSignal.addEventListener("abort", waiter.onAbort, { once: true });
      this.#waiters.add(waiter);
    });
  }
  #removeWaiter(waiter) {
    if (this.#waiters.delete(waiter)) {
      waiter.abortSignal.removeEventListener("abort", waiter.onAbort);
    }
  }
  #notifyWaiters(name) {
    for (const waiter of [...this.#waiters]) {
      if (waiter.nameSet && !waiter.nameSet.has(name)) {
        continue;
      }
      waiter.resolve();
    }
  }
  clear() {
    this.#messages = [];
    for (const waiter of [...this.#waiters]) {
      waiter.reject(new Error("cleared"));
    }
  }
};
var InMemoryDriver = (_class = class {constructor() { _class.prototype.__init.call(this);_class.prototype.__init2.call(this);_class.prototype.__init3.call(this);_class.prototype.__init4.call(this);_class.prototype.__init5.call(this); }
  // Map from hex-encoded key to { originalKey, value }
  __init() {this.kv = /* @__PURE__ */ new Map()}
  __init2() {this.alarms = /* @__PURE__ */ new Map()}
  #inMemoryMessageDriver = new InMemoryWorkflowMessageDriver();
  /** Simulated latency per operation (ms) */
  __init3() {this.latency = 10}
  /** How often the worker polls for work */
  __init4() {this.workerPollInterval = 100}
  __init5() {this.messageDriver = this.#inMemoryMessageDriver}
  async get(key) {
    await _chunkKQO2TD7Tcjs.sleep.call(void 0, this.latency);
    const entry = this.kv.get(_chunkKQO2TD7Tcjs.keyToHex.call(void 0, key));
    return _nullishCoalesce((entry == null ? void 0 : entry.value), () => ( null));
  }
  async set(key, value) {
    await _chunkKQO2TD7Tcjs.sleep.call(void 0, this.latency);
    this.kv.set(_chunkKQO2TD7Tcjs.keyToHex.call(void 0, key), { key, value });
  }
  async delete(key) {
    await _chunkKQO2TD7Tcjs.sleep.call(void 0, this.latency);
    this.kv.delete(_chunkKQO2TD7Tcjs.keyToHex.call(void 0, key));
  }
  async deletePrefix(prefix) {
    await _chunkKQO2TD7Tcjs.sleep.call(void 0, this.latency);
    for (const [hexKey, entry] of this.kv) {
      if (_chunkKQO2TD7Tcjs.keyStartsWith.call(void 0, entry.key, prefix)) {
        this.kv.delete(hexKey);
      }
    }
  }
  async list(prefix) {
    await _chunkKQO2TD7Tcjs.sleep.call(void 0, this.latency);
    const results = [];
    for (const entry of this.kv.values()) {
      if (_chunkKQO2TD7Tcjs.keyStartsWith.call(void 0, entry.key, prefix)) {
        results.push({ key: entry.key, value: entry.value });
      }
    }
    return results.sort((a, b) => _chunkKQO2TD7Tcjs.compareKeys.call(void 0, a.key, b.key));
  }
  async batch(writes) {
    await _chunkKQO2TD7Tcjs.sleep.call(void 0, this.latency);
    for (const { key, value } of writes) {
      this.kv.set(_chunkKQO2TD7Tcjs.keyToHex.call(void 0, key), { key, value });
    }
  }
  async setAlarm(workflowId, wakeAt) {
    await _chunkKQO2TD7Tcjs.sleep.call(void 0, this.latency);
    this.alarms.set(workflowId, wakeAt);
  }
  async clearAlarm(workflowId) {
    await _chunkKQO2TD7Tcjs.sleep.call(void 0, this.latency);
    this.alarms.delete(workflowId);
  }
  async waitForMessages(messageNames, abortSignal) {
    const driver = this.messageDriver;
    if (driver.waitForMessages) {
      await driver.waitForMessages(messageNames, abortSignal);
      return;
    }
    while (true) {
      if (abortSignal.aborted) {
        throw new (0, _chunkKQO2TD7Tcjs.EvictedError)();
      }
      const messages = await this.messageDriver.receiveMessages({
        names: messageNames.length > 0 ? messageNames : void 0,
        count: 1,
        completable: true
      });
      if (messages.length > 0) {
        return;
      }
      await _chunkKQO2TD7Tcjs.sleep.call(void 0, Math.max(1, this.latency));
    }
  }
  /**
   * Get the alarm time for a workflow (for testing).
   */
  getAlarm(workflowId) {
    return this.alarms.get(workflowId);
  }
  /**
   * Check if any alarms are due and return their workflow IDs.
   */
  getDueAlarms() {
    const now = Date.now();
    const due = [];
    for (const [workflowId, wakeAt] of this.alarms) {
      if (wakeAt <= now) {
        due.push(workflowId);
      }
    }
    return due;
  }
  /**
   * Clear all data (for testing).
   */
  clear() {
    this.kv.clear();
    this.alarms.clear();
    this.#inMemoryMessageDriver.clear();
  }
  /**
   * Get a snapshot of all data (for testing/debugging).
   */
  snapshot() {
    const kvSnapshot = {};
    for (const [hexKey, entry] of this.kv) {
      kvSnapshot[hexKey] = entry.value;
    }
    return {
      kv: kvSnapshot,
      alarms: Object.fromEntries(this.alarms)
    };
  }
  /**
   * Get all hex-encoded keys (for testing).
   */
  keys() {
    return [...this.kv.keys()];
  }
}, _class);














































exports.CancelledError = _chunkKQO2TD7Tcjs.CancelledError; exports.CriticalError = _chunkKQO2TD7Tcjs.CriticalError; exports.DEFAULT_LOOP_COMMIT_INTERVAL = _chunkKQO2TD7Tcjs.DEFAULT_LOOP_COMMIT_INTERVAL; exports.DEFAULT_LOOP_HISTORY_EVERY = _chunkKQO2TD7Tcjs.DEFAULT_LOOP_HISTORY_EVERY; exports.DEFAULT_LOOP_HISTORY_KEEP = _chunkKQO2TD7Tcjs.DEFAULT_LOOP_HISTORY_KEEP; exports.DEFAULT_MAX_RETRIES = _chunkKQO2TD7Tcjs.DEFAULT_MAX_RETRIES; exports.DEFAULT_RETRY_BACKOFF_BASE = _chunkKQO2TD7Tcjs.DEFAULT_RETRY_BACKOFF_BASE; exports.DEFAULT_RETRY_BACKOFF_MAX = _chunkKQO2TD7Tcjs.DEFAULT_RETRY_BACKOFF_MAX; exports.DEFAULT_STEP_TIMEOUT = _chunkKQO2TD7Tcjs.DEFAULT_STEP_TIMEOUT; exports.EntryInProgressError = _chunkKQO2TD7Tcjs.EntryInProgressError; exports.EvictedError = _chunkKQO2TD7Tcjs.EvictedError; exports.HistoryDivergedError = _chunkKQO2TD7Tcjs.HistoryDivergedError; exports.InMemoryDriver = InMemoryDriver; exports.JoinError = _chunkKQO2TD7Tcjs.JoinError; exports.Loop = _chunkKQO2TD7Tcjs.Loop; exports.MessageWaitError = _chunkKQO2TD7Tcjs.MessageWaitError; exports.RaceError = _chunkKQO2TD7Tcjs.RaceError; exports.RollbackCheckpointError = _chunkKQO2TD7Tcjs.RollbackCheckpointError; exports.RollbackError = _chunkKQO2TD7Tcjs.RollbackError; exports.SleepError = _chunkKQO2TD7Tcjs.SleepError; exports.StepExhaustedError = _chunkKQO2TD7Tcjs.StepExhaustedError; exports.StepFailedError = _chunkKQO2TD7Tcjs.StepFailedError; exports.WorkflowContextImpl = _chunkKQO2TD7Tcjs.WorkflowContextImpl; exports.appendLoopIteration = _chunkKQO2TD7Tcjs.appendLoopIteration; exports.appendName = _chunkKQO2TD7Tcjs.appendName; exports.createEntry = _chunkKQO2TD7Tcjs.createEntry; exports.createHistorySnapshot = _chunkKQO2TD7Tcjs.createHistorySnapshot; exports.createStorage = _chunkKQO2TD7Tcjs.createStorage; exports.deleteEntriesWithPrefix = _chunkKQO2TD7Tcjs.deleteEntriesWithPrefix; exports.emptyLocation = _chunkKQO2TD7Tcjs.emptyLocation; exports.flush = _chunkKQO2TD7Tcjs.flush; exports.generateId = _chunkKQO2TD7Tcjs.generateId; exports.getEntry = _chunkKQO2TD7Tcjs.getEntry; exports.getOrCreateMetadata = _chunkKQO2TD7Tcjs.getOrCreateMetadata; exports.isLocationPrefix = _chunkKQO2TD7Tcjs.isLocationPrefix; exports.isLoopIterationMarker = _chunkKQO2TD7Tcjs.isLoopIterationMarker; exports.loadMetadata = _chunkKQO2TD7Tcjs.loadMetadata; exports.loadStorage = _chunkKQO2TD7Tcjs.loadStorage; exports.locationToKey = _chunkKQO2TD7Tcjs.locationToKey; exports.locationsEqual = _chunkKQO2TD7Tcjs.locationsEqual; exports.parentLocation = _chunkKQO2TD7Tcjs.parentLocation; exports.registerName = _chunkKQO2TD7Tcjs.registerName; exports.resolveName = _chunkKQO2TD7Tcjs.resolveName; exports.runWorkflow = _chunkKQO2TD7Tcjs.runWorkflow; exports.setEntry = _chunkKQO2TD7Tcjs.setEntry;
