// rivetkit/dist/tsup/actor/errors.js
import {
  ActionNotFound,
  ActionTimedOut,
  ActorAborted,
  ActorDuplicateKey,
  ActorError,
  ActorNotFound,
  ActorStopping,
  ConnStateNotEnabled,
  DatabaseNotEnabled,
  EndpointMismatch,
  EventPayloadInvalid,
  FeatureNotImplemented,
  Forbidden,
  INTERNAL_ERROR_CODE,
  INTERNAL_ERROR_DESCRIPTION,
  IncomingMessageTooLong,
  InternalError,
  InvalidActionRequest,
  InvalidCanPublishResponse,
  InvalidCanSubscribeResponse,
  InvalidEncoding,
  InvalidParams,
  InvalidQueryJSON,
  InvalidRequest,
  InvalidRequestHandlerResponse,
  InvalidStateType,
  MalformedMessage,
  MissingActorHeader,
  NamespaceMismatch,
  OutgoingMessageTooLong,
  ProxyError,
  QueueAlreadyCompleted,
  QueueCompleteNotConfigured,
  QueueCompletionPayloadInvalid,
  QueueFull,
  QueueMessageInvalid,
  QueueMessageTooLarge,
  QueuePayloadInvalid,
  QueuePreviousMessageNotCompleted,
  RequestHandlerNotDefined,
  RestrictedFeature,
  RouteNotFound,
  StateNotEnabled,
  USER_ERROR_CODE,
  Unreachable,
  Unsupported,
  UserError,
  VarsNotEnabled,
  WebSocketHandlerNotDefined,
  WebSocketsNotEnabled
} from "../chunk-7HTNH26M.js";
import "../chunk-424PT5DM.js";
export {
  ActionNotFound,
  ActionTimedOut,
  ActorAborted,
  ActorDuplicateKey,
  ActorError,
  ActorNotFound,
  ActorStopping,
  ConnStateNotEnabled,
  DatabaseNotEnabled,
  EndpointMismatch,
  EventPayloadInvalid,
  FeatureNotImplemented,
  Forbidden,
  INTERNAL_ERROR_CODE,
  INTERNAL_ERROR_DESCRIPTION,
  IncomingMessageTooLong,
  InternalError,
  InvalidActionRequest,
  InvalidCanPublishResponse,
  InvalidCanSubscribeResponse,
  InvalidEncoding,
  InvalidParams,
  InvalidQueryJSON,
  InvalidRequest,
  InvalidRequestHandlerResponse,
  InvalidStateType,
  MalformedMessage,
  MissingActorHeader,
  NamespaceMismatch,
  OutgoingMessageTooLong,
  ProxyError,
  QueueAlreadyCompleted,
  QueueCompleteNotConfigured,
  QueueCompletionPayloadInvalid,
  QueueFull,
  QueueMessageInvalid,
  QueueMessageTooLarge,
  QueuePayloadInvalid,
  QueuePreviousMessageNotCompleted,
  RequestHandlerNotDefined,
  RestrictedFeature,
  RouteNotFound,
  StateNotEnabled,
  USER_ERROR_CODE,
  Unreachable,
  Unsupported,
  UserError,
  VarsNotEnabled,
  WebSocketHandlerNotDefined,
  WebSocketsNotEnabled
};
