// rivetkit/dist/tsup/actor/errors.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


















































var _chunkL47L3ZWJcjs = require('../chunk-L47L3ZWJ.cjs');
require('../chunk-HDQ2JUQT.cjs');



















































exports.ActionNotFound = _chunkL47L3ZWJcjs.ActionNotFound; exports.ActionTimedOut = _chunkL47L3ZWJcjs.ActionTimedOut; exports.ActorAborted = _chunkL47L3ZWJcjs.ActorAborted; exports.ActorDuplicateKey = _chunkL47L3ZWJcjs.ActorDuplicateKey; exports.ActorError = _chunkL47L3ZWJcjs.ActorError; exports.ActorNotFound = _chunkL47L3ZWJcjs.ActorNotFound; exports.ActorStopping = _chunkL47L3ZWJcjs.ActorStopping; exports.ConnStateNotEnabled = _chunkL47L3ZWJcjs.ConnStateNotEnabled; exports.DatabaseNotEnabled = _chunkL47L3ZWJcjs.DatabaseNotEnabled; exports.EndpointMismatch = _chunkL47L3ZWJcjs.EndpointMismatch; exports.EventPayloadInvalid = _chunkL47L3ZWJcjs.EventPayloadInvalid; exports.FeatureNotImplemented = _chunkL47L3ZWJcjs.FeatureNotImplemented; exports.Forbidden = _chunkL47L3ZWJcjs.Forbidden; exports.INTERNAL_ERROR_CODE = _chunkL47L3ZWJcjs.INTERNAL_ERROR_CODE; exports.INTERNAL_ERROR_DESCRIPTION = _chunkL47L3ZWJcjs.INTERNAL_ERROR_DESCRIPTION; exports.IncomingMessageTooLong = _chunkL47L3ZWJcjs.IncomingMessageTooLong; exports.InternalError = _chunkL47L3ZWJcjs.InternalError; exports.InvalidActionRequest = _chunkL47L3ZWJcjs.InvalidActionRequest; exports.InvalidCanPublishResponse = _chunkL47L3ZWJcjs.InvalidCanPublishResponse; exports.InvalidCanSubscribeResponse = _chunkL47L3ZWJcjs.InvalidCanSubscribeResponse; exports.InvalidEncoding = _chunkL47L3ZWJcjs.InvalidEncoding; exports.InvalidParams = _chunkL47L3ZWJcjs.InvalidParams; exports.InvalidQueryJSON = _chunkL47L3ZWJcjs.InvalidQueryJSON; exports.InvalidRequest = _chunkL47L3ZWJcjs.InvalidRequest; exports.InvalidRequestHandlerResponse = _chunkL47L3ZWJcjs.InvalidRequestHandlerResponse; exports.InvalidStateType = _chunkL47L3ZWJcjs.InvalidStateType; exports.MalformedMessage = _chunkL47L3ZWJcjs.MalformedMessage; exports.MissingActorHeader = _chunkL47L3ZWJcjs.MissingActorHeader; exports.NamespaceMismatch = _chunkL47L3ZWJcjs.NamespaceMismatch; exports.OutgoingMessageTooLong = _chunkL47L3ZWJcjs.OutgoingMessageTooLong; exports.ProxyError = _chunkL47L3ZWJcjs.ProxyError; exports.QueueAlreadyCompleted = _chunkL47L3ZWJcjs.QueueAlreadyCompleted; exports.QueueCompleteNotConfigured = _chunkL47L3ZWJcjs.QueueCompleteNotConfigured; exports.QueueCompletionPayloadInvalid = _chunkL47L3ZWJcjs.QueueCompletionPayloadInvalid; exports.QueueFull = _chunkL47L3ZWJcjs.QueueFull; exports.QueueMessageInvalid = _chunkL47L3ZWJcjs.QueueMessageInvalid; exports.QueueMessageTooLarge = _chunkL47L3ZWJcjs.QueueMessageTooLarge; exports.QueuePayloadInvalid = _chunkL47L3ZWJcjs.QueuePayloadInvalid; exports.QueuePreviousMessageNotCompleted = _chunkL47L3ZWJcjs.QueuePreviousMessageNotCompleted; exports.RequestHandlerNotDefined = _chunkL47L3ZWJcjs.RequestHandlerNotDefined; exports.RestrictedFeature = _chunkL47L3ZWJcjs.RestrictedFeature; exports.RouteNotFound = _chunkL47L3ZWJcjs.RouteNotFound; exports.StateNotEnabled = _chunkL47L3ZWJcjs.StateNotEnabled; exports.USER_ERROR_CODE = _chunkL47L3ZWJcjs.USER_ERROR_CODE; exports.Unreachable = _chunkL47L3ZWJcjs.Unreachable; exports.Unsupported = _chunkL47L3ZWJcjs.Unsupported; exports.UserError = _chunkL47L3ZWJcjs.UserError; exports.VarsNotEnabled = _chunkL47L3ZWJcjs.VarsNotEnabled; exports.WebSocketHandlerNotDefined = _chunkL47L3ZWJcjs.WebSocketHandlerNotDefined; exports.WebSocketsNotEnabled = _chunkL47L3ZWJcjs.WebSocketsNotEnabled;
