// rivetkit/dist/tsup/test/mod.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _chunkZ2X3VVCScjs = require('../chunk-Z2X3VVCS.cjs');


var _chunkGRHLSP4Acjs = require('../chunk-GRHLSP4A.cjs');
require('../chunk-DSXC3S7Q.cjs');


var _chunkTTVNEKBEcjs = require('../chunk-TTVNEKBE.cjs');
require('../chunk-NIYZDWMW.cjs');
require('../chunk-GVBTEL2D.cjs');


var _chunk2Q5WS6QHcjs = require('../chunk-2Q5WS6QH.cjs');
require('../chunk-L47L3ZWJ.cjs');
require('../chunk-HDQ2JUQT.cjs');

// src/test/mod.ts
var _nodeserver = require('@hono/node-server');
var _nodews = require('@hono/node-ws');
var _invariant = require('invariant'); var _invariant2 = _interopRequireDefault(_invariant);

// src/test/log.ts
function logger() {
  return _chunk2Q5WS6QHcjs.getLogger.call(void 0, "test");
}

// src/test/mod.ts
async function setupTest(c, registry) {
  var _a;
  registry.config.test = { ...registry.config.test, enabled: true };
  const driver = _chunkZ2X3VVCScjs.createFileSystemOrMemoryDriver.call(void 0, 
    true,
    { path: `/tmp/rivetkit-test-${crypto.randomUUID()}` }
  );
  let upgradeWebSocket;
  registry.config.driver = driver;
  registry.config.inspector = {
    enabled: true,
    token: () => "token"
  };
  const parsedConfig = registry.parseConfig();
  const managerDriver = (_a = driver.manager) == null ? void 0 : _a.call(driver, parsedConfig);
  _invariant2.default.call(void 0, managerDriver, "missing manager driver");
  const getUpgradeWebSocket = () => upgradeWebSocket;
  managerDriver.setGetUpgradeWebSocket(getUpgradeWebSocket);
  const { router } = _chunkTTVNEKBEcjs.buildManagerRouter.call(void 0, 
    parsedConfig,
    managerDriver,
    getUpgradeWebSocket
  );
  const nodeWebSocket = _nodews.createNodeWebSocket.call(void 0, { app: router });
  upgradeWebSocket = nodeWebSocket.upgradeWebSocket;
  const server = _nodeserver.serve.call(void 0, {
    fetch: router.fetch,
    hostname: "127.0.0.1",
    port: 0
  });
  if (!server.listening) {
    await new Promise((resolve) => {
      server.once("listening", () => resolve());
    });
  }
  _invariant2.default.call(void 0, 
    nodeWebSocket.injectWebSocket !== void 0,
    "should have injectWebSocket"
  );
  nodeWebSocket.injectWebSocket(server);
  const address = server.address();
  _invariant2.default.call(void 0, address && typeof address !== "string", "missing server address");
  const port = address.port;
  const endpoint = `http://127.0.0.1:${port}`;
  logger().info({ msg: "test server listening", port });
  c.onTestFinished(async () => {
    await new Promise((resolve) => server.close(() => resolve(void 0)));
  });
  const client = _chunkGRHLSP4Acjs.createClient.call(void 0, {
    endpoint,
    namespace: "default",
    runnerName: "default",
    disableMetadataLookup: true
  });
  c.onTestFinished(async () => await client.dispose());
  return { client };
}


exports.setupTest = setupTest;
